# scRNAseq_teaching

## Introduction

This is the github repository for the “Hands-on guide to single cell RNA seq analysis in R computational workshop” series given to the AMA and Steventon labs on 4th June 2020.

## Viewing in your browser

Instead of running the code in RStudio, you can also view the code and the associated outputs in your browser. Select any .html file in the <strong>_book_html</strong> folder. This will lead you to the html output that we knitted from the RMarkdown files. 

## Objectives

- The volume of online resources on the single-cell field is vast. We hope that this short practical session will help point you in the right direction in conducting your own scRNA-seq data analysis by:
- Highlighting the relevant reading material, scRNA-seq packages and other resources available for further study.
- Sharing practical advice on dealing with issues that you may encounter in your own analyses.
- Providing you with code as starting material to tinker with.



## Description
We will be studying the scRNA-seq dataset of 18hpf zebrafish embryos, with a focus on examining the gene expression states of the tailbud Neuromesodermal progenitors.

Additionally, we will be taking a closer look at some key techniques, highlighting important details to keep in mind both when implementing and evaluating single-cell anaylsis.

We hope that this will provide a good starting point for getting to grips with this fast-moving field, and encourage you to contact us for further disucssion or advice if needed.

## Links
[Reference paper](https://science.sciencemag.org/content/360/6392/981.full)

[Dataset repository](https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3067194)

[Presentation on 4th June 2020](https://drive.google.com/drive/folders/1_LYbUTTfQOl_BaLyNdtOO5AxdgFkDYyq)

## Prerequisites

To follow the tutorial, you should have R (necessary) and Rstudio (Not necessary, but ideal) installed. Next, you should clone the git repository and then run the install.R file to 
download the necessary packages. Once you have all the packages installed on your local machine, you are ready to start tinkering with the code yourself. To get a quick overview of the output,
you can view the gitbook in the _book_html folder. 

## Created by:
- Kane Toh
- Giles Harper-Donnelly 



