# Resources
The start of a collection of resources that are useful for learning single-cell RNAseq analysis in R.

## Online Tutorials
### scRNA-seq
Sanger single cell course:
https://scrnaseq-course.cog.sanger.ac.uk/website/index.html.

Extensive introduction to sc analysis from some great people, updated regularly:
[Orchestrating Single-Cell Analysis with Bioconductor](https://osca.bioconductor.org/).

Quick and dirty way of analysing sc data:
[A step-by-step workflow for low-level analysis of single-cell RNA-seq data with Bioconductor](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC5112579/)

### R
As with many languages [Codeacademy](https://www.codecademy.com/) is a great place to start particularly the [Learn R series](https://www.codecademy.com/learn/learn-r).

Hadley Wickham's [R for Data Science](https://r4ds.had.co.nz/). Hadley is the author of a number of very popular R packages, particularly within the '[tidyverse](https://www.tidyverse.org)', this guide is a great introduction to some of the tidyverse tools for handling and visualising data in R.

## Papers
- 2018: RNA Sequencing Data: Hitchhiker's Guide to Expression Analysis
- 2019: Current best practices in single‐cell RNA‐seq analysis: a tutorial | Molecular Systems Biology


## R Packages 
### Analysing SC data
https://github.com/seandavi/awesome-single-cell

### Data
R package for sc mouse gastrulation data from the Marioni group: [MouseGastrulationData](https://bioconductor.org/packages/devel/data/experiment/html/MouseGastrulationData.html)
