<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 6 Differentially Expressed Genes | NMps_book.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 6 Differentially Expressed Genes | NMps_book.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 6 Differentially Expressed Genes | NMps_book.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="optimisations.html"/>
<link rel="next" href="subclustering.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="downloading-data.html"><a href="downloading-data.html"><i class="fa fa-check"></i><b>2</b> Downloading data</a></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a><ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#importing-libraries"><i class="fa fa-check"></i><b>3.2</b> Importing libraries</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#importing-the-data"><i class="fa fa-check"></i><b>3.3</b> Importing the data</a></li>
<li class="chapter" data-level="3.4" data-path="quality-control.html"><a href="quality-control.html#preparing-the-column-metadata"><i class="fa fa-check"></i><b>3.4</b> Preparing the column metadata</a></li>
<li class="chapter" data-level="3.5" data-path="quality-control.html"><a href="quality-control.html#preparing-the-row-metadata"><i class="fa fa-check"></i><b>3.5</b> Preparing the row metadata</a></li>
<li class="chapter" data-level="3.6" data-path="quality-control.html"><a href="quality-control.html#constructing-the-single-cell-experiment-object"><i class="fa fa-check"></i><b>3.6</b> Constructing the single cell experiment object</a></li>
<li class="chapter" data-level="3.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-with-scater"><i class="fa fa-check"></i><b>3.7</b> Quality control with scater</a></li>
<li class="chapter" data-level="3.8" data-path="quality-control.html"><a href="quality-control.html#saving-the-rds-object"><i class="fa fa-check"></i><b>3.8</b> Saving the RDS object</a></li>
<li class="chapter" data-level="3.9" data-path="quality-control.html"><a href="quality-control.html#session-info"><i class="fa fa-check"></i><b>3.9</b> Session Info</a></li>
<li class="chapter" data-level="3.10" data-path="quality-control.html"><a href="quality-control.html#solutions"><i class="fa fa-check"></i><b>3.10</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>4</b> Clustering</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="clustering.html"><a href="clustering.html#importing-libraries-1"><i class="fa fa-check"></i><b>4.2</b> Importing libraries</a></li>
<li class="chapter" data-level="4.3" data-path="clustering.html"><a href="clustering.html#importing-data"><i class="fa fa-check"></i><b>4.3</b> Importing Data</a></li>
<li class="chapter" data-level="4.4" data-path="clustering.html"><a href="clustering.html#useful-seurat-object-interaction-methods"><i class="fa fa-check"></i><b>4.4</b> Useful seurat object interaction methods</a></li>
<li class="chapter" data-level="4.5" data-path="clustering.html"><a href="clustering.html#performing-dimensional-reduction"><i class="fa fa-check"></i><b>4.5</b> Performing dimensional reduction</a></li>
<li class="chapter" data-level="4.6" data-path="clustering.html"><a href="clustering.html#clustering-1"><i class="fa fa-check"></i><b>4.6</b> Clustering</a><ul>
<li class="chapter" data-level="4.6.1" data-path="clustering.html"><a href="clustering.html#constructing-a-shared-nearest-neighbour-snn-graph"><i class="fa fa-check"></i><b>4.6.1</b> Constructing a Shared-nearest neighbour (SNN) graph</a></li>
<li class="chapter" data-level="4.6.2" data-path="clustering.html"><a href="clustering.html#clustering-with-the-louvain-algorithm"><i class="fa fa-check"></i><b>4.6.2</b> Clustering with the Louvain algorithm</a></li>
<li class="chapter" data-level="4.6.3" data-path="clustering.html"><a href="clustering.html#resources"><i class="fa fa-check"></i><b>4.6.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="clustering.html"><a href="clustering.html#umap-with-authors-annotations"><i class="fa fa-check"></i><b>4.7</b> UMAP with authors’ annotations</a></li>
<li class="chapter" data-level="4.8" data-path="clustering.html"><a href="clustering.html#t-sne-with-authors-annotations"><i class="fa fa-check"></i><b>4.8</b> t-SNE with authors’ annotations</a></li>
<li class="chapter" data-level="4.9" data-path="clustering.html"><a href="clustering.html#assigning-our-own-cluster-identities"><i class="fa fa-check"></i><b>4.9</b> Assigning our own cluster identities</a></li>
<li class="chapter" data-level="4.10" data-path="clustering.html"><a href="clustering.html#umap-visualisation-of-our-annotations"><i class="fa fa-check"></i><b>4.10</b> UMAP visualisation of our annotations</a></li>
<li class="chapter" data-level="4.11" data-path="clustering.html"><a href="clustering.html#highlighting-tailbud-cells"><i class="fa fa-check"></i><b>4.11</b> Highlighting tailbud cells</a></li>
<li class="chapter" data-level="4.12" data-path="clustering.html"><a href="clustering.html#saving-rds-object"><i class="fa fa-check"></i><b>4.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="4.13" data-path="clustering.html"><a href="clustering.html#session-info-1"><i class="fa fa-check"></i><b>4.13</b> Session Info</a></li>
<li class="chapter" data-level="4.14" data-path="clustering.html"><a href="clustering.html#solutions-1"><i class="fa fa-check"></i><b>4.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="optimisations.html"><a href="optimisations.html"><i class="fa fa-check"></i><b>5</b> Optimisations</a><ul>
<li class="chapter" data-level="5.1" data-path="optimisations.html"><a href="optimisations.html#importing-libraries-2"><i class="fa fa-check"></i><b>5.1</b> Importing libraries</a></li>
<li class="chapter" data-level="5.2" data-path="optimisations.html"><a href="optimisations.html#importing-data-1"><i class="fa fa-check"></i><b>5.2</b> Importing Data</a></li>
<li class="chapter" data-level="5.3" data-path="optimisations.html"><a href="optimisations.html#number-of-pcs-to-retain"><i class="fa fa-check"></i><b>5.3</b> Number of PCs to retain</a><ul>
<li class="chapter" data-level="5.3.1" data-path="optimisations.html"><a href="optimisations.html#scree-plot"><i class="fa fa-check"></i><b>5.3.1</b> Scree plot</a></li>
<li class="chapter" data-level="5.3.2" data-path="optimisations.html"><a href="optimisations.html#examining-technical-noise"><i class="fa fa-check"></i><b>5.3.2</b> Examining technical noise</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="optimisations.html"><a href="optimisations.html#evaluating-clustering-results"><i class="fa fa-check"></i><b>5.4</b> Evaluating clustering results</a><ul>
<li class="chapter" data-level="5.4.1" data-path="optimisations.html"><a href="optimisations.html#cluster-stability"><i class="fa fa-check"></i><b>5.4.1</b> Cluster stability</a></li>
<li class="chapter" data-level="5.4.2" data-path="optimisations.html"><a href="optimisations.html#cluster-separability"><i class="fa fa-check"></i><b>5.4.2</b> Cluster separability</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="optimisations.html"><a href="optimisations.html#comparison-with-authors-labels"><i class="fa fa-check"></i><b>5.5</b> Comparison with authors’ labels</a><ul>
<li class="chapter" data-level="5.5.1" data-path="optimisations.html"><a href="optimisations.html#concordance-in-tailbud-cluster-assignment"><i class="fa fa-check"></i><b>5.5.1</b> Concordance in tailbud cluster assignment</a></li>
<li class="chapter" data-level="5.5.2" data-path="optimisations.html"><a href="optimisations.html#optimising-the-seurat-clustering-parameters"><i class="fa fa-check"></i><b>5.5.2</b> Optimising the seurat clustering parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="optimisations.html"><a href="optimisations.html#selecting-cells-for-downstream-analysis"><i class="fa fa-check"></i><b>5.6</b> Selecting cells for downstream analysis</a></li>
<li class="chapter" data-level="5.7" data-path="optimisations.html"><a href="optimisations.html#session-info-2"><i class="fa fa-check"></i><b>5.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html"><i class="fa fa-check"></i><b>6</b> Differentially Expressed Genes</a><ul>
<li class="chapter" data-level="6.1" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualisation-of-marker-genes-in-our-dataset"><i class="fa fa-check"></i><b>6.1</b> Visualisation of marker genes in our dataset</a></li>
<li class="chapter" data-level="6.2" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#obtaining-a-list-of-differentially-expressed-genes"><i class="fa fa-check"></i><b>6.2</b> Obtaining a list of differentially expressed genes</a></li>
<li class="chapter" data-level="6.3" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#analysing-output-of-the-wilcoxon-test"><i class="fa fa-check"></i><b>6.3</b> Analysing output of the wilcoxon test</a></li>
<li class="chapter" data-level="6.4" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualising-the-expression-of-the-marker-genes"><i class="fa fa-check"></i><b>6.4</b> Visualising the expression of the marker genes</a></li>
<li class="chapter" data-level="6.5" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#session-info-3"><i class="fa fa-check"></i><b>6.5</b> Session Info</a></li>
<li class="chapter" data-level="6.6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#solutions-2"><i class="fa fa-check"></i><b>6.6</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="subclustering.html"><a href="subclustering.html"><i class="fa fa-check"></i><b>7</b> Subclustering</a><ul>
<li class="chapter" data-level="7.1" data-path="subclustering.html"><a href="subclustering.html#importing-libraries-3"><i class="fa fa-check"></i><b>7.1</b> Importing libraries</a></li>
<li class="chapter" data-level="7.2" data-path="subclustering.html"><a href="subclustering.html#importing-data-2"><i class="fa fa-check"></i><b>7.2</b> Importing Data</a></li>
<li class="chapter" data-level="7.3" data-path="subclustering.html"><a href="subclustering.html#subclustering-the-tb-cells---using-only-cells-that-are-assigned-to-the-shared-cluster-across-all-the-clustering-resolutions."><i class="fa fa-check"></i><b>7.3</b> Subclustering the TB cells - Using only cells that are assigned to the ‘shared’ cluster across all the clustering resolutions.</a></li>
<li class="chapter" data-level="7.4" data-path="subclustering.html"><a href="subclustering.html#dimensional-reduction-with-umap-and-clustering"><i class="fa fa-check"></i><b>7.4</b> Dimensional reduction with UMAP and clustering</a></li>
<li class="chapter" data-level="7.5" data-path="subclustering.html"><a href="subclustering.html#identifying-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.5</b> Identifying marker genes for each cluster</a></li>
<li class="chapter" data-level="7.6" data-path="subclustering.html"><a href="subclustering.html#merging-clusters-that-do-not-appear-distinct"><i class="fa fa-check"></i><b>7.6</b> Merging clusters that do not appear distinct</a></li>
<li class="chapter" data-level="7.7" data-path="subclustering.html"><a href="subclustering.html#plotting-umap-and-assigning-cluster-labels"><i class="fa fa-check"></i><b>7.7</b> Plotting UMAP and assigning cluster labels</a></li>
<li class="chapter" data-level="7.8" data-path="subclustering.html"><a href="subclustering.html#checking-the-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.8</b> Checking the marker genes for each cluster</a></li>
<li class="chapter" data-level="7.9" data-path="subclustering.html"><a href="subclustering.html#sox2-tbxta-coexpression"><i class="fa fa-check"></i><b>7.9</b> Sox2 tbxta coexpression</a></li>
<li class="chapter" data-level="7.10" data-path="subclustering.html"><a href="subclustering.html#viewing-genes-of-interest-with-a-dotplot"><i class="fa fa-check"></i><b>7.10</b> Viewing genes of interest with a dotplot</a></li>
<li class="chapter" data-level="7.11" data-path="subclustering.html"><a href="subclustering.html#exporting-the-violin-plots-for-genes-of-interest."><i class="fa fa-check"></i><b>7.11</b> Exporting the violin plots for genes of interest.</a></li>
<li class="chapter" data-level="7.12" data-path="subclustering.html"><a href="subclustering.html#saving-rds-object-1"><i class="fa fa-check"></i><b>7.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="7.13" data-path="subclustering.html"><a href="subclustering.html#session-info-4"><i class="fa fa-check"></i><b>7.13</b> Session Info</a></li>
<li class="chapter" data-level="7.14" data-path="subclustering.html"><a href="subclustering.html#solutions-3"><i class="fa fa-check"></i><b>7.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="pseudotime.html"><a href="pseudotime.html"><i class="fa fa-check"></i><b>8</b> Pseudotime</a><ul>
<li class="chapter" data-level="8.1" data-path="pseudotime.html"><a href="pseudotime.html#importing-libraries-4"><i class="fa fa-check"></i><b>8.1</b> Importing libraries</a></li>
<li class="chapter" data-level="8.2" data-path="pseudotime.html"><a href="pseudotime.html#importing-data-3"><i class="fa fa-check"></i><b>8.2</b> Importing data</a></li>
<li class="chapter" data-level="8.3" data-path="pseudotime.html"><a href="pseudotime.html#tradeseq-workflow-analyses"><i class="fa fa-check"></i><b>8.3</b> TradeSeq workflow analyses</a></li>
<li class="chapter" data-level="8.4" data-path="pseudotime.html"><a href="pseudotime.html#session-info-5"><i class="fa fa-check"></i><b>8.4</b> Session Info</a></li>
<li class="chapter" data-level="8.5" data-path="pseudotime.html"><a href="pseudotime.html#solutions-4"><i class="fa fa-check"></i><b>8.5</b> Solutions</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="differentially-expressed-genes" class="section level1">
<h1><span class="header-section-number">Chapter 6</span> Differentially Expressed Genes</h1>
<div id="visualisation-of-marker-genes-in-our-dataset" class="section level2">
<h2><span class="header-section-number">6.1</span> Visualisation of marker genes in our dataset</h2>
<p>Genes that are differentially expressed between clusters drive the separation of the clusters. We call these genes the <code>marker genes</code> for the cluster. If the clusters correspond to our biologically meaningful notion of a cell type, then these marker genes are akin to the cell-type specific markers that identifies these cells. By using our existing biological knowledge, we can annotate the function of these genes and use this to assign a cell type identity for each cluster. Alternatively, we can obtain a list of marker genes and check for novel genes that mark the cell type.</p>
<p>We can easily assess the accuracy of our clustering by looking at the distribution of known marker genes across the different clusters. For instance, <code>tbx16</code> and <code>msgn1</code> mark the nascent presomitic mesoderm whilst <code>sox2</code> and <code>sox3</code> are the neuronal markers.</p>
<div class="sourceCode" id="cb63"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb63-1" data-line-number="1"><span class="kw">library</span>(Seurat)</a>
<a class="sourceLine" id="cb63-2" data-line-number="2"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb63-3" data-line-number="3"><span class="kw">library</span>(gridExtra)</a>
<a class="sourceLine" id="cb63-4" data-line-number="4"><span class="kw">library</span>(scales)</a>
<a class="sourceLine" id="cb63-5" data-line-number="5"><span class="kw">library</span>(extrafont) </a>
<a class="sourceLine" id="cb63-6" data-line-number="6"><span class="kw">loadfonts</span>(<span class="dt">quiet =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<p><em>Q1) What other markers would be interesting to visualise for this dataset?</em></p>
<div class="sourceCode" id="cb64"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb64-1" data-line-number="1">hpf18_seurat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;rds/hpf18_seurat_dimred_new.rds&quot;</span>)</a>
<a class="sourceLine" id="cb64-2" data-line-number="2">hpf18_seurat&lt;-<span class="st"> </span><span class="kw">SetIdent</span>(hpf18_seurat, <span class="dt">value =</span> hpf18_seurat<span class="op">@</span>meta.data<span class="op">$</span>seurat_clusters_names)</a>
<a class="sourceLine" id="cb64-3" data-line-number="3"></a>
<a class="sourceLine" id="cb64-4" data-line-number="4">markergenes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;tbx16&quot;</span>, <span class="st">&quot;msgn1&quot;</span>, <span class="st">&quot;ta&quot;</span>, <span class="st">&quot;sox2&quot;</span>, <span class="st">&quot;sox3&quot;</span>, <span class="st">&quot;pax6a&quot;</span>, <span class="st">&quot;krt8&quot;</span>, <span class="st">&quot;pfn1&quot;</span>, <span class="st">&quot;krt4&quot;</span>, <span class="st">&quot;pitx1&quot;</span>, <span class="st">&quot;hand2&quot;</span>, <span class="st">&quot;entpd1&quot;</span>)</a>
<a class="sourceLine" id="cb64-5" data-line-number="5"></a>
<a class="sourceLine" id="cb64-6" data-line-number="6"><span class="co">#Visualizing marker genes </span></a>
<a class="sourceLine" id="cb64-7" data-line-number="7"><span class="kw">FeaturePlot</span>(<span class="dt">object =</span> hpf18_seurat, <span class="dt">features=</span> markergenes, <span class="dt">cols =</span> colorspace<span class="op">::</span><span class="kw">sequential_hcl</span>(<span class="dv">4</span>, <span class="st">&quot;Viridis&quot;</span>), <span class="dt">reduction=</span> <span class="st">&quot;umap&quot;</span>, <span class="dt">label =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<div class="figure"><span id="fig:clusteringAnalysis"></span>
<img src="NMps_book_files/figure-html/clusteringAnalysis-1.png" alt="Figure 1- UMAP visualisation of cells labelled with the expression of marker genes." width="1440" />
<p class="caption">
Figure 6.1: Figure 1- UMAP visualisation of cells labelled with the expression of marker genes.
</p>
</div>
<p><em>Q2) What additional information does the violin plot provide over the feature plot? How does the violin plot differ from a box plot?</em></p>
<div class="sourceCode" id="cb65"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb65-1" data-line-number="1"><span class="kw">VlnPlot</span>(<span class="dt">object =</span> hpf18_seurat, <span class="dt">features =</span> markergenes, <span class="dt">ncol =</span> <span class="dv">3</span>)</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/all_violinplot-1.png" alt="Figure 2- Violin plot of the marker genes." width="1440" />
<p class="caption">
(#fig:all_violinplot)Figure 2- Violin plot of the marker genes.
</p>
</div>
</div>
<div id="obtaining-a-list-of-differentially-expressed-genes" class="section level2">
<h2><span class="header-section-number">6.2</span> Obtaining a list of differentially expressed genes</h2>
<p>Seurat’s <code>FindAllMarkers</code> function allows us to easily compute these DEGs. The important arguments in this function are:</p>
<ol style="list-style-type: decimal">
<li>only.pos: Setting this to TRUE returns only the marker genes that are upregulated in the cluster.
2.test: This determines the statistical test for differential expression.</li>
<li>min.pct: The default is 0.25. This means that it will only test genes that are detected in a minimum fraction of 0.25 cells in either of the 2 clusters being compared.</li>
</ol>
<p>For a complete list, see the <a href="https://www.rdocumentation.org/packages/Seurat/versions/3.1.4/topics/FindAllMarkers">documentation</a></p>
<p><em>Q3) How would you go about comparing a cluster of interest with another cluster?</em></p>
<div class="sourceCode" id="cb66"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb66-1" data-line-number="1"><span class="co">#Obtaining DEGs</span></a>
<a class="sourceLine" id="cb66-2" data-line-number="2"></a>
<a class="sourceLine" id="cb66-3" data-line-number="3">cells.markers.wilcox &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(<span class="dt">object =</span> hpf18_seurat, <span class="dt">only.pos =</span> <span class="ot">TRUE</span>, <span class="dt">test =</span> <span class="st">&quot;wilcox&quot;</span>, <span class="dt">min.pct =</span> <span class="fl">0.25</span>, </a>
<a class="sourceLine" id="cb66-4" data-line-number="4">    <span class="dt">thresh.use =</span> <span class="fl">0.25</span>)</a></code></pre></div>
<pre><code>## Calculating cluster 18hpf-hatching gland</code></pre>
<pre><code>## Calculating cluster 18hpf-neural crest - grem2</code></pre>
<pre><code>## Calculating cluster 18hpf-differentiating neurons</code></pre>
<pre><code>## Calculating cluster 18hpf-tailbud - spinal cord</code></pre>
<pre><code>## Calculating cluster 18hpf-epidermal - anterior</code></pre>
<pre><code>## Calculating cluster 18hpf-neural - fore-midbrain</code></pre>
<pre><code>## Calculating cluster 18hpf-pharyngeal arch</code></pre>
<pre><code>## Calculating cluster 18hpf-myotome</code></pre>
<pre><code>## Calculating cluster 18hpf-tailbud - PSM</code></pre>
<pre><code>## Calculating cluster 18hpf-neural - ventral hindbrain</code></pre>
<pre><code>## Calculating cluster 18hpf-epidermal - pfn1</code></pre>
<pre><code>## Calculating cluster 18hpf-epidermal - sox4b</code></pre>
<pre><code>## Calculating cluster 18hpf-neural - mid-hindbrain</code></pre>
<pre><code>## Calculating cluster 18hpf-erythroid</code></pre>
<pre><code>## Calculating cluster 18hpf-neural - telencephalon</code></pre>
<pre><code>## Calculating cluster 18hpf-macrophage</code></pre>
<pre><code>## Calculating cluster 18hpf-neural - midbrain</code></pre>
<pre><code>## Calculating cluster 18hpf-endothelial</code></pre>
<div class="sourceCode" id="cb85"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb85-1" data-line-number="1"><span class="kw">write.table</span>(<span class="dt">file =</span> <span class="st">&quot;hpf18_wilcoxon.txt&quot;</span>, cells.markers.wilcox, <span class="dt">sep =</span> <span class="st">&quot;</span><span class="ch">\t</span><span class="st">&quot;</span>)</a>
<a class="sourceLine" id="cb85-2" data-line-number="2"></a>
<a class="sourceLine" id="cb85-3" data-line-number="3"><span class="kw">View</span>(cells.markers.wilcox)</a></code></pre></div>
<p><em>Q4) Have a go at customizing the heatmap. How would you quickly check for the arguments in the function? </em></p>
<div class="sourceCode" id="cb86"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb86-1" data-line-number="1">top3_wilcox &lt;-<span class="st"> </span>cells.markers.wilcox <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">3</span>, avg_logFC)</a>
<a class="sourceLine" id="cb86-2" data-line-number="2"><span class="kw">DoHeatmap</span>(<span class="dt">object =</span> hpf18_seurat, <span class="dt">features =</span> top3_wilcox<span class="op">$</span>gene, <span class="dt">label =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/DEG%20heatmap-1.png" alt="Figure 3- Heatmap of differentially expressed genes in each cluster of the single-cell data. " width="1152" />
<p class="caption">
(#fig:DEG heatmap)Figure 3- Heatmap of differentially expressed genes in each cluster of the single-cell data.
</p>
</div>
</div>
<div id="analysing-output-of-the-wilcoxon-test" class="section level2">
<h2><span class="header-section-number">6.3</span> Analysing output of the wilcoxon test</h2>
<p>Text below taken from Seurat’s <a href="https://satijalab.org/seurat/v3.0/de_vignette.html">differential expression vignette</a></p>
<p>The results data frame has the following columns:</p>
<ul>
<li>p_val : p_val (unadjusted)
-bavg_logFC : log fold-chage of the average expression between the two groups. Positive values indicate that the feature is more highly expressed in the first group.</li>
<li>pct.1 : The percentage of cells where the feature is detected in the first group</li>
<li>pct.2 : The percentage of cells where the feature is detected in the second group</li>
<li>p_val_adj : Adjusted p-value, based on bonferroni correction using all features in the dataset.</li>
</ul>
<div class="sourceCode" id="cb87"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb87-1" data-line-number="1">wilcoxon &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;hpf18_wilcoxon.txt&quot;</span>)</a>
<a class="sourceLine" id="cb87-2" data-line-number="2">wilcoxon &lt;-<span class="st"> </span>wilcoxon <span class="op">%&gt;%</span><span class="st"> </span>dplyr<span class="op">::</span><span class="kw">select</span>(gene, cluster, p_val_adj, avg_logFC) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb87-3" data-line-number="3"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">test =</span> <span class="st">&quot;wilcoxon&quot;</span>) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb87-4" data-line-number="4"><span class="st">  </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb87-5" data-line-number="5"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">arrange</span>(cluster, <span class="kw">desc</span>(avg_logFC))<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb87-6" data-line-number="6"><span class="st">  </span><span class="kw">top_n</span>(<span class="dv">20</span>, avg_logFC)</a>
<a class="sourceLine" id="cb87-7" data-line-number="7"></a>
<a class="sourceLine" id="cb87-8" data-line-number="8"><span class="kw">write.csv</span>(wilcoxon, <span class="dt">file =</span> <span class="st">&quot;wilcoxon_dim25.csv&quot;</span>)</a></code></pre></div>
</div>
<div id="visualising-the-expression-of-the-marker-genes" class="section level2">
<h2><span class="header-section-number">6.4</span> Visualising the expression of the marker genes</h2>
<p>Figure 2b) of Cao et al., 2019- Cluster vs marker gene dot plot</p>
<div class="sourceCode" id="cb88"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb88-1" data-line-number="1">hpf18_seurat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;rds/hpf18_seurat_dimred_new.rds&quot;</span>)</a>
<a class="sourceLine" id="cb88-2" data-line-number="2">wilcoxon &lt;-<span class="st"> </span><span class="kw">read.table</span>(<span class="st">&quot;hpf18_wilcoxon.txt&quot;</span>)</a>
<a class="sourceLine" id="cb88-3" data-line-number="3"></a>
<a class="sourceLine" id="cb88-4" data-line-number="4"><span class="kw">Idents</span>(hpf18_seurat) &lt;-<span class="st"> &quot;seurat_clusters_names&quot;</span></a>
<a class="sourceLine" id="cb88-5" data-line-number="5"></a>
<a class="sourceLine" id="cb88-6" data-line-number="6">features &lt;-<span class="st"> </span>wilcoxon <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb88-7" data-line-number="7"><span class="st">  </span><span class="kw">filter</span>(<span class="op">!</span>(<span class="kw">str_detect</span>(gene, <span class="st">&quot;si:&quot;</span>)<span class="op">|</span><span class="kw">str_detect</span>(gene, <span class="st">&quot;wu:&quot;</span>)<span class="op">|</span><span class="kw">str_detect</span>(gene, <span class="st">&quot;LOC&quot;</span>)<span class="op">|</span><span class="kw">str_detect</span>(gene, <span class="st">&quot;MGC&quot;</span>)<span class="op">|</span><span class="kw">str_detect</span>(gene, <span class="st">&quot;zgc:&quot;</span>))) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb88-8" data-line-number="8"><span class="st">  </span><span class="kw">group_by</span>(cluster)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb88-9" data-line-number="9"><span class="st">  </span><span class="kw">top_n</span>(<span class="dt">n=</span><span class="op">-</span><span class="dv">1</span>, <span class="dt">wt=</span> p_val_adj) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb88-10" data-line-number="10"><span class="st">  </span><span class="kw">top_n</span>(<span class="dt">n=</span><span class="dv">1</span>, <span class="dt">wt=</span>avg_logFC) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb88-11" data-line-number="11"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(cluster,gene)</a>
<a class="sourceLine" id="cb88-12" data-line-number="12"></a>
<a class="sourceLine" id="cb88-13" data-line-number="13">a &lt;-<span class="st"> </span><span class="kw">DotPlot</span>(hpf18_seurat, <span class="dt">features =</span> features<span class="op">$</span>gene) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb88-14" data-line-number="14"><span class="st">  </span><span class="co">#RotatedAxis() + </span></a>
<a class="sourceLine" id="cb88-15" data-line-number="15"><span class="st">  </span><span class="kw">scale_color_gradient2</span>(<span class="dt">low =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">mid =</span> <span class="st">&quot;white&quot;</span>,<span class="dt">high =</span> <span class="kw">muted</span>(<span class="st">&quot;blue&quot;</span>), <span class="dt">midpoint =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb88-16" data-line-number="16"><span class="st">  </span><span class="kw">guides</span>(<span class="dt">size =</span> <span class="kw">guide_legend</span>(<span class="dt">title =</span> <span class="st">&#39;Percent of cells in cluster expressing feature&#39;</span>))<span class="op">+</span></a>
<a class="sourceLine" id="cb88-17" data-line-number="17"><span class="st">  </span><span class="kw">theme_bw</span>(<span class="dt">base_size=</span><span class="dv">14</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb88-18" data-line-number="18"><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;Cluster Name&quot;</span>)<span class="op">+</span></a>
<a class="sourceLine" id="cb88-19" data-line-number="19"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.text=</span><span class="kw">element_text</span>(<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>), <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb88-20" data-line-number="20">        <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>),</a>
<a class="sourceLine" id="cb88-21" data-line-number="21">        <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust=</span><span class="dv">1</span>)) </a></code></pre></div>
<pre><code>## Scale for &#39;colour&#39; is already present. Adding another scale for &#39;colour&#39;,
## which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb90"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb90-1" data-line-number="1">a</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/dotplot_all-1.png" alt="Figure 4- Dot plot displaying the top marker gene for each cluster." width="1440" />
<p class="caption">
(#fig:dotplot_all)Figure 4- Dot plot displaying the top marker gene for each cluster.
</p>
</div>
<div class="sourceCode" id="cb91"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb91-1" data-line-number="1"><span class="co">#unique(wilcoxon$cluster)</span></a>
<a class="sourceLine" id="cb91-2" data-line-number="2">TB_features &lt;-<span class="st"> </span>wilcoxon <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb91-3" data-line-number="3"><span class="st">  </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb91-4" data-line-number="4"><span class="st">  </span><span class="kw">filter</span>(cluster <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - spinal cord&quot;</span> <span class="op">|</span><span class="st"> </span>cluster <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - PSM&quot;</span> ) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb91-5" data-line-number="5"><span class="st">  </span><span class="kw">top_n</span>(<span class="dt">n=</span><span class="op">-</span><span class="dv">10</span>, <span class="dt">wt=</span> p_val_adj) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb91-6" data-line-number="6"><span class="st">  </span><span class="kw">top_n</span>(<span class="dt">n=</span><span class="op">-</span><span class="dv">10</span>, <span class="dt">wt=</span>avg_logFC) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb91-7" data-line-number="7"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(cluster,gene)</a>
<a class="sourceLine" id="cb91-8" data-line-number="8"></a>
<a class="sourceLine" id="cb91-9" data-line-number="9">b &lt;-<span class="st"> </span><span class="kw">DotPlot</span>(hpf18_seurat, <span class="dt">features =</span> TB_features<span class="op">$</span>gene) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb91-10" data-line-number="10"><span class="st">  </span><span class="kw">RotatedAxis</span>() <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb91-11" data-line-number="11"><span class="st">  </span><span class="kw">scale_color_gradient2</span>(<span class="dt">low =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">mid =</span> <span class="st">&quot;white&quot;</span>,<span class="dt">high =</span> <span class="kw">muted</span>(<span class="st">&quot;blue&quot;</span>), <span class="dt">midpoint =</span> <span class="dv">0</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb91-12" data-line-number="12"><span class="st">  </span><span class="kw">guides</span>(<span class="dt">size =</span> <span class="kw">guide_legend</span>(<span class="dt">title =</span> <span class="st">&#39;Percent of cells in cluster expressing feature&#39;</span>))<span class="op">+</span></a>
<a class="sourceLine" id="cb91-13" data-line-number="13"><span class="st">  </span><span class="kw">theme_bw</span>(<span class="dt">base_size=</span><span class="dv">14</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb91-14" data-line-number="14"><span class="st">  </span><span class="kw">ylab</span>(<span class="st">&quot;Cluster Name&quot;</span>)<span class="op">+</span></a>
<a class="sourceLine" id="cb91-15" data-line-number="15"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.text=</span><span class="kw">element_text</span>(<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>), <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb91-16" data-line-number="16">        <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>),</a>
<a class="sourceLine" id="cb91-17" data-line-number="17">        <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle =</span> <span class="dv">90</span>, <span class="dt">hjust=</span><span class="dv">1</span>)) </a></code></pre></div>
<pre><code>## Scale for &#39;colour&#39; is already present. Adding another scale for &#39;colour&#39;,
## which will replace the existing scale.</code></pre>
<div class="sourceCode" id="cb93"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb93-1" data-line-number="1">b</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/dotplot_TB%20-1.png" alt="Figure 5- Dot plot showing the distribution of the tailbud marker genes across all clusters" width="1440" />
<p class="caption">
(#fig:dotplot_TB )Figure 5- Dot plot showing the distribution of the tailbud marker genes across all clusters
</p>
</div>
<div class="sourceCode" id="cb94"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb94-1" data-line-number="1"><span class="co">#DoHeatmap(object = hpf18_seurat_cc, slot = &quot;scale.data&quot;, assay = &quot;SCT&quot;, features = TB_features$gene, size = 3) + RotatedAxis() </span></a></code></pre></div>
</div>
<div id="session-info-3" class="section level2">
<h2><span class="header-section-number">6.5</span> Session Info</h2>
<details>
<p><summary>View Session Info</summary></p>
<div class="sourceCode" id="cb95"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb95-1" data-line-number="1">devtools<span class="op">::</span><span class="kw">session_info</span>()</a>
<a class="sourceLine" id="cb95-2" data-line-number="2"><span class="co">## - Session info ---------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb95-3" data-line-number="3"><span class="co">##  setting  value                       </span></a>
<a class="sourceLine" id="cb95-4" data-line-number="4"><span class="co">##  version  R version 4.0.0 (2020-04-24)</span></a>
<a class="sourceLine" id="cb95-5" data-line-number="5"><span class="co">##  os       Windows 10 x64              </span></a>
<a class="sourceLine" id="cb95-6" data-line-number="6"><span class="co">##  system   x86_64, mingw32             </span></a>
<a class="sourceLine" id="cb95-7" data-line-number="7"><span class="co">##  ui       RTerm                       </span></a>
<a class="sourceLine" id="cb95-8" data-line-number="8"><span class="co">##  language (EN)                        </span></a>
<a class="sourceLine" id="cb95-9" data-line-number="9"><span class="co">##  collate  English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb95-10" data-line-number="10"><span class="co">##  ctype    English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb95-11" data-line-number="11"><span class="co">##  tz       Asia/Kuala_Lumpur           </span></a>
<a class="sourceLine" id="cb95-12" data-line-number="12"><span class="co">##  date     2020-06-03                  </span></a>
<a class="sourceLine" id="cb95-13" data-line-number="13"><span class="co">## </span></a>
<a class="sourceLine" id="cb95-14" data-line-number="14"><span class="co">## - Packages -------------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb95-15" data-line-number="15"><span class="co">##  package              * version  date       lib source        </span></a>
<a class="sourceLine" id="cb95-16" data-line-number="16"><span class="co">##  AnnotationDbi        * 1.50.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-17" data-line-number="17"><span class="co">##  ape                    5.3      2019-03-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-18" data-line-number="18"><span class="co">##  assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-19" data-line-number="19"><span class="co">##  backports              1.1.6    2020-04-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-20" data-line-number="20"><span class="co">##  beeswarm               0.2.3    2016-04-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-21" data-line-number="21"><span class="co">##  bibtex                 0.4.2.2  2020-01-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-22" data-line-number="22"><span class="co">##  Biobase              * 2.48.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-23" data-line-number="23"><span class="co">##  BiocGenerics         * 0.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-24" data-line-number="24"><span class="co">##  BiocNeighbors          1.6.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-25" data-line-number="25"><span class="co">##  BiocParallel           1.22.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-26" data-line-number="26"><span class="co">##  BiocSingular           1.4.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-27" data-line-number="27"><span class="co">##  bit                    1.1-15.2 2020-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-28" data-line-number="28"><span class="co">##  bit64                  0.9-7    2017-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-29" data-line-number="29"><span class="co">##  bitops                 1.0-6    2013-08-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-30" data-line-number="30"><span class="co">##  blob                   1.2.1    2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-31" data-line-number="31"><span class="co">##  bookdown               0.19     2020-05-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-32" data-line-number="32"><span class="co">##  broom                  0.5.6    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-33" data-line-number="33"><span class="co">##  callr                  3.4.3    2020-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-34" data-line-number="34"><span class="co">##  cellranger             1.1.0    2016-07-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-35" data-line-number="35"><span class="co">##  cli                    2.0.2    2020-02-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-36" data-line-number="36"><span class="co">##  cluster              * 2.1.0    2019-06-19 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-37" data-line-number="37"><span class="co">##  codetools              0.2-16   2018-12-24 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-38" data-line-number="38"><span class="co">##  colorspace             1.4-1    2019-03-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-39" data-line-number="39"><span class="co">##  cowplot                1.0.0    2019-07-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-40" data-line-number="40"><span class="co">##  crayon                 1.3.4    2017-09-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-41" data-line-number="41"><span class="co">##  data.table             1.12.8   2019-12-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-42" data-line-number="42"><span class="co">##  DBI                    1.1.0    2019-12-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-43" data-line-number="43"><span class="co">##  dbplyr                 1.4.3    2020-04-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-44" data-line-number="44"><span class="co">##  DelayedArray         * 0.14.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-45" data-line-number="45"><span class="co">##  DelayedMatrixStats     1.10.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-46" data-line-number="46"><span class="co">##  desc                   1.2.0    2018-05-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-47" data-line-number="47"><span class="co">##  devtools               2.3.0    2020-04-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-48" data-line-number="48"><span class="co">##  digest                 0.6.25   2020-02-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-49" data-line-number="49"><span class="co">##  doParallel             1.0.15   2019-08-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-50" data-line-number="50"><span class="co">##  dplyr                * 0.8.5    2020-03-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-51" data-line-number="51"><span class="co">##  dqrng                  0.2.1    2019-05-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-52" data-line-number="52"><span class="co">##  edgeR                  3.30.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-53" data-line-number="53"><span class="co">##  ellipsis               0.3.0    2019-09-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-54" data-line-number="54"><span class="co">##  evaluate               0.14     2019-05-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-55" data-line-number="55"><span class="co">##  extrafont            * 0.17     2014-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-56" data-line-number="56"><span class="co">##  extrafontdb            1.0      2012-06-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-57" data-line-number="57"><span class="co">##  fansi                  0.4.1    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-58" data-line-number="58"><span class="co">##  farver                 2.0.3    2020-01-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-59" data-line-number="59"><span class="co">##  fitdistrplus           1.0-14   2019-01-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-60" data-line-number="60"><span class="co">##  forcats              * 0.5.0    2020-03-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-61" data-line-number="61"><span class="co">##  foreach                1.5.0    2020-03-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-62" data-line-number="62"><span class="co">##  fs                     1.4.1    2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-63" data-line-number="63"><span class="co">##  future                 1.17.0   2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-64" data-line-number="64"><span class="co">##  future.apply           1.5.0    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-65" data-line-number="65"><span class="co">##  generics               0.0.2    2018-11-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-66" data-line-number="66"><span class="co">##  GenomeInfoDb         * 1.24.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-67" data-line-number="67"><span class="co">##  GenomeInfoDbData       1.2.3    2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-68" data-line-number="68"><span class="co">##  GenomicRanges        * 1.40.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-69" data-line-number="69"><span class="co">##  ggbeeswarm             0.6.0    2017-08-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-70" data-line-number="70"><span class="co">##  ggplot2              * 3.3.0    2020-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-71" data-line-number="71"><span class="co">##  ggrepel                0.8.2    2020-03-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-72" data-line-number="72"><span class="co">##  ggridges               0.5.2    2020-01-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-73" data-line-number="73"><span class="co">##  globals                0.12.5   2019-12-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-74" data-line-number="74"><span class="co">##  glue                   1.4.0    2020-04-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-75" data-line-number="75"><span class="co">##  gridBase               0.4-7    2014-02-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-76" data-line-number="76"><span class="co">##  gridExtra            * 2.3      2017-09-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-77" data-line-number="77"><span class="co">##  gtable                 0.3.0    2019-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-78" data-line-number="78"><span class="co">##  gtools                 3.8.2    2020-03-31 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-79" data-line-number="79"><span class="co">##  haven                  2.2.0    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-80" data-line-number="80"><span class="co">##  highr                  0.8      2019-03-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-81" data-line-number="81"><span class="co">##  hms                    0.5.3    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-82" data-line-number="82"><span class="co">##  htmltools              0.4.0    2019-10-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-83" data-line-number="83"><span class="co">##  htmlwidgets            1.5.1    2019-10-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-84" data-line-number="84"><span class="co">##  httr                   1.4.1    2019-08-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-85" data-line-number="85"><span class="co">##  ica                    1.0-2    2018-05-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-86" data-line-number="86"><span class="co">##  igraph                 1.2.5    2020-03-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-87" data-line-number="87"><span class="co">##  IRanges              * 2.22.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-88" data-line-number="88"><span class="co">##  irlba                  2.3.3    2019-02-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-89" data-line-number="89"><span class="co">##  iterators              1.0.12   2019-07-26 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-90" data-line-number="90"><span class="co">##  jsonlite               1.6.1    2020-02-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-91" data-line-number="91"><span class="co">##  KernSmooth             2.23-16  2019-10-15 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-92" data-line-number="92"><span class="co">##  knitr                  1.28     2020-02-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-93" data-line-number="93"><span class="co">##  labeling               0.3      2014-08-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-94" data-line-number="94"><span class="co">##  lattice                0.20-41  2020-04-02 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-95" data-line-number="95"><span class="co">##  lazyeval               0.2.2    2019-03-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-96" data-line-number="96"><span class="co">##  leiden                 0.3.3    2020-02-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-97" data-line-number="97"><span class="co">##  lifecycle              0.2.0    2020-03-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-98" data-line-number="98"><span class="co">##  limma                  3.44.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-99" data-line-number="99"><span class="co">##  listenv                0.8.0    2019-12-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-100" data-line-number="100"><span class="co">##  lmtest                 0.9-37   2019-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-101" data-line-number="101"><span class="co">##  locfit                 1.5-9.4  2020-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-102" data-line-number="102"><span class="co">##  lsei                   1.2-0.1  2020-05-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-103" data-line-number="103"><span class="co">##  lubridate              1.7.8    2020-04-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-104" data-line-number="104"><span class="co">##  magrittr               1.5      2014-11-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-105" data-line-number="105"><span class="co">##  MASS                   7.3-51.5 2019-12-20 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-106" data-line-number="106"><span class="co">##  Matrix                 1.2-18   2019-11-27 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-107" data-line-number="107"><span class="co">##  matrixStats          * 0.56.0   2020-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-108" data-line-number="108"><span class="co">##  mclust               * 5.4.6    2020-04-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-109" data-line-number="109"><span class="co">##  memoise                1.1.0    2017-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-110" data-line-number="110"><span class="co">##  modelr                 0.1.7    2020-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-111" data-line-number="111"><span class="co">##  munsell                0.5.0    2018-06-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-112" data-line-number="112"><span class="co">##  nlme                   3.1-147  2020-04-13 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-113" data-line-number="113"><span class="co">##  NMF                  * 0.22.0   2020-02-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-114" data-line-number="114"><span class="co">##  npsurv                 0.4-0.1  2020-05-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-115" data-line-number="115"><span class="co">##  org.Dr.eg.db         * 3.11.1   2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-116" data-line-number="116"><span class="co">##  patchwork              1.0.0    2019-12-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-117" data-line-number="117"><span class="co">##  pbapply                1.4-2    2019-08-31 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-118" data-line-number="118"><span class="co">##  pheatmap             * 1.0.12   2019-01-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-119" data-line-number="119"><span class="co">##  pillar                 1.4.4    2020-05-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-120" data-line-number="120"><span class="co">##  pkgbuild               1.0.8    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-121" data-line-number="121"><span class="co">##  pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-122" data-line-number="122"><span class="co">##  pkgload                1.0.2    2018-10-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-123" data-line-number="123"><span class="co">##  pkgmaker             * 0.31.1   2020-03-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-124" data-line-number="124"><span class="co">##  plotly                 4.9.2.1  2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-125" data-line-number="125"><span class="co">##  plyr                   1.8.6    2020-03-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-126" data-line-number="126"><span class="co">##  png                    0.1-7    2013-12-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-127" data-line-number="127"><span class="co">##  prettyunits            1.1.1    2020-01-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-128" data-line-number="128"><span class="co">##  processx               3.4.2    2020-02-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-129" data-line-number="129"><span class="co">##  ps                     1.3.3    2020-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-130" data-line-number="130"><span class="co">##  purrr                * 0.3.4    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-131" data-line-number="131"><span class="co">##  R.methodsS3          * 1.8.0    2020-02-14 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-132" data-line-number="132"><span class="co">##  R.oo                 * 1.23.0   2019-11-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-133" data-line-number="133"><span class="co">##  R.utils              * 2.9.2    2019-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-134" data-line-number="134"><span class="co">##  R6                     2.4.1    2019-11-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-135" data-line-number="135"><span class="co">##  RANN                   2.6.1    2019-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-136" data-line-number="136"><span class="co">##  rappdirs               0.3.1    2016-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-137" data-line-number="137"><span class="co">##  RColorBrewer           1.1-2    2014-12-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-138" data-line-number="138"><span class="co">##  Rcpp                   1.0.4.6  2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-139" data-line-number="139"><span class="co">##  RcppAnnoy              0.0.16   2020-03-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-140" data-line-number="140"><span class="co">##  RCurl                  1.98-1.2 2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-141" data-line-number="141"><span class="co">##  readr                * 1.3.1    2018-12-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-142" data-line-number="142"><span class="co">##  readxl                 1.3.1    2019-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-143" data-line-number="143"><span class="co">##  registry             * 0.5-1    2019-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-144" data-line-number="144"><span class="co">##  remotes                2.1.1    2020-02-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-145" data-line-number="145"><span class="co">##  reprex                 0.3.0    2019-05-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-146" data-line-number="146"><span class="co">##  reshape2               1.4.4    2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-147" data-line-number="147"><span class="co">##  reticulate             1.15     2020-04-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-148" data-line-number="148"><span class="co">##  rlang                  0.4.6    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-149" data-line-number="149"><span class="co">##  rmarkdown              2.1      2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-150" data-line-number="150"><span class="co">##  rngtools             * 1.5      2020-01-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-151" data-line-number="151"><span class="co">##  ROCR                   1.0-11   2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-152" data-line-number="152"><span class="co">##  rprojroot              1.3-2    2018-01-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-153" data-line-number="153"><span class="co">##  RSpectra               0.16-0   2019-12-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-154" data-line-number="154"><span class="co">##  RSQLite                2.2.0    2020-01-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-155" data-line-number="155"><span class="co">##  rstudioapi             0.11     2020-02-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-156" data-line-number="156"><span class="co">##  rsvd                   1.0.3    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-157" data-line-number="157"><span class="co">##  Rtsne                  0.15     2018-11-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-158" data-line-number="158"><span class="co">##  Rttf2pt1               1.3.8    2020-01-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-159" data-line-number="159"><span class="co">##  rvest                  0.3.5    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-160" data-line-number="160"><span class="co">##  S4Vectors            * 0.26.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-161" data-line-number="161"><span class="co">##  scales               * 1.1.1    2020-05-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-162" data-line-number="162"><span class="co">##  scater               * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-163" data-line-number="163"><span class="co">##  scran                * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-164" data-line-number="164"><span class="co">##  sctransform            0.2.1    2019-12-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-165" data-line-number="165"><span class="co">##  sessioninfo            1.1.1    2018-11-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-166" data-line-number="166"><span class="co">##  Seurat               * 3.1.5    2020-04-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-167" data-line-number="167"><span class="co">##  SingleCellExperiment * 1.10.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-168" data-line-number="168"><span class="co">##  statmod                1.4.34   2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-169" data-line-number="169"><span class="co">##  stringi                1.4.6    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-170" data-line-number="170"><span class="co">##  stringr              * 1.4.0    2019-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-171" data-line-number="171"><span class="co">##  SummarizedExperiment * 1.18.1   2020-04-30 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-172" data-line-number="172"><span class="co">##  survival               3.1-12   2020-04-10 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-173" data-line-number="173"><span class="co">##  testthat               2.3.2    2020-03-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-174" data-line-number="174"><span class="co">##  tibble               * 3.0.1    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-175" data-line-number="175"><span class="co">##  tidyr                * 1.0.3    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-176" data-line-number="176"><span class="co">##  tidyselect             1.0.0    2020-01-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-177" data-line-number="177"><span class="co">##  tidyverse            * 1.3.0    2019-11-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-178" data-line-number="178"><span class="co">##  tsne                   0.1-3    2016-07-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-179" data-line-number="179"><span class="co">##  usethis                1.6.1    2020-04-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-180" data-line-number="180"><span class="co">##  utf8                   1.1.4    2018-05-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-181" data-line-number="181"><span class="co">##  uwot                   0.1.8    2020-03-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-182" data-line-number="182"><span class="co">##  vctrs                  0.2.4    2020-03-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-183" data-line-number="183"><span class="co">##  vipor                  0.4.5    2017-03-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-184" data-line-number="184"><span class="co">##  viridis              * 0.5.1    2018-03-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-185" data-line-number="185"><span class="co">##  viridisLite          * 0.3.0    2018-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-186" data-line-number="186"><span class="co">##  withr                  2.2.0    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-187" data-line-number="187"><span class="co">##  xfun                   0.13     2020-04-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-188" data-line-number="188"><span class="co">##  xml2                   1.3.2    2020-04-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-189" data-line-number="189"><span class="co">##  xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-190" data-line-number="190"><span class="co">##  XVector                0.28.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-191" data-line-number="191"><span class="co">##  yaml                   2.2.1    2020-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-192" data-line-number="192"><span class="co">##  zlibbioc               1.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb95-193" data-line-number="193"><span class="co">##  zoo                    1.8-8    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb95-194" data-line-number="194"><span class="co">## </span></a>
<a class="sourceLine" id="cb95-195" data-line-number="195"><span class="co">## [1] D:/R_packages</span></a>
<a class="sourceLine" id="cb95-196" data-line-number="196"><span class="co">## [2] C:/Program Files/R/R-4.0.0/library</span></a></code></pre></div>
</details>
</div>
<div id="solutions-2" class="section level2">
<h2><span class="header-section-number">6.6</span> Solutions</h2>
<ol style="list-style-type: decimal">
<li>Any marker that you like- that’s part of the exploration! :)</li>
<li>The feature plot gives a qualitative view of the expression distribution; the violin plot quantifies the distribution. The violin plot It shows the full distribution of the data, which is especially useful in the case of multimodality (multiple peaks in the data)</li>
<li>Use the <code>FindMarkers</code> function. To obtain the DEGs for cluster A against cluster B, set the ident.1 to cluster A and ident.2 to the target cluster B.</li>
<li>?DoHeatmap</li>
</ol>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="optimisations.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="subclustering.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["NMps_book.pdf", "NMps_book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
