<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 1 Introduction | scRNAseq_NMps.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 1 Introduction | scRNAseq_NMps.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 1 Introduction | scRNAseq_NMps.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  

<link rel="next" href="downloading-data.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./"> Hands-on guide to single cell RNA seq analysis in R</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Introduction</a><ul>
<li class="chapter" data-level="1.1" data-path="index.html"><a href="index.html#objectives"><i class="fa fa-check"></i><b>1.1</b> Objectives</a></li>
<li class="chapter" data-level="1.2" data-path="index.html"><a href="index.html#description"><i class="fa fa-check"></i><b>1.2</b> Description:</a></li>
<li class="chapter" data-level="1.3" data-path="index.html"><a href="index.html#links"><i class="fa fa-check"></i><b>1.3</b> Links</a></li>
<li class="chapter" data-level="1.4" data-path="index.html"><a href="index.html#created-by"><i class="fa fa-check"></i><b>1.4</b> Created by:</a></li>
</ul></li>
<li class="chapter" data-level="2" data-path="downloading-data.html"><a href="downloading-data.html"><i class="fa fa-check"></i><b>2</b> Downloading data</a></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a><ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#importing-libraries"><i class="fa fa-check"></i><b>3.2</b> Importing libraries</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#importing-the-data"><i class="fa fa-check"></i><b>3.3</b> Importing the data</a></li>
<li class="chapter" data-level="3.4" data-path="quality-control.html"><a href="quality-control.html#preparing-the-column-metadata"><i class="fa fa-check"></i><b>3.4</b> Preparing the column metadata</a></li>
<li class="chapter" data-level="3.5" data-path="quality-control.html"><a href="quality-control.html#preparing-the-row-metadata"><i class="fa fa-check"></i><b>3.5</b> Preparing the row metadata</a></li>
<li class="chapter" data-level="3.6" data-path="quality-control.html"><a href="quality-control.html#constructing-the-single-cell-experiment-object"><i class="fa fa-check"></i><b>3.6</b> Constructing the single cell experiment object</a></li>
<li class="chapter" data-level="3.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-with-scater"><i class="fa fa-check"></i><b>3.7</b> Quality control with scater</a></li>
<li class="chapter" data-level="3.8" data-path="quality-control.html"><a href="quality-control.html#saving-the-rds-object"><i class="fa fa-check"></i><b>3.8</b> Saving the RDS object</a></li>
<li class="chapter" data-level="3.9" data-path="quality-control.html"><a href="quality-control.html#session-info"><i class="fa fa-check"></i><b>3.9</b> Session Info</a></li>
<li class="chapter" data-level="3.10" data-path="quality-control.html"><a href="quality-control.html#solutions"><i class="fa fa-check"></i><b>3.10</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>4</b> Clustering</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering.html"><a href="clustering.html#introduction-1"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="clustering.html"><a href="clustering.html#importing-libraries-1"><i class="fa fa-check"></i><b>4.2</b> Importing libraries</a></li>
<li class="chapter" data-level="4.3" data-path="clustering.html"><a href="clustering.html#importing-data"><i class="fa fa-check"></i><b>4.3</b> Importing Data</a></li>
<li class="chapter" data-level="4.4" data-path="clustering.html"><a href="clustering.html#useful-seurat-object-interaction-methods"><i class="fa fa-check"></i><b>4.4</b> Useful seurat object interaction methods</a></li>
<li class="chapter" data-level="4.5" data-path="clustering.html"><a href="clustering.html#performing-dimensional-reduction"><i class="fa fa-check"></i><b>4.5</b> Performing dimensional reduction</a></li>
<li class="chapter" data-level="4.6" data-path="clustering.html"><a href="clustering.html#clustering-1"><i class="fa fa-check"></i><b>4.6</b> Clustering</a><ul>
<li class="chapter" data-level="4.6.1" data-path="clustering.html"><a href="clustering.html#constructing-a-shared-nearest-neighbour-snn-graph"><i class="fa fa-check"></i><b>4.6.1</b> Constructing a Shared-nearest neighbour (SNN) graph</a></li>
<li class="chapter" data-level="4.6.2" data-path="clustering.html"><a href="clustering.html#clustering-with-the-louvain-algorithm"><i class="fa fa-check"></i><b>4.6.2</b> Clustering with the Louvain algorithm</a></li>
<li class="chapter" data-level="4.6.3" data-path="clustering.html"><a href="clustering.html#resources"><i class="fa fa-check"></i><b>4.6.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="clustering.html"><a href="clustering.html#umap-with-authors-annotations"><i class="fa fa-check"></i><b>4.7</b> UMAP with authors’ annotations</a></li>
<li class="chapter" data-level="4.8" data-path="clustering.html"><a href="clustering.html#t-sne-with-authors-annotations"><i class="fa fa-check"></i><b>4.8</b> t-SNE with authors’ annotations</a></li>
<li class="chapter" data-level="4.9" data-path="clustering.html"><a href="clustering.html#assigning-our-own-cluster-identities"><i class="fa fa-check"></i><b>4.9</b> Assigning our own cluster identities</a></li>
<li class="chapter" data-level="4.10" data-path="clustering.html"><a href="clustering.html#umap-visualisation-of-our-annotations"><i class="fa fa-check"></i><b>4.10</b> UMAP visualisation of our annotations</a></li>
<li class="chapter" data-level="4.11" data-path="clustering.html"><a href="clustering.html#highlighting-tailbud-cells"><i class="fa fa-check"></i><b>4.11</b> Highlighting tailbud cells</a></li>
<li class="chapter" data-level="4.12" data-path="clustering.html"><a href="clustering.html#saving-rds-object"><i class="fa fa-check"></i><b>4.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="4.13" data-path="clustering.html"><a href="clustering.html#session-info-1"><i class="fa fa-check"></i><b>4.13</b> Session Info</a></li>
<li class="chapter" data-level="4.14" data-path="clustering.html"><a href="clustering.html#solutions-1"><i class="fa fa-check"></i><b>4.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html"><i class="fa fa-check"></i><b>5</b> Differentially Expressed Genes</a><ul>
<li class="chapter" data-level="5.1" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualisation-of-marker-genes-in-our-dataset"><i class="fa fa-check"></i><b>5.1</b> Visualisation of marker genes in our dataset</a></li>
<li class="chapter" data-level="5.2" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#obtaining-a-list-of-differentially-expressed-genes"><i class="fa fa-check"></i><b>5.2</b> Obtaining a list of differentially expressed genes</a></li>
<li class="chapter" data-level="5.3" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#analysing-output-of-the-wilcoxon-test"><i class="fa fa-check"></i><b>5.3</b> Analysing output of the wilcoxon test</a></li>
<li class="chapter" data-level="5.4" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualising-the-expression-of-the-marker-genes"><i class="fa fa-check"></i><b>5.4</b> Visualising the expression of the marker genes</a></li>
<li class="chapter" data-level="5.5" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#session-info-2"><i class="fa fa-check"></i><b>5.5</b> Session Info</a></li>
<li class="chapter" data-level="5.6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#solutions-2"><i class="fa fa-check"></i><b>5.6</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="exploring-dimensionality-reduction.html"><a href="exploring-dimensionality-reduction.html"><i class="fa fa-check"></i><b>6</b> Exploring dimensionality reduction</a><ul>
<li class="chapter" data-level="6.1" data-path="exploring-dimensionality-reduction.html"><a href="exploring-dimensionality-reduction.html#load-data"><i class="fa fa-check"></i><b>6.1</b> Load Data</a></li>
<li class="chapter" data-level="6.2" data-path="exploring-dimensionality-reduction.html"><a href="exploring-dimensionality-reduction.html#pca"><i class="fa fa-check"></i><b>6.2</b> PCA</a></li>
<li class="chapter" data-level="6.3" data-path="exploring-dimensionality-reduction.html"><a href="exploring-dimensionality-reduction.html#t-stochastic-neighbor-embedding"><i class="fa fa-check"></i><b>6.3</b> <span class="math inline">\(t\)</span>-stochastic neighbor embedding</a></li>
<li class="chapter" data-level="6.4" data-path="exploring-dimensionality-reduction.html"><a href="exploring-dimensionality-reduction.html#umap"><i class="fa fa-check"></i><b>6.4</b> UMAP</a></li>
<li class="chapter" data-level="6.5" data-path="exploring-dimensionality-reduction.html"><a href="exploring-dimensionality-reduction.html#bibliography"><i class="fa fa-check"></i><b>6.5</b> Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="exploring-clustering.html"><a href="exploring-clustering.html"><i class="fa fa-check"></i><b>7</b> Exploring clustering</a><ul>
<li class="chapter" data-level="7.1" data-path="exploring-clustering.html"><a href="exploring-clustering.html#load-data-1"><i class="fa fa-check"></i><b>7.1</b> Load Data</a></li>
<li class="chapter" data-level="7.2" data-path="exploring-clustering.html"><a href="exploring-clustering.html#clustering-2"><i class="fa fa-check"></i><b>7.2</b> Clustering</a></li>
<li class="chapter" data-level="7.3" data-path="exploring-clustering.html"><a href="exploring-clustering.html#k-means"><i class="fa fa-check"></i><b>7.3</b> k-means</a><ul>
<li class="chapter" data-level="7.3.1" data-path="exploring-clustering.html"><a href="exploring-clustering.html#choosing-k"><i class="fa fa-check"></i><b>7.3.1</b> Choosing k</a></li>
<li class="chapter" data-level="7.3.2" data-path="exploring-clustering.html"><a href="exploring-clustering.html#which-subspace"><i class="fa fa-check"></i><b>7.3.2</b> Which subspace?</a></li>
</ul></li>
<li class="chapter" data-level="7.4" data-path="exploring-clustering.html"><a href="exploring-clustering.html#graph-based-clustering"><i class="fa fa-check"></i><b>7.4</b> Graph-based clustering</a><ul>
<li class="chapter" data-level="7.4.1" data-path="exploring-clustering.html"><a href="exploring-clustering.html#force-directed-layout"><i class="fa fa-check"></i><b>7.4.1</b> Force-directed layout</a></li>
</ul></li>
<li class="chapter" data-level="7.5" data-path="exploring-clustering.html"><a href="exploring-clustering.html#bibliography-1"><i class="fa fa-check"></i><b>7.5</b> Bibliography</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="optimisations.html"><a href="optimisations.html"><i class="fa fa-check"></i><b>8</b> Optimisations</a><ul>
<li class="chapter" data-level="8.1" data-path="optimisations.html"><a href="optimisations.html#importing-libraries-2"><i class="fa fa-check"></i><b>8.1</b> Importing libraries</a></li>
<li class="chapter" data-level="8.2" data-path="optimisations.html"><a href="optimisations.html#importing-data-1"><i class="fa fa-check"></i><b>8.2</b> Importing Data</a></li>
<li class="chapter" data-level="8.3" data-path="optimisations.html"><a href="optimisations.html#number-of-pcs-to-retain"><i class="fa fa-check"></i><b>8.3</b> Number of PCs to retain</a><ul>
<li class="chapter" data-level="8.3.1" data-path="optimisations.html"><a href="optimisations.html#scree-plot"><i class="fa fa-check"></i><b>8.3.1</b> Scree plot</a></li>
<li class="chapter" data-level="8.3.2" data-path="optimisations.html"><a href="optimisations.html#examining-technical-noise"><i class="fa fa-check"></i><b>8.3.2</b> Examining technical noise</a></li>
</ul></li>
<li class="chapter" data-level="8.4" data-path="optimisations.html"><a href="optimisations.html#evaluating-clustering-results"><i class="fa fa-check"></i><b>8.4</b> Evaluating clustering results</a><ul>
<li class="chapter" data-level="8.4.1" data-path="optimisations.html"><a href="optimisations.html#cluster-stability"><i class="fa fa-check"></i><b>8.4.1</b> Cluster stability</a></li>
<li class="chapter" data-level="8.4.2" data-path="optimisations.html"><a href="optimisations.html#cluster-separability"><i class="fa fa-check"></i><b>8.4.2</b> Cluster separability</a></li>
</ul></li>
<li class="chapter" data-level="8.5" data-path="optimisations.html"><a href="optimisations.html#comparison-with-authors-labels"><i class="fa fa-check"></i><b>8.5</b> Comparison with authors’ labels</a><ul>
<li class="chapter" data-level="8.5.1" data-path="optimisations.html"><a href="optimisations.html#concordance-in-tailbud-cluster-assignment"><i class="fa fa-check"></i><b>8.5.1</b> Concordance in tailbud cluster assignment</a></li>
<li class="chapter" data-level="8.5.2" data-path="optimisations.html"><a href="optimisations.html#optimising-the-seurat-clustering-parameters"><i class="fa fa-check"></i><b>8.5.2</b> Optimising the seurat clustering parameters</a></li>
</ul></li>
<li class="chapter" data-level="8.6" data-path="optimisations.html"><a href="optimisations.html#selecting-cells-for-downstream-analysis"><i class="fa fa-check"></i><b>8.6</b> Selecting cells for downstream analysis</a></li>
<li class="chapter" data-level="8.7" data-path="optimisations.html"><a href="optimisations.html#session-info-3"><i class="fa fa-check"></i><b>8.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="9" data-path="subclustering.html"><a href="subclustering.html"><i class="fa fa-check"></i><b>9</b> Subclustering</a><ul>
<li class="chapter" data-level="9.1" data-path="subclustering.html"><a href="subclustering.html#importing-libraries-3"><i class="fa fa-check"></i><b>9.1</b> Importing libraries</a></li>
<li class="chapter" data-level="9.2" data-path="subclustering.html"><a href="subclustering.html#importing-data-2"><i class="fa fa-check"></i><b>9.2</b> Importing Data</a></li>
<li class="chapter" data-level="9.3" data-path="subclustering.html"><a href="subclustering.html#subclustering-the-tb-cells---using-only-cells-that-are-assigned-to-the-shared-cluster-across-all-the-clustering-resolutions."><i class="fa fa-check"></i><b>9.3</b> Subclustering the TB cells - Using only cells that are assigned to the ‘shared’ cluster across all the clustering resolutions.</a></li>
<li class="chapter" data-level="9.4" data-path="subclustering.html"><a href="subclustering.html#dimensional-reduction-with-umap-and-clustering"><i class="fa fa-check"></i><b>9.4</b> Dimensional reduction with UMAP and clustering</a></li>
<li class="chapter" data-level="9.5" data-path="subclustering.html"><a href="subclustering.html#identifying-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>9.5</b> Identifying marker genes for each cluster</a></li>
<li class="chapter" data-level="9.6" data-path="subclustering.html"><a href="subclustering.html#merging-clusters-that-do-not-appear-distinct"><i class="fa fa-check"></i><b>9.6</b> Merging clusters that do not appear distinct</a></li>
<li class="chapter" data-level="9.7" data-path="subclustering.html"><a href="subclustering.html#plotting-umap-and-assigning-cluster-labels"><i class="fa fa-check"></i><b>9.7</b> Plotting UMAP and assigning cluster labels</a></li>
<li class="chapter" data-level="9.8" data-path="subclustering.html"><a href="subclustering.html#checking-the-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>9.8</b> Checking the marker genes for each cluster</a></li>
<li class="chapter" data-level="9.9" data-path="subclustering.html"><a href="subclustering.html#sox2-tbxta-coexpression"><i class="fa fa-check"></i><b>9.9</b> Sox2 tbxta coexpression</a></li>
<li class="chapter" data-level="9.10" data-path="subclustering.html"><a href="subclustering.html#viewing-genes-of-interest-with-a-dotplot"><i class="fa fa-check"></i><b>9.10</b> Viewing genes of interest with a dotplot</a></li>
<li class="chapter" data-level="9.11" data-path="subclustering.html"><a href="subclustering.html#exporting-the-violin-plots-for-genes-of-interest."><i class="fa fa-check"></i><b>9.11</b> Exporting the violin plots for genes of interest.</a></li>
<li class="chapter" data-level="9.12" data-path="subclustering.html"><a href="subclustering.html#saving-rds-object-1"><i class="fa fa-check"></i><b>9.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="9.13" data-path="subclustering.html"><a href="subclustering.html#session-info-4"><i class="fa fa-check"></i><b>9.13</b> Session Info</a></li>
<li class="chapter" data-level="9.14" data-path="subclustering.html"><a href="subclustering.html#solutions-3"><i class="fa fa-check"></i><b>9.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="10" data-path="pseudotime.html"><a href="pseudotime.html"><i class="fa fa-check"></i><b>10</b> Pseudotime</a><ul>
<li class="chapter" data-level="10.1" data-path="pseudotime.html"><a href="pseudotime.html#importing-libraries-4"><i class="fa fa-check"></i><b>10.1</b> Importing libraries</a></li>
<li class="chapter" data-level="10.2" data-path="pseudotime.html"><a href="pseudotime.html#importing-data-3"><i class="fa fa-check"></i><b>10.2</b> Importing data</a></li>
<li class="chapter" data-level="10.3" data-path="pseudotime.html"><a href="pseudotime.html#tradeseq-workflow-analyses"><i class="fa fa-check"></i><b>10.3</b> TradeSeq workflow analyses</a></li>
<li class="chapter" data-level="10.4" data-path="pseudotime.html"><a href="pseudotime.html#session-info-5"><i class="fa fa-check"></i><b>10.4</b> Session Info</a></li>
<li class="chapter" data-level="10.5" data-path="pseudotime.html"><a href="pseudotime.html#solutions-4"><i class="fa fa-check"></i><b>10.5</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="11" data-path="random-seeds.html"><a href="random-seeds.html"><i class="fa fa-check"></i><b>11</b> Random seeds</a><ul>
<li class="chapter" data-level="11.1" data-path="random-seeds.html"><a href="random-seeds.html#what-is-a-random-seed"><i class="fa fa-check"></i><b>11.1</b> What is a random seed?</a></li>
<li class="chapter" data-level="11.2" data-path="random-seeds.html"><a href="random-seeds.html#simulations"><i class="fa fa-check"></i><b>11.2</b> Simulations</a></li>
<li class="chapter" data-level="11.3" data-path="random-seeds.html"><a href="random-seeds.html#box-muller-transform"><i class="fa fa-check"></i><b>11.3</b> Box-Muller transform</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<!--bookdown:title:end-->
<!--bookdown:title:start-->
<div id="introduction" class="section level1">
<h1><span class="header-section-number">Chapter 1</span> Introduction</h1>
<p>This is the github repository for the “Hands-on guide to single cell RNA seq analysis in R computational workshop” series given to the AMA and Steventon labs on 4th June 2020.</p>
<div id="objectives" class="section level2">
<h2><span class="header-section-number">1.1</span> Objectives</h2>
<ul>
<li>The volume of online resources on the single-cell field is vast. We hope that this short practical session will help point you in the right direction in conducting your own scRNA-seq data analysis by:</li>
<li>Highlighting the relevant reading material, scRNA-seq packages and other resources available for further study.</li>
<li>Sharing practical advice on dealing with issues that you may encounter in your own analyses.</li>
<li>Providing you with code as starting material to tinker with.</li>
</ul>
</div>
<div id="description" class="section level2">
<h2><span class="header-section-number">1.2</span> Description:</h2>
<p>We will be studying the scRNA-seq dataset of 18hpf zebrafish embryos, with a focus on examining the gene expression states of the tailbud Neuromesodermal progenitors.</p>
<p>Additioanlly we will be taking a closer look at some key techniques, highlighting important details to keep in mind both when implementing and evaluating single-cell anaylsis.</p>
<p>We hope that this will provide a good starting point for getting to grips with this fast-moving field, and encourage you to contact us for further disucssion or advice if needed.</p>
</div>
<div id="links" class="section level2">
<h2><span class="header-section-number">1.3</span> Links</h2>
<p><a href="https://science.sciencemag.org/content/360/6392/981.full">Reference paper</a></p>
<p><a href="https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSM3067194">Dataset repository</a></p>
</div>
<div id="created-by" class="section level2">
<h2><span class="header-section-number">1.4</span> Created by:</h2>
<ul>
<li>Kane Toh</li>
<li>Giles Harper-Donnelly</li>
</ul>

</div>
</div>
            </section>

          </div>
        </div>
      </div>

<a href="downloading-data.html" class="navigation navigation-next navigation-unique" aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["scRNAseq_NMps.pdf", "scRNAseq_NMps.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
