<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 5 Optimisations | NMps_book.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 5 Optimisations | NMps_book.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 5 Optimisations | NMps_book.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="clustering.html"/>
<link rel="next" href="differentially-expressed-genes.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="downloading-data.html"><a href="downloading-data.html"><i class="fa fa-check"></i><b>2</b> Downloading data</a></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a><ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#importing-libraries"><i class="fa fa-check"></i><b>3.2</b> Importing libraries</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#importing-the-data"><i class="fa fa-check"></i><b>3.3</b> Importing the data</a></li>
<li class="chapter" data-level="3.4" data-path="quality-control.html"><a href="quality-control.html#preparing-the-column-metadata"><i class="fa fa-check"></i><b>3.4</b> Preparing the column metadata</a></li>
<li class="chapter" data-level="3.5" data-path="quality-control.html"><a href="quality-control.html#preparing-the-row-metadata"><i class="fa fa-check"></i><b>3.5</b> Preparing the row metadata</a></li>
<li class="chapter" data-level="3.6" data-path="quality-control.html"><a href="quality-control.html#constructing-the-single-cell-experiment-object"><i class="fa fa-check"></i><b>3.6</b> Constructing the single cell experiment object</a></li>
<li class="chapter" data-level="3.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-with-scater"><i class="fa fa-check"></i><b>3.7</b> Quality control with scater</a></li>
<li class="chapter" data-level="3.8" data-path="quality-control.html"><a href="quality-control.html#saving-the-rds-object"><i class="fa fa-check"></i><b>3.8</b> Saving the RDS object</a></li>
<li class="chapter" data-level="3.9" data-path="quality-control.html"><a href="quality-control.html#session-info"><i class="fa fa-check"></i><b>3.9</b> Session Info</a></li>
<li class="chapter" data-level="3.10" data-path="quality-control.html"><a href="quality-control.html#solutions"><i class="fa fa-check"></i><b>3.10</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>4</b> Clustering</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="clustering.html"><a href="clustering.html#importing-libraries-1"><i class="fa fa-check"></i><b>4.2</b> Importing libraries</a></li>
<li class="chapter" data-level="4.3" data-path="clustering.html"><a href="clustering.html#importing-data"><i class="fa fa-check"></i><b>4.3</b> Importing Data</a></li>
<li class="chapter" data-level="4.4" data-path="clustering.html"><a href="clustering.html#useful-seurat-object-interaction-methods"><i class="fa fa-check"></i><b>4.4</b> Useful seurat object interaction methods</a></li>
<li class="chapter" data-level="4.5" data-path="clustering.html"><a href="clustering.html#performing-dimensional-reduction"><i class="fa fa-check"></i><b>4.5</b> Performing dimensional reduction</a></li>
<li class="chapter" data-level="4.6" data-path="clustering.html"><a href="clustering.html#clustering-1"><i class="fa fa-check"></i><b>4.6</b> Clustering</a><ul>
<li class="chapter" data-level="4.6.1" data-path="clustering.html"><a href="clustering.html#constructing-a-shared-nearest-neighbour-snn-graph"><i class="fa fa-check"></i><b>4.6.1</b> Constructing a Shared-nearest neighbour (SNN) graph</a></li>
<li class="chapter" data-level="4.6.2" data-path="clustering.html"><a href="clustering.html#clustering-with-the-louvain-algorithm"><i class="fa fa-check"></i><b>4.6.2</b> Clustering with the Louvain algorithm</a></li>
<li class="chapter" data-level="4.6.3" data-path="clustering.html"><a href="clustering.html#resources"><i class="fa fa-check"></i><b>4.6.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="clustering.html"><a href="clustering.html#umap-with-authors-annotations"><i class="fa fa-check"></i><b>4.7</b> UMAP with authors’ annotations</a></li>
<li class="chapter" data-level="4.8" data-path="clustering.html"><a href="clustering.html#t-sne-with-authors-annotations"><i class="fa fa-check"></i><b>4.8</b> t-SNE with authors’ annotations</a></li>
<li class="chapter" data-level="4.9" data-path="clustering.html"><a href="clustering.html#assigning-our-own-cluster-identities"><i class="fa fa-check"></i><b>4.9</b> Assigning our own cluster identities</a></li>
<li class="chapter" data-level="4.10" data-path="clustering.html"><a href="clustering.html#umap-visualisation-of-our-annotations"><i class="fa fa-check"></i><b>4.10</b> UMAP visualisation of our annotations</a></li>
<li class="chapter" data-level="4.11" data-path="clustering.html"><a href="clustering.html#highlighting-tailbud-cells"><i class="fa fa-check"></i><b>4.11</b> Highlighting tailbud cells</a></li>
<li class="chapter" data-level="4.12" data-path="clustering.html"><a href="clustering.html#saving-rds-object"><i class="fa fa-check"></i><b>4.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="4.13" data-path="clustering.html"><a href="clustering.html#session-info-1"><i class="fa fa-check"></i><b>4.13</b> Session Info</a></li>
<li class="chapter" data-level="4.14" data-path="clustering.html"><a href="clustering.html#solutions-1"><i class="fa fa-check"></i><b>4.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="optimisations.html"><a href="optimisations.html"><i class="fa fa-check"></i><b>5</b> Optimisations</a><ul>
<li class="chapter" data-level="5.1" data-path="optimisations.html"><a href="optimisations.html#importing-libraries-2"><i class="fa fa-check"></i><b>5.1</b> Importing libraries</a></li>
<li class="chapter" data-level="5.2" data-path="optimisations.html"><a href="optimisations.html#importing-data-1"><i class="fa fa-check"></i><b>5.2</b> Importing Data</a></li>
<li class="chapter" data-level="5.3" data-path="optimisations.html"><a href="optimisations.html#number-of-pcs-to-retain"><i class="fa fa-check"></i><b>5.3</b> Number of PCs to retain</a><ul>
<li class="chapter" data-level="5.3.1" data-path="optimisations.html"><a href="optimisations.html#scree-plot"><i class="fa fa-check"></i><b>5.3.1</b> Scree plot</a></li>
<li class="chapter" data-level="5.3.2" data-path="optimisations.html"><a href="optimisations.html#examining-technical-noise"><i class="fa fa-check"></i><b>5.3.2</b> Examining technical noise</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="optimisations.html"><a href="optimisations.html#evaluating-clustering-results"><i class="fa fa-check"></i><b>5.4</b> Evaluating clustering results</a><ul>
<li class="chapter" data-level="5.4.1" data-path="optimisations.html"><a href="optimisations.html#cluster-stability"><i class="fa fa-check"></i><b>5.4.1</b> Cluster stability</a></li>
<li class="chapter" data-level="5.4.2" data-path="optimisations.html"><a href="optimisations.html#cluster-separability"><i class="fa fa-check"></i><b>5.4.2</b> Cluster separability</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="optimisations.html"><a href="optimisations.html#comparison-with-authors-labels"><i class="fa fa-check"></i><b>5.5</b> Comparison with authors’ labels</a><ul>
<li class="chapter" data-level="5.5.1" data-path="optimisations.html"><a href="optimisations.html#concordance-in-tailbud-cluster-assignment"><i class="fa fa-check"></i><b>5.5.1</b> Concordance in tailbud cluster assignment</a></li>
<li class="chapter" data-level="5.5.2" data-path="optimisations.html"><a href="optimisations.html#optimising-the-seurat-clustering-parameters"><i class="fa fa-check"></i><b>5.5.2</b> Optimising the seurat clustering parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="optimisations.html"><a href="optimisations.html#selecting-cells-for-downstream-analysis"><i class="fa fa-check"></i><b>5.6</b> Selecting cells for downstream analysis</a></li>
<li class="chapter" data-level="5.7" data-path="optimisations.html"><a href="optimisations.html#session-info-2"><i class="fa fa-check"></i><b>5.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html"><i class="fa fa-check"></i><b>6</b> Differentially Expressed Genes</a><ul>
<li class="chapter" data-level="6.1" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualisation-of-marker-genes-in-our-dataset"><i class="fa fa-check"></i><b>6.1</b> Visualisation of marker genes in our dataset</a></li>
<li class="chapter" data-level="6.2" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#obtaining-a-list-of-differentially-expressed-genes"><i class="fa fa-check"></i><b>6.2</b> Obtaining a list of differentially expressed genes</a></li>
<li class="chapter" data-level="6.3" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#analysing-output-of-the-wilcoxon-test"><i class="fa fa-check"></i><b>6.3</b> Analysing output of the wilcoxon test</a></li>
<li class="chapter" data-level="6.4" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualising-the-expression-of-the-marker-genes"><i class="fa fa-check"></i><b>6.4</b> Visualising the expression of the marker genes</a></li>
<li class="chapter" data-level="6.5" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#session-info-3"><i class="fa fa-check"></i><b>6.5</b> Session Info</a></li>
<li class="chapter" data-level="6.6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#solutions-2"><i class="fa fa-check"></i><b>6.6</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="subclustering.html"><a href="subclustering.html"><i class="fa fa-check"></i><b>7</b> Subclustering</a><ul>
<li class="chapter" data-level="7.1" data-path="subclustering.html"><a href="subclustering.html#importing-libraries-3"><i class="fa fa-check"></i><b>7.1</b> Importing libraries</a></li>
<li class="chapter" data-level="7.2" data-path="subclustering.html"><a href="subclustering.html#importing-data-2"><i class="fa fa-check"></i><b>7.2</b> Importing Data</a></li>
<li class="chapter" data-level="7.3" data-path="subclustering.html"><a href="subclustering.html#subclustering-the-tb-cells---using-only-cells-that-are-assigned-to-the-shared-cluster-across-all-the-clustering-resolutions."><i class="fa fa-check"></i><b>7.3</b> Subclustering the TB cells - Using only cells that are assigned to the ‘shared’ cluster across all the clustering resolutions.</a></li>
<li class="chapter" data-level="7.4" data-path="subclustering.html"><a href="subclustering.html#dimensional-reduction-with-umap-and-clustering"><i class="fa fa-check"></i><b>7.4</b> Dimensional reduction with UMAP and clustering</a></li>
<li class="chapter" data-level="7.5" data-path="subclustering.html"><a href="subclustering.html#identifying-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.5</b> Identifying marker genes for each cluster</a></li>
<li class="chapter" data-level="7.6" data-path="subclustering.html"><a href="subclustering.html#merging-clusters-that-do-not-appear-distinct"><i class="fa fa-check"></i><b>7.6</b> Merging clusters that do not appear distinct</a></li>
<li class="chapter" data-level="7.7" data-path="subclustering.html"><a href="subclustering.html#plotting-umap-and-assigning-cluster-labels"><i class="fa fa-check"></i><b>7.7</b> Plotting UMAP and assigning cluster labels</a></li>
<li class="chapter" data-level="7.8" data-path="subclustering.html"><a href="subclustering.html#checking-the-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.8</b> Checking the marker genes for each cluster</a></li>
<li class="chapter" data-level="7.9" data-path="subclustering.html"><a href="subclustering.html#sox2-tbxta-coexpression"><i class="fa fa-check"></i><b>7.9</b> Sox2 tbxta coexpression</a></li>
<li class="chapter" data-level="7.10" data-path="subclustering.html"><a href="subclustering.html#viewing-genes-of-interest-with-a-dotplot"><i class="fa fa-check"></i><b>7.10</b> Viewing genes of interest with a dotplot</a></li>
<li class="chapter" data-level="7.11" data-path="subclustering.html"><a href="subclustering.html#exporting-the-violin-plots-for-genes-of-interest."><i class="fa fa-check"></i><b>7.11</b> Exporting the violin plots for genes of interest.</a></li>
<li class="chapter" data-level="7.12" data-path="subclustering.html"><a href="subclustering.html#saving-rds-object-1"><i class="fa fa-check"></i><b>7.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="7.13" data-path="subclustering.html"><a href="subclustering.html#session-info-4"><i class="fa fa-check"></i><b>7.13</b> Session Info</a></li>
<li class="chapter" data-level="7.14" data-path="subclustering.html"><a href="subclustering.html#solutions-3"><i class="fa fa-check"></i><b>7.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="pseudotime.html"><a href="pseudotime.html"><i class="fa fa-check"></i><b>8</b> Pseudotime</a><ul>
<li class="chapter" data-level="8.1" data-path="pseudotime.html"><a href="pseudotime.html#importing-libraries-4"><i class="fa fa-check"></i><b>8.1</b> Importing libraries</a></li>
<li class="chapter" data-level="8.2" data-path="pseudotime.html"><a href="pseudotime.html#importing-data-3"><i class="fa fa-check"></i><b>8.2</b> Importing data</a></li>
<li class="chapter" data-level="8.3" data-path="pseudotime.html"><a href="pseudotime.html#tradeseq-workflow-analyses"><i class="fa fa-check"></i><b>8.3</b> TradeSeq workflow analyses</a></li>
<li class="chapter" data-level="8.4" data-path="pseudotime.html"><a href="pseudotime.html#session-info-5"><i class="fa fa-check"></i><b>8.4</b> Session Info</a></li>
<li class="chapter" data-level="8.5" data-path="pseudotime.html"><a href="pseudotime.html#solutions-4"><i class="fa fa-check"></i><b>8.5</b> Solutions</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="optimisations" class="section level1">
<h1><span class="header-section-number">Chapter 5</span> Optimisations</h1>
<div id="importing-libraries-2" class="section level2">
<h2><span class="header-section-number">5.1</span> Importing libraries</h2>
<div class="sourceCode" id="cb35"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb35-1" data-line-number="1"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb35-2" data-line-number="2"><span class="kw">library</span>(Seurat)</a>
<a class="sourceLine" id="cb35-3" data-line-number="3"><span class="kw">library</span>(scran)</a>
<a class="sourceLine" id="cb35-4" data-line-number="4"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb35-5" data-line-number="5"><span class="kw">library</span>(scater)</a>
<a class="sourceLine" id="cb35-6" data-line-number="6"><span class="kw">library</span>(NMF)</a>
<a class="sourceLine" id="cb35-7" data-line-number="7"><span class="kw">library</span>(mclust)</a>
<a class="sourceLine" id="cb35-8" data-line-number="8"><span class="kw">library</span>(pheatmap)</a>
<a class="sourceLine" id="cb35-9" data-line-number="9"><span class="kw">library</span>(extrafont) </a>
<a class="sourceLine" id="cb35-10" data-line-number="10"><span class="kw">loadfonts</span>(<span class="dt">quiet =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
</div>
<div id="importing-data-1" class="section level2">
<h2><span class="header-section-number">5.2</span> Importing Data</h2>
<div class="sourceCode" id="cb36"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb36-1" data-line-number="1">hpf18_seurat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span><span class="st">&quot;rds/hpf18_seurat_dimred_new.rds&quot;</span>)</a>
<a class="sourceLine" id="cb36-2" data-line-number="2">hpf18_sce &lt;-<span class="st"> </span><span class="kw">as.SingleCellExperiment</span>(hpf18_seurat)</a></code></pre></div>
</div>
<div id="number-of-pcs-to-retain" class="section level2">
<h2><span class="header-section-number">5.3</span> Number of PCs to retain</h2>
<div id="scree-plot" class="section level3">
<h3><span class="header-section-number">5.3.1</span> Scree plot</h3>
<p>We plot the % of variance explained against the number of principal components. If the top few PCs capture most of the biological signal in our dataset, we expect that there should be a sharp drop between these PCs and the rest of the PCs that capture the technical noise. This drop resembles an ‘elbow’, and hence we are searching for the elbow in the scree plot. In general, methods that involve plotting the % of variance explained against some parameter of interest are called elbow methods, since the point of optimisation is where the said elbow occurs. In very high-dimensional datasets such as ours, it is not uncommon for there to be no clear elbow in the data.</p>
<div class="sourceCode" id="cb37"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb37-1" data-line-number="1">ep &lt;-<span class="st"> </span><span class="kw">ElbowPlot</span>(hpf18_seurat, <span class="dt">ndims =</span> <span class="dv">50</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb37-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">12</span>,  <span class="dt">color =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">size=</span><span class="fl">1.5</span>, <span class="dt">linetype=</span><span class="st">&quot;dotted&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb37-3" data-line-number="3"><span class="st">  </span><span class="kw">ylab</span>(<span class="dt">label =</span> <span class="st">&quot;Percentage of variance explained&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb37-4" data-line-number="4"><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.line =</span> <span class="kw">element_line</span>(<span class="dt">colour =</span> <span class="st">&quot;black&quot;</span>),</a>
<a class="sourceLine" id="cb37-5" data-line-number="5">    <span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb37-6" data-line-number="6">    <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb37-7" data-line-number="7">    <span class="dt">panel.border =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb37-8" data-line-number="8">    <span class="dt">panel.background =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb37-9" data-line-number="9">    <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>))</a>
<a class="sourceLine" id="cb37-10" data-line-number="10">ep</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/PCA_elbow-1.png" alt="Figure 1- Elbow plot for determining the number of principal components. There is a gradual drop in the % of variance explained in the data. The % variance explained for 12 PCs is highlighted by the red line. " width="672" />
<p class="caption">
(#fig:PCA_elbow)Figure 1- Elbow plot for determining the number of principal components. There is a gradual drop in the % of variance explained in the data. The % variance explained for 12 PCs is highlighted by the red line.
</p>
</div>
</div>
<div id="examining-technical-noise" class="section level3">
<h3><span class="header-section-number">5.3.2</span> Examining technical noise</h3>
<p>We use the functions available from the <code>scran</code> package in this section.</p>
<p>In this alternative method, we first model the mean-variance relationship of the technical noise in our dataset. This is how the relationship would look like, were there to be no biological signal at all. We can do this by a parametric method in the <code>modelGeneVar</code> function. That is to say, we assume the functional form of the noise as a non-linear curve of the form: <span class="math inline">\(y = \frac{ax}{x^n + b}\)</span></p>
<p>Alternatively, we can make a more stringent assumption that the data trend should follow a poisson distribution. This is implemented in the <code>modelGeneVarByPoisson</code> function.</p>
<div class="sourceCode" id="cb38"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb38-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">2020</span>) <span class="co">## For reproducibiity</span></a>
<a class="sourceLine" id="cb38-2" data-line-number="2"></a>
<a class="sourceLine" id="cb38-3" data-line-number="3"><span class="co">## We fit a non-linear curve to the variances against the means of the data.</span></a>
<a class="sourceLine" id="cb38-4" data-line-number="4"></a>
<a class="sourceLine" id="cb38-5" data-line-number="5">dec_sce &lt;-<span class="st"> </span>scran<span class="op">::</span><span class="kw">modelGeneVar</span>(hpf18_sce)</a>
<a class="sourceLine" id="cb38-6" data-line-number="6">fit_sce &lt;-<span class="st"> </span><span class="kw">metadata</span>(dec_sce)</a>
<a class="sourceLine" id="cb38-7" data-line-number="7"><span class="kw">plot</span>(fit_sce<span class="op">$</span>mean, fit_sce<span class="op">$</span>var, <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb38-8" data-line-number="8">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb38-9" data-line-number="9"><span class="kw">curve</span>(fit_sce<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>, <span class="dt">lwd=</span><span class="dv">2</span>)</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/PCA_tech_noise-1.png" alt="Figure 2- Variance in our dataset against the mean. Each point represents a gene and the blue curve is the fitted trend to all of the genes with a non-linear model." width="50%" />
<p class="caption">
(#fig:PCA_tech_noise)Figure 2- Variance in our dataset against the mean. Each point represents a gene and the blue curve is the fitted trend to all of the genes with a non-linear model.
</p>
</div>
<div class="sourceCode" id="cb39"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb39-1" data-line-number="1"><span class="co">### Here, we assume that the technical noise can be modelled with a poisson distribution</span></a>
<a class="sourceLine" id="cb39-2" data-line-number="2"></a>
<a class="sourceLine" id="cb39-3" data-line-number="3">dec_pois_pbmc &lt;-<span class="st"> </span><span class="kw">modelGeneVarByPoisson</span>(hpf18_sce)</a>
<a class="sourceLine" id="cb39-4" data-line-number="4"><span class="kw">plot</span>(dec_pois_pbmc<span class="op">$</span>mean, dec_pois_pbmc<span class="op">$</span>total, <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">xlab=</span><span class="st">&quot;Mean of log-expression&quot;</span>,</a>
<a class="sourceLine" id="cb39-5" data-line-number="5">    <span class="dt">ylab=</span><span class="st">&quot;Variance of log-expression&quot;</span>)</a>
<a class="sourceLine" id="cb39-6" data-line-number="6"><span class="kw">curve</span>(<span class="kw">metadata</span>(dec_pois_pbmc)<span class="op">$</span><span class="kw">trend</span>(x), <span class="dt">col=</span><span class="st">&quot;dodgerblue&quot;</span>, <span class="dt">add=</span><span class="ot">TRUE</span>)</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/PCA_tech_poisson-1.png" alt="Figure 3- Variance in our dataset against the mean. Each point represents a gene and the blue curve is the fitted trend to all of the genes with a poisson model that is assumed to be the distributional form of the mean-variance relationship." width="672" />
<p class="caption">
(#fig:PCA_tech_poisson)Figure 3- Variance in our dataset against the mean. Each point represents a gene and the blue curve is the fitted trend to all of the genes with a poisson model that is assumed to be the distributional form of the mean-variance relationship.
</p>
</div>
<p>Using the <code>denoisePCA</code> function, we denoise the log-expression data by removing the principal components that correspond to technical noise.</p>
<div class="sourceCode" id="cb40"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb40-1" data-line-number="1">denoised_sce&lt;-<span class="st"> </span><span class="kw">denoisePCA</span>(hpf18_sce, <span class="dt">technical=</span>dec_sce, <span class="dt">subset.row=</span><span class="kw">getTopHVGs</span>(dec_sce, <span class="dt">prop=</span><span class="fl">0.1</span>))</a>
<a class="sourceLine" id="cb40-2" data-line-number="2"><span class="kw">ncol</span>(<span class="kw">reducedDim</span>(denoised_sce)) </a></code></pre></div>
<pre><code>## [1] 12</code></pre>
<div class="sourceCode" id="cb42"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb42-1" data-line-number="1">denoised_sce_pois&lt;-<span class="st"> </span><span class="kw">denoisePCA</span>(hpf18_sce, <span class="dt">technical=</span>dec_pois_pbmc, <span class="dt">subset.row=</span><span class="kw">getTopHVGs</span>(dec_pois_pbmc, <span class="dt">prop=</span><span class="fl">0.1</span>))</a>
<a class="sourceLine" id="cb42-2" data-line-number="2"><span class="kw">ncol</span>(<span class="kw">reducedDim</span>(denoised_sce_pois)) <span class="co"># We hit the cap on the number of PCs</span></a></code></pre></div>
<pre><code>## [1] 50</code></pre>
<p>We can visualise the loadings (coefficients in the linear combination) of the top genes in each principal component with Seurat’s <code>VizDimLoadings</code> function.</p>
<p>We can also plot a heatmap for each principal component. This heatmap sorts cells and genes by their principal component scores, allowing us to identify the drivers of variation in each PC.</p>
<div class="sourceCode" id="cb44"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb44-1" data-line-number="1"><span class="co">#Seurat::VizDimLoadings(hpf18_seurat, dims = 1:10, reduction = &quot;pca&quot;)</span></a>
<a class="sourceLine" id="cb44-2" data-line-number="2">Seurat<span class="op">::</span><span class="kw">DimHeatmap</span>(hpf18_seurat, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">2</span>, <span class="dt">cells =</span> <span class="dv">500</span>, <span class="dt">balanced =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/PCA_dim_visual-1.png" alt="Figure 4- PCA heatmaps of the first two components" width="672" />
<p class="caption">
(#fig:PCA_dim_visual)Figure 4- PCA heatmaps of the first two components
</p>
</div>
</div>
</div>
<div id="evaluating-clustering-results" class="section level2">
<h2><span class="header-section-number">5.4</span> Evaluating clustering results</h2>
<p>The explanation and code below is adapted from <a href="https://osca.bioconductor.org/clustering.html#cluster-bootstrapping">osca</a></p>
<div id="cluster-stability" class="section level3">
<h3><span class="header-section-number">5.4.1</span> Cluster stability</h3>
<p>Stable clusters are desirable as small changes in our data preprocessing will not cause a major change in the cell assignment. To do this, scran performs sampling with replacement (bootstrap) on the dataset with the <code>bootstrapCluster</code> function and then recomputes the clustering on this newly sampled dataset.</p>
<p>The heatmap displays a matrix (we call it <span class="math inline">\(A\)</span>) of coassignment probabilities. The entry <span class="math inline">\(A_{ij}\)</span> in <span class="math inline">\(A\)</span> corresponds to the probability that a randomly chosen cell from cluster <span class="math inline">\(i\)</span> is found in cluster <span class="math inline">\(j\)</span>. In the extreme case, a perfect clustering would have a probability of 1 for all diagonal entries, and 0 elsewhere.</p>
<div class="sourceCode" id="cb45"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb45-1" data-line-number="1">seurat_clusters &lt;-<span class="st"> </span><span class="kw">colData</span>(hpf18_sce)<span class="op">$</span>seurat_clusters</a>
<a class="sourceLine" id="cb45-2" data-line-number="2"></a>
<a class="sourceLine" id="cb45-3" data-line-number="3">myClusterFun &lt;-<span class="st"> </span><span class="cf">function</span>(x){</a>
<a class="sourceLine" id="cb45-4" data-line-number="4">  g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(x, <span class="dt">use.dimred =</span> <span class="st">&quot;PCA&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;jaccard&quot;</span>)</a>
<a class="sourceLine" id="cb45-5" data-line-number="5">  igraph<span class="op">::</span><span class="kw">cluster_louvain</span>(g)<span class="op">$</span>membership</a>
<a class="sourceLine" id="cb45-6" data-line-number="6">}</a>
<a class="sourceLine" id="cb45-7" data-line-number="7"></a>
<a class="sourceLine" id="cb45-8" data-line-number="8"><span class="kw">set.seed</span>(<span class="dv">2020</span>)</a>
<a class="sourceLine" id="cb45-9" data-line-number="9">coassign &lt;-<span class="st"> </span><span class="kw">bootstrapCluster</span>(hpf18_sce, <span class="dt">FUN =</span> myClusterFun, <span class="dt">clusters =</span> seurat_clusters)</a>
<a class="sourceLine" id="cb45-10" data-line-number="10"></a>
<a class="sourceLine" id="cb45-11" data-line-number="11"><span class="kw">pheatmap</span>(coassign, <span class="dt">cluster_row=</span><span class="ot">FALSE</span>, <span class="dt">cluster_col=</span><span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb45-12" data-line-number="12">    <span class="dt">color=</span><span class="kw">rev</span>(viridis<span class="op">::</span><span class="kw">magma</span>(<span class="dv">100</span>)))</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/clust_stability-1.png" alt="Figure 5- Heatmap of coassignment probabilities from bootstrapping of the dataset and reclustering cells " width="672" />
<p class="caption">
(#fig:clust_stability)Figure 5- Heatmap of coassignment probabilities from bootstrapping of the dataset and reclustering cells
</p>
</div>
</div>
<div id="cluster-separability" class="section level3">
<h3><span class="header-section-number">5.4.2</span> Cluster separability</h3>
<p>Ideally, our clusters are well-separated from each other. The definition of modularity was provided in chapter 2. In essence, it is the difference between the observed edge weights between nodes belonging to the same cluster and expected edge weights if the edges were wired randomly. A high modulaity score would imply that most edges occur between cells in the same cluster, suggesting that the clusters are well-separated as edges are not formed between cells of different clusters.</p>
<p>Recall from our previous discussion that in the SNN graph, an edge is drawn between two cells if they have a neighbor in common. The edge weight describes the degree of similarity via the rank of its shared neighbors.</p>
<div class="sourceCode" id="cb46"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb46-1" data-line-number="1">g &lt;-<span class="st"> </span><span class="kw">buildSNNGraph</span>(hpf18_sce, <span class="dt">use.dimred =</span> <span class="st">&quot;PCA&quot;</span>, <span class="dt">type =</span> <span class="st">&quot;jaccard&quot;</span>)</a>
<a class="sourceLine" id="cb46-2" data-line-number="2">ratio &lt;-<span class="st"> </span><span class="kw">clusterModularity</span>(g, seurat_clusters, <span class="dt">as.ratio =</span> <span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb46-3" data-line-number="3"><span class="kw">pheatmap</span>(<span class="kw">log2</span>(ratio<span class="op">+</span><span class="dv">1</span>), <span class="dt">cluster_rows=</span><span class="ot">FALSE</span>, <span class="dt">cluster_cols=</span><span class="ot">FALSE</span>,</a>
<a class="sourceLine" id="cb46-4" data-line-number="4">    <span class="dt">color=</span><span class="kw">colorRampPalette</span>(<span class="kw">c</span>(<span class="st">&quot;white&quot;</span>, <span class="st">&quot;blue&quot;</span>))(<span class="dv">100</span>))</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/clust_separation-1.png" alt="Figure 6- Heatmap of the cluster modularity. Each entry along the heatmap diagonal represents the ratio of the total weight between nodes in the same cluster relative to a null model of randomly connected edges." width="672" />
<p class="caption">
(#fig:clust_separation)Figure 6- Heatmap of the cluster modularity. Each entry along the heatmap diagonal represents the ratio of the total weight between nodes in the same cluster relative to a null model of randomly connected edges.
</p>
</div>
<p>We can create a graph in which the nodes are clusters, instead of cells. Here, a high edge weight is represented by a thicker edge, and corresponds to a greater degree of reallocation of cells between the 2 clusters.</p>
<div class="sourceCode" id="cb47"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb47-1" data-line-number="1">cluster.gr &lt;-<span class="st"> </span>igraph<span class="op">::</span><span class="kw">graph_from_adjacency_matrix</span>(<span class="kw">log2</span>(ratio<span class="op">+</span><span class="dv">1</span>), </a>
<a class="sourceLine" id="cb47-2" data-line-number="2">    <span class="dt">mode=</span><span class="st">&quot;upper&quot;</span>, <span class="dt">weighted=</span><span class="ot">TRUE</span>, <span class="dt">diag=</span><span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb47-3" data-line-number="3"></a>
<a class="sourceLine" id="cb47-4" data-line-number="4"><span class="co"># Increasing the weight to increase the visibility of the lines.</span></a>
<a class="sourceLine" id="cb47-5" data-line-number="5"><span class="kw">set.seed</span>(<span class="dv">2020</span>)</a>
<a class="sourceLine" id="cb47-6" data-line-number="6"><span class="kw">plot</span>(cluster.gr, <span class="dt">edge.width=</span>igraph<span class="op">::</span><span class="kw">E</span>(cluster.gr)<span class="op">$</span>weight<span class="op">*</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb47-7" data-line-number="7">    <span class="dt">layout=</span>igraph<span class="op">::</span>layout_with_fr)</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/graph_adj-1.png" alt="Figure 7- Force-based layout depicting the relationships between clusters. Each node is a cluster, and the edge weight represents the ratio of the observed to expected sum of weights between each pair of clusters." width="672" />
<p class="caption">
(#fig:graph_adj)Figure 7- Force-based layout depicting the relationships between clusters. Each node is a cluster, and the edge weight represents the ratio of the observed to expected sum of weights between each pair of clusters.
</p>
</div>
</div>
</div>
<div id="comparison-with-authors-labels" class="section level2">
<h2><span class="header-section-number">5.5</span> Comparison with authors’ labels</h2>
<div id="concordance-in-tailbud-cluster-assignment" class="section level3">
<h3><span class="header-section-number">5.5.1</span> Concordance in tailbud cluster assignment</h3>
<p>In this section, we seek the answer to the question: to what extent does our clustering labels correspond to the authors’ original clustering results? In the original paper, the authors used t-SNE to reduce the dimensions of the data, followed by clustering the cells on this reduced dimensions with density peak clustering.</p>
<p>In our pipeline, we utilised Seurat’s approach of first constructing a shared-nearest neighour graph (SNN graph) using the PCA components, prior to clustering with the louvain algorithm, a graph-based clustering method. To what extent are these methods concordant?</p>
<p>In the table below, the <code>clusterNames</code> column indicates the cell’s original cluster label. Recall that the identities of these two clusters (tailbud-PSM and tailbud-spinal cord) correspond to the cluster name that occurs most frequently. Therefore, with the exception of a perfect correspondence with the authors’ clusters, we expect that each cluster should be somewhat heterogeneous in the cell labels as some of the cell labels should differ with the assigned cluster labels. The <code>seurat_clusters_names</code> column indicates the assigned cluster name.</p>
<div class="sourceCode" id="cb48"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb48-1" data-line-number="1">allLabelsAgainstClusters &lt;-<span class="st"> </span>hpf18_seurat<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb48-2" data-line-number="2"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(clusterNames, seurat_clusters, seurat_clusters_names)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-3" data-line-number="3"><span class="st">  </span><span class="kw">na.omit</span>()<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-4" data-line-number="4"><span class="st">  </span><span class="kw">group_by</span>(clusterNames, seurat_clusters_names) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-5" data-line-number="5"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">Cell_Numbers=</span><span class="kw">n</span>())<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-6" data-line-number="6"><span class="st">  </span><span class="kw">group_by</span>(seurat_clusters_names)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-7" data-line-number="7"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">Percentage_seurat =</span> Cell_Numbers <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(Cell_Numbers)<span class="op">*</span><span class="dv">100</span>)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-8" data-line-number="8"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(seurat_clusters_names <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - PSM&quot;</span> <span class="op">|</span></a>
<a class="sourceLine" id="cb48-9" data-line-number="9"><span class="st">                </span>seurat_clusters_names <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - spinal cord&quot;</span>)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-10" data-line-number="10"><span class="st">  </span><span class="kw">ungroup</span>()<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb48-11" data-line-number="11"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">complete</span>(clusterNames, seurat_clusters_names, <span class="dt">fill =</span> <span class="kw">list</span>(<span class="dt">Cell_Numbers =</span> <span class="dv">0</span>, <span class="dt">Percentage_seurat =</span> <span class="dv">0</span>))</a>
<a class="sourceLine" id="cb48-12" data-line-number="12"></a>
<a class="sourceLine" id="cb48-13" data-line-number="13">allLabelsAgainstClusters</a></code></pre></div>
<pre><code>## # A tibble: 20 x 4
##    clusterNames              seurat_clusters_names Cell_Numbers Percentage_seur~
##    &lt;chr&gt;                     &lt;chr&gt;                        &lt;dbl&gt;            &lt;dbl&gt;
##  1 18hpf-epidermal - anteri~ 18hpf-tailbud - PSM              0            0    
##  2 18hpf-epidermal - anteri~ 18hpf-tailbud - spin~            1            0.450
##  3 18hpf-neural - dorsal hi~ 18hpf-tailbud - PSM              0            0    
##  4 18hpf-neural - dorsal hi~ 18hpf-tailbud - spin~            7            3.15 
##  5 18hpf-neural - hindbrain~ 18hpf-tailbud - PSM              0            0    
##  6 18hpf-neural - hindbrain~ 18hpf-tailbud - spin~            2            0.901
##  7 18hpf-neural - midbrain   18hpf-tailbud - PSM              2            0.528
##  8 18hpf-neural - midbrain   18hpf-tailbud - spin~            0            0    
##  9 18hpf-neural - telenceph~ 18hpf-tailbud - PSM              0            0    
## 10 18hpf-neural - telenceph~ 18hpf-tailbud - spin~            1            0.450
## 11 18hpf-neural - ventral h~ 18hpf-tailbud - PSM              1            0.264
## 12 18hpf-neural - ventral h~ 18hpf-tailbud - spin~            7            3.15 
## 13 18hpf-notocord            18hpf-tailbud - PSM              2            0.528
## 14 18hpf-notocord            18hpf-tailbud - spin~            0            0    
## 15 18hpf-optic cup           18hpf-tailbud - PSM              0            0    
## 16 18hpf-optic cup           18hpf-tailbud - spin~            2            0.901
## 17 18hpf-tailbud - PSM       18hpf-tailbud - PSM            371           97.9  
## 18 18hpf-tailbud - PSM       18hpf-tailbud - spin~            7            3.15 
## 19 18hpf-tailbud - spinal c~ 18hpf-tailbud - PSM              3            0.792
## 20 18hpf-tailbud - spinal c~ 18hpf-tailbud - spin~          195           87.8</code></pre>
<p>We visualise our results with a ggplot heatmap using the <code>geom_tile</code> function.</p>
<div class="sourceCode" id="cb50"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb50-1" data-line-number="1">a1 &lt;-<span class="st"> </span><span class="kw">ggplot</span>(allLabelsAgainstClusters, <span class="kw">aes</span>(<span class="dt">x =</span> seurat_clusters_names, <span class="dt">y =</span> clusterNames, <span class="dt">fill =</span> Cell_Numbers)) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb50-2" data-line-number="2"><span class="st">  </span><span class="kw">geom_tile</span>(<span class="kw">aes</span>(<span class="dt">fill =</span> Cell_Numbers), <span class="dt">size=</span><span class="fl">0.5</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb50-3" data-line-number="3"><span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label=</span>Cell_Numbers), <span class="dt">alpha=</span><span class="fl">1.0</span>, <span class="dt">size=</span><span class="dv">3</span>, <span class="dt">colour =</span> <span class="st">&quot;gray78&quot;</span>)<span class="op">+</span></a>
<a class="sourceLine" id="cb50-4" data-line-number="4"><span class="st">  </span><span class="kw">labs</span>(<span class="dt">x=</span><span class="st">&quot;Clusters&quot;</span>,<span class="dt">y=</span><span class="st">&quot;Author&#39;s labels&quot;</span>)</a>
<a class="sourceLine" id="cb50-5" data-line-number="5"></a>
<a class="sourceLine" id="cb50-6" data-line-number="6">a2 &lt;-<span class="st"> </span>a1 <span class="op">+</span>viridis<span class="op">::</span><span class="kw">scale_fill_viridis</span>(<span class="dt">name =</span> <span class="st">&quot;Cell Number&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb50-7" data-line-number="7"><span class="st">  </span><span class="kw">scale_x_discrete</span>(<span class="dt">expand=</span><span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">0</span>), <span class="dt">name =</span> <span class="st">&quot;Clusters&quot;</span>, <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;18hpf-tailbud - PSM&quot;</span>, <span class="st">&quot;18hpf-tailbud - spinal cord&quot;</span>))<span class="op">+</span></a>
<a class="sourceLine" id="cb50-8" data-line-number="8"><span class="st">  </span><span class="kw">theme_bw</span>(<span class="dt">base_size=</span><span class="dv">10</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb50-9" data-line-number="9"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">legend.text=</span><span class="kw">element_text</span>(<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>), <span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">hjust =</span> <span class="fl">0.5</span>),</a>
<a class="sourceLine" id="cb50-10" data-line-number="10">        <span class="dt">panel.border=</span><span class="kw">element_blank</span>(), <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>))</a>
<a class="sourceLine" id="cb50-11" data-line-number="11"></a>
<a class="sourceLine" id="cb50-12" data-line-number="12">a2</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/concordance%20heatmap-1.png" alt="Figure 8- Heatmap of the authors cell labels against our assigned cluster labels. The majority of cells in our assigned tailbud clusters were also mapped to the same clusters in the authors' original clustering." width="672" />
<p class="caption">
(#fig:concordance heatmap)Figure 8- Heatmap of the authors cell labels against our assigned cluster labels. The majority of cells in our assigned tailbud clusters were also mapped to the same clusters in the authors’ original clustering.
</p>
</div>
</div>
<div id="optimising-the-seurat-clustering-parameters" class="section level3">
<h3><span class="header-section-number">5.5.2</span> Optimising the seurat clustering parameters</h3>
<p>One reason why we observe a disparity between the clustering results could be because our chosen parameters were not optimised for the correspondence. In other words, it is possible that the degree of concordance could be improved with a better clustering resolution in the <code>FindClusters</code> function for example.</p>
<p>The function below first assigns a clustername to the clusters identified from seurat. Then, we filter for cells that are assigned to the two tailbud clusters, either in our assignment or the authors’ original assignment. Finally, we group cells into one of 3 categories:
- Present in the tailbud clusters from our assignment but not in the authors (<code>in_mine_only</code>)
- Present in the tailbud clusters from the authors’ dataset but not in ours (<code>in_authors_only</code>)
- Present in the tailbud clusters from both approaches (<code>in_both</code>)</p>
<div class="sourceCode" id="cb51"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb51-1" data-line-number="1">myoptim &lt;-<span class="st"> </span><span class="cf">function</span>(reso,myObj){</a>
<a class="sourceLine" id="cb51-2" data-line-number="2">  </a>
<a class="sourceLine" id="cb51-3" data-line-number="3">  myObj &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">FindNeighbors</span>(hpf18_seurat,  <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">12</span>, <span class="dt">k.param =</span> <span class="dv">20</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb51-4" data-line-number="4">  myObj &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">FindClusters</span>(hpf18_seurat, <span class="dt">algorithm =</span> <span class="dv">1</span>, <span class="dt">resolution =</span> reso, <span class="dt">verbose =</span> <span class="ot">FALSE</span>) </a>
<a class="sourceLine" id="cb51-5" data-line-number="5">  </a>
<a class="sourceLine" id="cb51-6" data-line-number="6">  <span class="co">#Finding correspondence between cluster numbers AND authors&#39; labels</span></a>
<a class="sourceLine" id="cb51-7" data-line-number="7">  clusterIDlabels &lt;-<span class="st"> </span>myObj<span class="op">@</span>meta.data <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb51-8" data-line-number="8"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(clusterNames, seurat_clusters)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-9" data-line-number="9"><span class="st">  </span><span class="kw">na.omit</span>()  <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-10" data-line-number="10"><span class="st">  </span><span class="kw">group_by</span>(clusterNames, seurat_clusters) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-11" data-line-number="11"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">Cell_Numbers=</span><span class="kw">n</span>())<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-12" data-line-number="12"><span class="st">  </span>tidyr<span class="op">::</span><span class="kw">complete</span>(seurat_clusters)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-13" data-line-number="13"><span class="st">  </span><span class="kw">replace_na</span>(<span class="kw">list</span>(<span class="dt">Cell_Numbers =</span> <span class="dv">0</span>))<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-14" data-line-number="14"><span class="st">  </span><span class="kw">group_by</span>(seurat_clusters)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-15" data-line-number="15"><span class="st">  </span><span class="kw">top_n</span>(<span class="dt">n=</span><span class="dv">1</span>, <span class="dt">wt=</span>Cell_Numbers) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-16" data-line-number="16"><span class="st">  </span><span class="kw">arrange</span>(seurat_clusters) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-17" data-line-number="17"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(seurat_clusters, clusterNames)</a>
<a class="sourceLine" id="cb51-18" data-line-number="18"></a>
<a class="sourceLine" id="cb51-19" data-line-number="19">  <span class="co"># Assigning mapping to seurat object</span></a>
<a class="sourceLine" id="cb51-20" data-line-number="20">  myObj<span class="op">@</span>meta.data<span class="op">$</span>seurat_clusters_names &lt;-<span class="st"> </span>clusterIDlabels<span class="op">$</span>clusterNames[<span class="kw">match</span>(myObj<span class="op">@</span>meta.data<span class="op">$</span>seurat_clusters,  clusterIDlabels<span class="op">$</span>seurat_clusters)]</a>
<a class="sourceLine" id="cb51-21" data-line-number="21">  </a>
<a class="sourceLine" id="cb51-22" data-line-number="22">  </a>
<a class="sourceLine" id="cb51-23" data-line-number="23">  <span class="co"># Selecting cells that are either assigned the authors&#39;TB labels OR fall into the TB clusters</span></a>
<a class="sourceLine" id="cb51-24" data-line-number="24">  myObj_subsetted &lt;-<span class="st"> </span>myObj<span class="op">@</span>meta.data <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-25" data-line-number="25"><span class="st">    </span><span class="kw">rownames_to_column</span>(<span class="st">&quot;cell&quot;</span>)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-26" data-line-number="26"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">filter</span>(seurat_clusters_names <span class="op">==</span><span class="st">    &quot;18hpf-tailbud - PSM&quot;</span> <span class="op">|</span><span class="st"> </span></a>
<a class="sourceLine" id="cb51-27" data-line-number="27"><span class="st">                  </span>seurat_clusters_names <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - spinal cord&quot;</span> <span class="op">|</span></a>
<a class="sourceLine" id="cb51-28" data-line-number="28"><span class="st">                  </span>clusterNames <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - PSM&quot;</span> <span class="op">|</span></a>
<a class="sourceLine" id="cb51-29" data-line-number="29"><span class="st">                  </span>clusterNames <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - spinal cord&quot;</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-30" data-line-number="30"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">mutate</span>(<span class="dt">WithinTBCluster =</span> <span class="kw">ifelse</span>((seurat_clusters_names <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - PSM&quot;</span> <span class="op">|</span>seurat_clusters_names <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - spinal cord&quot;</span>) <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb51-31" data-line-number="31"><span class="st">                                           </span><span class="op">!</span>(clusterNames <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - PSM&quot;</span> <span class="op">|</span></a>
<a class="sourceLine" id="cb51-32" data-line-number="32"><span class="st">                                             </span>clusterNames <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - spinal cord&quot;</span>),</a>
<a class="sourceLine" id="cb51-33" data-line-number="33">                                            <span class="st">&quot;in_mine_only&quot;</span>, </a>
<a class="sourceLine" id="cb51-34" data-line-number="34">                                           <span class="kw">ifelse</span>(<span class="op">!</span>(seurat_clusters_names <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - PSM&quot;</span> <span class="op">|</span></a>
<a class="sourceLine" id="cb51-35" data-line-number="35"><span class="st">                                           </span>seurat_clusters_names <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - spinal cord&quot;</span>) <span class="op">&amp;</span></a>
<a class="sourceLine" id="cb51-36" data-line-number="36"><span class="st">                                            </span></a>
<a class="sourceLine" id="cb51-37" data-line-number="37"><span class="st">                                             </span>(clusterNames <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - PSM&quot;</span> <span class="op">|</span></a>
<a class="sourceLine" id="cb51-38" data-line-number="38"><span class="st">                                             </span>clusterNames <span class="op">==</span><span class="st"> &quot;18hpf-tailbud - spinal cord&quot;</span>),</a>
<a class="sourceLine" id="cb51-39" data-line-number="39">                                           <span class="st">&quot;in_authors_only&quot;</span>, <span class="st">&quot;in_both&quot;</span>)), </a>
<a class="sourceLine" id="cb51-40" data-line-number="40">                                            <span class="dt">reso_used =</span> reso) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb51-41" data-line-number="41"><span class="st">    </span>dplyr<span class="op">::</span><span class="kw">select</span>(cell, clusterNames, seurat_clusters_names, WithinTBCluster,reso_used) </a>
<a class="sourceLine" id="cb51-42" data-line-number="42">  <span class="kw">return</span> (myObj_subsetted)</a>
<a class="sourceLine" id="cb51-43" data-line-number="43">}</a></code></pre></div>
<p>We provide a range of clustering resolutions to assess their impact on the concordance rate.</p>
<div class="sourceCode" id="cb52"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb52-1" data-line-number="1"><span class="co">#Inputting parameters</span></a>
<a class="sourceLine" id="cb52-2" data-line-number="2"></a>
<a class="sourceLine" id="cb52-3" data-line-number="3">cluster_reso &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="fl">0.4</span>,<span class="fl">0.8</span>, <span class="fl">1.2</span>, <span class="fl">1.6</span>, <span class="fl">2.0</span>, <span class="fl">2.4</span>)</a>
<a class="sourceLine" id="cb52-4" data-line-number="4"></a>
<a class="sourceLine" id="cb52-5" data-line-number="5">myDF &lt;-<span class="st"> </span><span class="kw">mapply</span>(myoptim, <span class="dt">reso =</span> cluster_reso, <span class="dt">MoreArgs =</span> <span class="kw">list</span>(<span class="dt">myObj=</span><span class="st">&quot;hpf18_seurat&quot;</span>),</a>
<a class="sourceLine" id="cb52-6" data-line-number="6">               <span class="dt">SIMPLIFY =</span> <span class="ot">FALSE</span>) </a>
<a class="sourceLine" id="cb52-7" data-line-number="7"></a>
<a class="sourceLine" id="cb52-8" data-line-number="8">myDF &lt;-<span class="st"> </span><span class="kw">do.call</span>(rbind, myDF)</a></code></pre></div>
<p>To assess the clustering result, we examine the following two measures:</p>
<ol style="list-style-type: decimal">
<li><a href="https://davetang.org/muse/2017/09/21/the-rand-index/">Adjusted Rand Index</a></li>
<li><a href="https://stackoverflow.com/questions/35709562/how-to-calculate-clustering-entropy-a-working-example-or-software-code">Entropy</a></li>
</ol>
<div class="sourceCode" id="cb53"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb53-1" data-line-number="1">clustering_optim &lt;-<span class="st"> </span>myDF <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-2" data-line-number="2"><span class="st">  </span><span class="kw">group_by</span>(reso_used, WithinTBCluster)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-3" data-line-number="3"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">numbers =</span> <span class="kw">n</span>())<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-4" data-line-number="4"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">percentage =</span> numbers <span class="op">/</span><span class="st"> </span><span class="kw">sum</span>(numbers) <span class="op">*</span><span class="st"> </span><span class="dv">100</span>)</a>
<a class="sourceLine" id="cb53-5" data-line-number="5"></a>
<a class="sourceLine" id="cb53-6" data-line-number="6">adjusted_rand_index &lt;-<span class="st"> </span>myDF <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-7" data-line-number="7"><span class="st">  </span><span class="kw">group_by</span>(reso_used) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-8" data-line-number="8"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">ARI =</span> <span class="kw">adjustedRandIndex</span>(seurat_clusters_names, clusterNames)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-9" data-line-number="9"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(reso_used, ARI) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-10" data-line-number="10"><span class="st">  </span><span class="kw">distinct</span>() </a>
<a class="sourceLine" id="cb53-11" data-line-number="11"></a>
<a class="sourceLine" id="cb53-12" data-line-number="12">entropy &lt;-<span class="st"> </span>myDF <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-13" data-line-number="13"><span class="st">  </span><span class="kw">group_by</span>(reso_used)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-14" data-line-number="14"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">entropy =</span> NMF<span class="op">::</span><span class="kw">entropy</span>(<span class="kw">as.factor</span>(seurat_clusters_names),   <span class="kw">as.factor</span>(clusterNames), <span class="dt">method =</span> <span class="st">&quot;mean&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-15" data-line-number="15"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(reso_used,entropy) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb53-16" data-line-number="16"><span class="st">  </span><span class="kw">distinct</span>()</a>
<a class="sourceLine" id="cb53-17" data-line-number="17"></a>
<a class="sourceLine" id="cb53-18" data-line-number="18">clustering_optim &lt;-<span class="st"> </span><span class="kw">list</span>(adjusted_rand_index, clustering_optim, entropy) <span class="op">%&gt;%</span><span class="st"> </span>purrr<span class="op">::</span><span class="kw">reduce</span>(left_join, <span class="dt">by =</span> <span class="st">&quot;reso_used&quot;</span>) </a>
<a class="sourceLine" id="cb53-19" data-line-number="19"></a>
<a class="sourceLine" id="cb53-20" data-line-number="20">clustering_optim<span class="op">$</span>ARI[<span class="kw">duplicated</span>(clustering_optim<span class="op">$</span>ARI)] &lt;-<span class="st"> </span><span class="ot">NA</span></a>
<a class="sourceLine" id="cb53-21" data-line-number="21">clustering_optim<span class="op">$</span>entropy[<span class="kw">duplicated</span>(clustering_optim<span class="op">$</span>entropy)] &lt;-<span class="st"> </span><span class="ot">NA</span></a>
<a class="sourceLine" id="cb53-22" data-line-number="22">clustering_optim<span class="op">$</span>reso_used &lt;-<span class="kw">paste</span>(<span class="st">&quot;res=&quot;</span>, clustering_optim<span class="op">$</span>reso_used)</a>
<a class="sourceLine" id="cb53-23" data-line-number="23"></a>
<a class="sourceLine" id="cb53-24" data-line-number="24"><span class="co"># Adjusting order of factors</span></a>
<a class="sourceLine" id="cb53-25" data-line-number="25">clustering_optim<span class="op">$</span>WithinTBCluster &lt;-<span class="st"> </span><span class="kw">factor</span>(clustering_optim<span class="op">$</span>WithinTBCluster, </a>
<a class="sourceLine" id="cb53-26" data-line-number="26">                                           <span class="dt">levels =</span> <span class="kw">c</span>(<span class="st">&quot;in_both&quot;</span>, <span class="st">&quot;in_authors_only&quot;</span>, <span class="st">&quot;in_mine_only&quot;</span>))</a></code></pre></div>
<div class="sourceCode" id="cb54"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb54-1" data-line-number="1"><span class="co"># Facet plot: proportion falling into diff categories while varying k and res</span></a>
<a class="sourceLine" id="cb54-2" data-line-number="2"></a>
<a class="sourceLine" id="cb54-3" data-line-number="3">multiplot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(clustering_optim, <span class="kw">aes</span>(<span class="dt">x =</span> WithinTBCluster, <span class="dt">y =</span> percentage)) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">data =</span> clustering_optim, <span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">fill =</span> WithinTBCluster), <span class="dt">stat =</span>     </a>
<a class="sourceLine" id="cb54-5" data-line-number="5">             <span class="st">&quot;identity&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-6" data-line-number="6"><span class="st">  </span><span class="kw">facet_wrap</span>(<span class="kw">vars</span>(reso_used), <span class="dt">ncol=</span><span class="dv">3</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-7" data-line-number="7"><span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label =</span> <span class="kw">sprintf</span>(<span class="st">&quot;%.1f&quot;</span>, percentage), <span class="dt">group =</span> WithinTBCluster),<span class="dt">angle=</span><span class="dv">0</span>,  </a>
<a class="sourceLine" id="cb54-8" data-line-number="8">            <span class="dt">size =</span> <span class="dv">3</span>, <span class="dt">position =</span> <span class="kw">position_dodge</span>(<span class="dt">width =</span> <span class="dv">1</span>),<span class="dt">vjust =</span> <span class="fl">0.7</span>, <span class="dt">colour =</span> </a>
<a class="sourceLine" id="cb54-9" data-line-number="9">            <span class="st">&quot;grey5&quot;</span>,<span class="dt">fontface =</span> <span class="st">&quot;bold&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-10" data-line-number="10"><span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label =</span> <span class="kw">ifelse</span>(<span class="kw">is.na</span>(ARI), <span class="st">&quot;&quot;</span>,  <span class="kw">paste</span>(<span class="st">&quot;ARI =&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%.3f&quot;</span>, ARI), <span class="dt">sep =</span><span class="st">&quot; &quot;</span>)), <span class="dt">group =</span> reso_used),<span class="dt">inherit.aes=</span><span class="ot">FALSE</span>, <span class="dt">x =</span> <span class="fl">2.7</span>, <span class="dt">y=</span> <span class="dv">80</span>, <span class="dt">size =</span><span class="dv">3</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-11" data-line-number="11"><span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label =</span> <span class="kw">ifelse</span>(<span class="kw">is.na</span>(entropy), <span class="st">&quot;&quot;</span>,  <span class="kw">paste</span>(<span class="st">&quot;H =&quot;</span>, <span class="kw">sprintf</span>(<span class="st">&quot;%.3f&quot;</span>, </a>
<a class="sourceLine" id="cb54-12" data-line-number="12">                                                                         entropy), sep </a>
<a class="sourceLine" id="cb54-13" data-line-number="13">                                                          =<span class="st">&quot; &quot;</span>)), <span class="dt">group =</span> reso_used), </a>
<a class="sourceLine" id="cb54-14" data-line-number="14">            <span class="dt">inherit.aes=</span><span class="ot">FALSE</span>, <span class="dt">x =</span> <span class="fl">2.7</span>, <span class="dt">y=</span> <span class="dv">70</span>, <span class="dt">size =</span><span class="dv">3</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-15" data-line-number="15"><span class="st">  </span><span class="kw">scale_fill_brewer</span>(<span class="dt">palette =</span> <span class="st">&quot;Accent&quot;</span>, <span class="dt">labels =</span> <span class="kw">c</span>(<span class="st">&quot;shared&quot;</span>, <span class="st">&quot;in authors&#39; cluster only&quot;</span>, </a>
<a class="sourceLine" id="cb54-16" data-line-number="16">                                                   <span class="st">&quot;in my assigned cluster only&quot;</span>),<span class="dt">name =</span> </a>
<a class="sourceLine" id="cb54-17" data-line-number="17">                    <span class="st">&quot;Cluster Assignments&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb54-18" data-line-number="18"><span class="st">  </span><span class="kw">theme_bw</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb54-19" data-line-number="19"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">60</span>, <span class="dt">hjust=</span><span class="dv">1</span>),<span class="dt">axis.title.x=</span><span class="kw">element_blank</span>(), <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>))</a>
<a class="sourceLine" id="cb54-20" data-line-number="20"></a>
<a class="sourceLine" id="cb54-21" data-line-number="21">multiplot</a></code></pre></div>
<div class="figure">
<img src="NMps_book_files/figure-html/facet_optim-1.png" alt="Figure 9- Facetted bar plots of the cluster assignments across 6 louvain clustering resolutions. The Adjusted rand index and entropy for the cluster assignments are displayed." width="672" />
<p class="caption">
(#fig:facet_optim)Figure 9- Facetted bar plots of the cluster assignments across 6 louvain clustering resolutions. The Adjusted rand index and entropy for the cluster assignments are displayed.
</p>
</div>
<div class="sourceCode" id="cb55"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb55-1" data-line-number="1">entropy &lt;-<span class="st"> </span>myDF <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb55-2" data-line-number="2"><span class="st">  </span><span class="kw">group_by</span>(reso_used)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb55-3" data-line-number="3"><span class="st">  </span><span class="kw">mutate</span>(<span class="dt">entropy =</span> NMF<span class="op">::</span><span class="kw">entropy</span>(<span class="kw">factor</span>(seurat_clusters_names), clusterNames, <span class="dt">method =</span> <span class="st">&quot;mean&quot;</span>)) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb55-4" data-line-number="4"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">select</span>(reso_used, entropy) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb55-5" data-line-number="5"><span class="st">  </span><span class="kw">distinct</span>()</a>
<a class="sourceLine" id="cb55-6" data-line-number="6"></a>
<a class="sourceLine" id="cb55-7" data-line-number="7">entropyPlot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> entropy, <span class="kw">aes</span>(<span class="dt">x =</span> reso_used, <span class="dt">y =</span> entropy)) <span class="op">+</span></a>
<a class="sourceLine" id="cb55-8" data-line-number="8"><span class="st">  </span><span class="kw">geom_point</span>(<span class="kw">aes</span>(<span class="dt">color =</span> <span class="kw">as.factor</span>(reso_used)))<span class="op">+</span></a>
<a class="sourceLine" id="cb55-9" data-line-number="9"><span class="st">  </span><span class="kw">geom_line</span>(<span class="dt">alpha =</span> <span class="fl">0.5</span>, <span class="dt">linetype =</span> <span class="st">&quot;dotted&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb55-10" data-line-number="10"><span class="st">  </span><span class="kw">scale_color_viridis_d</span>(<span class="st">&quot;Louvain</span><span class="ch">\n</span><span class="st">Resolution&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb55-11" data-line-number="11"><span class="st">  </span><span class="kw">scale_x_continuous</span>(<span class="dt">breaks =</span> cluster_reso)<span class="op">+</span></a>
<a class="sourceLine" id="cb55-12" data-line-number="12"><span class="st">  </span><span class="kw">scale_y_continuous</span>(<span class="dt">limits =</span> <span class="kw">c</span>(<span class="fl">0.07</span>,<span class="fl">0.3</span>))<span class="op">+</span></a>
<a class="sourceLine" id="cb55-13" data-line-number="13"><span class="st">  </span><span class="kw">theme_bw</span>()<span class="op">+</span></a>
<a class="sourceLine" id="cb55-14" data-line-number="14"><span class="st">  </span><span class="kw">theme</span>(<span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">angle=</span><span class="dv">90</span>, <span class="dt">hjust=</span><span class="dv">1</span>, <span class="dt">vjust =</span> <span class="dv">1</span>),</a>
<a class="sourceLine" id="cb55-15" data-line-number="15">        <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>))</a>
<a class="sourceLine" id="cb55-16" data-line-number="16">entropyPlot</a></code></pre></div>
<div class="figure"><span id="fig:entropy"></span>
<img src="NMps_book_files/figure-html/entropy-1.png" alt="Figure 10- Entropy plot across 6 louvain clustering resolutions." width="672" />
<p class="caption">
Figure 5.1: Figure 10- Entropy plot across 6 louvain clustering resolutions.
</p>
</div>
</div>
</div>
<div id="selecting-cells-for-downstream-analysis" class="section level2">
<h2><span class="header-section-number">5.6</span> Selecting cells for downstream analysis</h2>
<p>For our downstream analysis, we would like to keep cells that are consistently identifed across both analyses, across all clustering resolutions, to belong to the tailbud cluster. The code below calculates this.</p>
<div class="sourceCode" id="cb56"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb56-1" data-line-number="1"><span class="co">#myDF$WithinTBCluster &lt;- as.factor(myDF$WithinTBCluster)</span></a>
<a class="sourceLine" id="cb56-2" data-line-number="2"></a>
<a class="sourceLine" id="cb56-3" data-line-number="3">labelCount &lt;-<span class="st"> </span>myDF <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb56-4" data-line-number="4"><span class="st">  </span><span class="kw">group_by</span>(cell, WithinTBCluster)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb56-5" data-line-number="5"><span class="st">  </span><span class="kw">summarize</span>(<span class="dt">numbers =</span> <span class="kw">n</span>()) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb56-6" data-line-number="6"><span class="st">  </span><span class="kw">filter</span>(WithinTBCluster <span class="op">==</span><span class="st"> &quot;in_both&quot;</span>)</a>
<a class="sourceLine" id="cb56-7" data-line-number="7"></a>
<a class="sourceLine" id="cb56-8" data-line-number="8">max_count &lt;-<span class="st"> </span><span class="kw">max</span>(labelCount<span class="op">$</span>numbers)</a>
<a class="sourceLine" id="cb56-9" data-line-number="9">labelCountTable &lt;-<span class="st"> </span><span class="kw">table</span>(<span class="kw">factor</span>(labelCount<span class="op">$</span>numbers, <span class="dt">levels =</span> <span class="dv">0</span><span class="op">:</span>max_count))</a>
<a class="sourceLine" id="cb56-10" data-line-number="10">labelCountdf &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="st">&quot;Occurences&quot;</span> =<span class="st"> </span><span class="dv">0</span><span class="op">:</span>max_count,</a>
<a class="sourceLine" id="cb56-11" data-line-number="11">                           <span class="st">&quot;Number_of_Cells&quot;</span> =<span class="st"> </span><span class="kw">as.numeric</span>(labelCountTable))</a>
<a class="sourceLine" id="cb56-12" data-line-number="12">labelCountdf</a></code></pre></div>
<pre><code>##   Occurences Number_of_Cells
## 1          0               0
## 2          1              20
## 3          2              28
## 4          3               0
## 5          4               2
## 6          5               8
## 7          6             551</code></pre>
<div class="sourceCode" id="cb58"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb58-1" data-line-number="1">cols &lt;-<span class="st"> </span><span class="kw">c</span>(scales<span class="op">::</span><span class="kw">hue_pal</span>()(<span class="kw">nrow</span>(labelCountdf)), <span class="st">&quot;#FFFFFF&quot;</span>)</a>
<a class="sourceLine" id="cb58-2" data-line-number="2"></a>
<a class="sourceLine" id="cb58-3" data-line-number="3">labelCountPlot &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> labelCountdf, <span class="kw">aes</span>(<span class="dt">x=</span>Occurences, <span class="dt">y =</span> Number_of_Cells)) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_bar</span>(<span class="dt">mapping =</span> <span class="kw">aes</span>(<span class="dt">fill =</span> <span class="kw">as.factor</span>(Occurences)), <span class="dt">stat =</span> <span class="st">&quot;identity&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb58-5" data-line-number="5"><span class="st">  </span><span class="kw">geom_text</span>(<span class="kw">aes</span>(<span class="dt">label =</span> Number_of_Cells, <span class="dt">group =</span> Occurences),<span class="dt">angle=</span><span class="dv">0</span>, <span class="dt">size =</span> <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb58-6" data-line-number="6">            <span class="dt">colour =</span> <span class="st">&quot;grey5&quot;</span>, <span class="dt">vjust =</span> <span class="fl">-0.6</span>, <span class="dt">fontface =</span> <span class="st">&quot;bold&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-7" data-line-number="7"><span class="st">  </span><span class="kw">scale_fill_manual</span>(<span class="dt">values =</span> cols) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-8" data-line-number="8"><span class="st">  </span><span class="kw">scale_x_discrete</span>(<span class="dt">name =</span> <span class="st">&quot;Total number of iterations&quot;</span> , <span class="dt">limits =</span> <span class="kw">c</span>(<span class="dv">0</span><span class="op">:</span>max_count)) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-9" data-line-number="9"><span class="st">  </span><span class="kw">ylab</span>(<span class="dt">label =</span> <span class="st">&quot;Cell Number&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb58-10" data-line-number="10"><span class="st">  </span><span class="kw">scale_y_continuous</span>(<span class="dt">expand =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">0</span>), <span class="dt">limits =</span> <span class="kw">c</span>(<span class="dv">0</span>, <span class="dv">600</span>))<span class="op">+</span></a>
<a class="sourceLine" id="cb58-11" data-line-number="11"><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span></a>
<a class="sourceLine" id="cb58-12" data-line-number="12"><span class="st">  </span><span class="kw">theme</span>(</a>
<a class="sourceLine" id="cb58-13" data-line-number="13">    <span class="dt">legend.position =</span> <span class="st">&quot;none&quot;</span>,</a>
<a class="sourceLine" id="cb58-14" data-line-number="14">    <span class="dt">axis.line =</span> <span class="kw">element_line</span>(<span class="dt">colour =</span> <span class="st">&quot;black&quot;</span>),</a>
<a class="sourceLine" id="cb58-15" data-line-number="15">    <span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb58-16" data-line-number="16">    <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb58-17" data-line-number="17">    <span class="dt">panel.border =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb58-18" data-line-number="18">    <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>))</a>
<a class="sourceLine" id="cb58-19" data-line-number="19"></a>
<a class="sourceLine" id="cb58-20" data-line-number="20">labelCountPlot</a></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-18"></span>
<img src="NMps_book_files/figure-html/unnamed-chunk-18-1.png" alt="Figure 11- Bar plots of number of cells against the total number of iterations. The height of each bar represents the number of cells that are assigned to the tailbud clusters by both the authors' and my labels, for a given number of clustering resolutions. For instance, 551 cells were assigned by the authors' and me to belong to the tailbud cluster across all 6 clustering resolutions." width="672" />
<p class="caption">
Figure 5.2: Figure 11- Bar plots of number of cells against the total number of iterations. The height of each bar represents the number of cells that are assigned to the tailbud clusters by both the authors’ and my labels, for a given number of clustering resolutions. For instance, 551 cells were assigned by the authors’ and me to belong to the tailbud cluster across all 6 clustering resolutions.
</p>
</div>
<div class="sourceCode" id="cb59"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb59-1" data-line-number="1">myCells &lt;-<span class="st"> </span>labelCount[labelCount<span class="op">$</span>numbers <span class="op">==</span><span class="st"> </span>max_count, ]<span class="op">$</span>cell</a>
<a class="sourceLine" id="cb59-2" data-line-number="2"><span class="kw">length</span>(myCells)</a></code></pre></div>
<pre><code>## [1] 551</code></pre>
<div class="sourceCode" id="cb61"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb61-1" data-line-number="1">hpf18_seurat<span class="op">$</span>TB &lt;-<span class="st"> </span><span class="kw">colnames</span>(hpf18_seurat) <span class="op">%in%</span><span class="st"> </span>myCells <span class="co"># Creating this column to facilitate subclustering in chapter 5.</span></a>
<a class="sourceLine" id="cb61-2" data-line-number="2"><span class="kw">saveRDS</span>(hpf18_seurat, <span class="dt">file =</span> <span class="st">&quot;rds/hpf18_seurat_optimized551.rds&quot;</span> )</a></code></pre></div>
</div>
<div id="session-info-2" class="section level2">
<h2><span class="header-section-number">5.7</span> Session Info</h2>
<details>
<p><summary>View Session Info</summary></p>
<div class="sourceCode" id="cb62"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb62-1" data-line-number="1">devtools<span class="op">::</span><span class="kw">session_info</span>()</a>
<a class="sourceLine" id="cb62-2" data-line-number="2"><span class="co">## - Session info ---------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb62-3" data-line-number="3"><span class="co">##  setting  value                       </span></a>
<a class="sourceLine" id="cb62-4" data-line-number="4"><span class="co">##  version  R version 4.0.0 (2020-04-24)</span></a>
<a class="sourceLine" id="cb62-5" data-line-number="5"><span class="co">##  os       Windows 10 x64              </span></a>
<a class="sourceLine" id="cb62-6" data-line-number="6"><span class="co">##  system   x86_64, mingw32             </span></a>
<a class="sourceLine" id="cb62-7" data-line-number="7"><span class="co">##  ui       RTerm                       </span></a>
<a class="sourceLine" id="cb62-8" data-line-number="8"><span class="co">##  language (EN)                        </span></a>
<a class="sourceLine" id="cb62-9" data-line-number="9"><span class="co">##  collate  English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb62-10" data-line-number="10"><span class="co">##  ctype    English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb62-11" data-line-number="11"><span class="co">##  tz       Asia/Kuala_Lumpur           </span></a>
<a class="sourceLine" id="cb62-12" data-line-number="12"><span class="co">##  date     2020-06-03                  </span></a>
<a class="sourceLine" id="cb62-13" data-line-number="13"><span class="co">## </span></a>
<a class="sourceLine" id="cb62-14" data-line-number="14"><span class="co">## - Packages -------------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb62-15" data-line-number="15"><span class="co">##  package              * version  date       lib source        </span></a>
<a class="sourceLine" id="cb62-16" data-line-number="16"><span class="co">##  AnnotationDbi        * 1.50.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-17" data-line-number="17"><span class="co">##  ape                    5.3      2019-03-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-18" data-line-number="18"><span class="co">##  assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-19" data-line-number="19"><span class="co">##  backports              1.1.6    2020-04-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-20" data-line-number="20"><span class="co">##  beeswarm               0.2.3    2016-04-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-21" data-line-number="21"><span class="co">##  bibtex                 0.4.2.2  2020-01-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-22" data-line-number="22"><span class="co">##  Biobase              * 2.48.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-23" data-line-number="23"><span class="co">##  BiocGenerics         * 0.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-24" data-line-number="24"><span class="co">##  BiocNeighbors          1.6.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-25" data-line-number="25"><span class="co">##  BiocParallel           1.22.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-26" data-line-number="26"><span class="co">##  BiocSingular           1.4.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-27" data-line-number="27"><span class="co">##  bit                    1.1-15.2 2020-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-28" data-line-number="28"><span class="co">##  bit64                  0.9-7    2017-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-29" data-line-number="29"><span class="co">##  bitops                 1.0-6    2013-08-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-30" data-line-number="30"><span class="co">##  blob                   1.2.1    2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-31" data-line-number="31"><span class="co">##  bookdown               0.19     2020-05-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-32" data-line-number="32"><span class="co">##  broom                  0.5.6    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-33" data-line-number="33"><span class="co">##  callr                  3.4.3    2020-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-34" data-line-number="34"><span class="co">##  cellranger             1.1.0    2016-07-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-35" data-line-number="35"><span class="co">##  cli                    2.0.2    2020-02-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-36" data-line-number="36"><span class="co">##  cluster              * 2.1.0    2019-06-19 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-37" data-line-number="37"><span class="co">##  codetools              0.2-16   2018-12-24 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-38" data-line-number="38"><span class="co">##  colorspace             1.4-1    2019-03-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-39" data-line-number="39"><span class="co">##  cowplot                1.0.0    2019-07-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-40" data-line-number="40"><span class="co">##  crayon                 1.3.4    2017-09-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-41" data-line-number="41"><span class="co">##  data.table             1.12.8   2019-12-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-42" data-line-number="42"><span class="co">##  DBI                    1.1.0    2019-12-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-43" data-line-number="43"><span class="co">##  dbplyr                 1.4.3    2020-04-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-44" data-line-number="44"><span class="co">##  DelayedArray         * 0.14.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-45" data-line-number="45"><span class="co">##  DelayedMatrixStats     1.10.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-46" data-line-number="46"><span class="co">##  desc                   1.2.0    2018-05-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-47" data-line-number="47"><span class="co">##  devtools               2.3.0    2020-04-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-48" data-line-number="48"><span class="co">##  digest                 0.6.25   2020-02-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-49" data-line-number="49"><span class="co">##  doParallel             1.0.15   2019-08-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-50" data-line-number="50"><span class="co">##  dplyr                * 0.8.5    2020-03-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-51" data-line-number="51"><span class="co">##  dqrng                  0.2.1    2019-05-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-52" data-line-number="52"><span class="co">##  edgeR                  3.30.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-53" data-line-number="53"><span class="co">##  ellipsis               0.3.0    2019-09-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-54" data-line-number="54"><span class="co">##  evaluate               0.14     2019-05-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-55" data-line-number="55"><span class="co">##  extrafont            * 0.17     2014-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-56" data-line-number="56"><span class="co">##  extrafontdb            1.0      2012-06-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-57" data-line-number="57"><span class="co">##  fansi                  0.4.1    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-58" data-line-number="58"><span class="co">##  farver                 2.0.3    2020-01-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-59" data-line-number="59"><span class="co">##  fitdistrplus           1.0-14   2019-01-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-60" data-line-number="60"><span class="co">##  forcats              * 0.5.0    2020-03-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-61" data-line-number="61"><span class="co">##  foreach                1.5.0    2020-03-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-62" data-line-number="62"><span class="co">##  fs                     1.4.1    2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-63" data-line-number="63"><span class="co">##  future                 1.17.0   2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-64" data-line-number="64"><span class="co">##  future.apply           1.5.0    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-65" data-line-number="65"><span class="co">##  generics               0.0.2    2018-11-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-66" data-line-number="66"><span class="co">##  GenomeInfoDb         * 1.24.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-67" data-line-number="67"><span class="co">##  GenomeInfoDbData       1.2.3    2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-68" data-line-number="68"><span class="co">##  GenomicRanges        * 1.40.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-69" data-line-number="69"><span class="co">##  ggbeeswarm             0.6.0    2017-08-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-70" data-line-number="70"><span class="co">##  ggplot2              * 3.3.0    2020-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-71" data-line-number="71"><span class="co">##  ggrepel                0.8.2    2020-03-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-72" data-line-number="72"><span class="co">##  ggridges               0.5.2    2020-01-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-73" data-line-number="73"><span class="co">##  globals                0.12.5   2019-12-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-74" data-line-number="74"><span class="co">##  glue                   1.4.0    2020-04-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-75" data-line-number="75"><span class="co">##  gridBase               0.4-7    2014-02-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-76" data-line-number="76"><span class="co">##  gridExtra            * 2.3      2017-09-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-77" data-line-number="77"><span class="co">##  gtable                 0.3.0    2019-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-78" data-line-number="78"><span class="co">##  gtools                 3.8.2    2020-03-31 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-79" data-line-number="79"><span class="co">##  haven                  2.2.0    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-80" data-line-number="80"><span class="co">##  highr                  0.8      2019-03-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-81" data-line-number="81"><span class="co">##  hms                    0.5.3    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-82" data-line-number="82"><span class="co">##  htmltools              0.4.0    2019-10-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-83" data-line-number="83"><span class="co">##  htmlwidgets            1.5.1    2019-10-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-84" data-line-number="84"><span class="co">##  httr                   1.4.1    2019-08-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-85" data-line-number="85"><span class="co">##  ica                    1.0-2    2018-05-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-86" data-line-number="86"><span class="co">##  igraph                 1.2.5    2020-03-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-87" data-line-number="87"><span class="co">##  IRanges              * 2.22.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-88" data-line-number="88"><span class="co">##  irlba                  2.3.3    2019-02-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-89" data-line-number="89"><span class="co">##  iterators              1.0.12   2019-07-26 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-90" data-line-number="90"><span class="co">##  jsonlite               1.6.1    2020-02-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-91" data-line-number="91"><span class="co">##  KernSmooth             2.23-16  2019-10-15 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-92" data-line-number="92"><span class="co">##  knitr                  1.28     2020-02-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-93" data-line-number="93"><span class="co">##  labeling               0.3      2014-08-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-94" data-line-number="94"><span class="co">##  lattice                0.20-41  2020-04-02 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-95" data-line-number="95"><span class="co">##  lazyeval               0.2.2    2019-03-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-96" data-line-number="96"><span class="co">##  leiden                 0.3.3    2020-02-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-97" data-line-number="97"><span class="co">##  lifecycle              0.2.0    2020-03-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-98" data-line-number="98"><span class="co">##  limma                  3.44.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-99" data-line-number="99"><span class="co">##  listenv                0.8.0    2019-12-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-100" data-line-number="100"><span class="co">##  lmtest                 0.9-37   2019-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-101" data-line-number="101"><span class="co">##  locfit                 1.5-9.4  2020-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-102" data-line-number="102"><span class="co">##  lsei                   1.2-0.1  2020-05-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-103" data-line-number="103"><span class="co">##  lubridate              1.7.8    2020-04-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-104" data-line-number="104"><span class="co">##  magrittr               1.5      2014-11-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-105" data-line-number="105"><span class="co">##  MASS                   7.3-51.5 2019-12-20 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-106" data-line-number="106"><span class="co">##  Matrix                 1.2-18   2019-11-27 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-107" data-line-number="107"><span class="co">##  matrixStats          * 0.56.0   2020-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-108" data-line-number="108"><span class="co">##  mclust               * 5.4.6    2020-04-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-109" data-line-number="109"><span class="co">##  memoise                1.1.0    2017-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-110" data-line-number="110"><span class="co">##  modelr                 0.1.7    2020-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-111" data-line-number="111"><span class="co">##  munsell                0.5.0    2018-06-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-112" data-line-number="112"><span class="co">##  nlme                   3.1-147  2020-04-13 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-113" data-line-number="113"><span class="co">##  NMF                  * 0.22.0   2020-02-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-114" data-line-number="114"><span class="co">##  npsurv                 0.4-0.1  2020-05-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-115" data-line-number="115"><span class="co">##  org.Dr.eg.db         * 3.11.1   2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-116" data-line-number="116"><span class="co">##  patchwork              1.0.0    2019-12-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-117" data-line-number="117"><span class="co">##  pbapply                1.4-2    2019-08-31 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-118" data-line-number="118"><span class="co">##  pheatmap             * 1.0.12   2019-01-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-119" data-line-number="119"><span class="co">##  pillar                 1.4.4    2020-05-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-120" data-line-number="120"><span class="co">##  pkgbuild               1.0.8    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-121" data-line-number="121"><span class="co">##  pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-122" data-line-number="122"><span class="co">##  pkgload                1.0.2    2018-10-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-123" data-line-number="123"><span class="co">##  pkgmaker             * 0.31.1   2020-03-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-124" data-line-number="124"><span class="co">##  plotly                 4.9.2.1  2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-125" data-line-number="125"><span class="co">##  plyr                   1.8.6    2020-03-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-126" data-line-number="126"><span class="co">##  png                    0.1-7    2013-12-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-127" data-line-number="127"><span class="co">##  prettyunits            1.1.1    2020-01-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-128" data-line-number="128"><span class="co">##  processx               3.4.2    2020-02-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-129" data-line-number="129"><span class="co">##  ps                     1.3.3    2020-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-130" data-line-number="130"><span class="co">##  purrr                * 0.3.4    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-131" data-line-number="131"><span class="co">##  R.methodsS3          * 1.8.0    2020-02-14 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-132" data-line-number="132"><span class="co">##  R.oo                 * 1.23.0   2019-11-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-133" data-line-number="133"><span class="co">##  R.utils              * 2.9.2    2019-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-134" data-line-number="134"><span class="co">##  R6                     2.4.1    2019-11-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-135" data-line-number="135"><span class="co">##  RANN                   2.6.1    2019-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-136" data-line-number="136"><span class="co">##  rappdirs               0.3.1    2016-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-137" data-line-number="137"><span class="co">##  RColorBrewer           1.1-2    2014-12-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-138" data-line-number="138"><span class="co">##  Rcpp                   1.0.4.6  2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-139" data-line-number="139"><span class="co">##  RcppAnnoy              0.0.16   2020-03-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-140" data-line-number="140"><span class="co">##  RCurl                  1.98-1.2 2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-141" data-line-number="141"><span class="co">##  readr                * 1.3.1    2018-12-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-142" data-line-number="142"><span class="co">##  readxl                 1.3.1    2019-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-143" data-line-number="143"><span class="co">##  registry             * 0.5-1    2019-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-144" data-line-number="144"><span class="co">##  remotes                2.1.1    2020-02-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-145" data-line-number="145"><span class="co">##  reprex                 0.3.0    2019-05-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-146" data-line-number="146"><span class="co">##  reshape2               1.4.4    2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-147" data-line-number="147"><span class="co">##  reticulate             1.15     2020-04-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-148" data-line-number="148"><span class="co">##  rlang                  0.4.6    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-149" data-line-number="149"><span class="co">##  rmarkdown              2.1      2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-150" data-line-number="150"><span class="co">##  rngtools             * 1.5      2020-01-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-151" data-line-number="151"><span class="co">##  ROCR                   1.0-11   2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-152" data-line-number="152"><span class="co">##  rprojroot              1.3-2    2018-01-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-153" data-line-number="153"><span class="co">##  RSpectra               0.16-0   2019-12-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-154" data-line-number="154"><span class="co">##  RSQLite                2.2.0    2020-01-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-155" data-line-number="155"><span class="co">##  rstudioapi             0.11     2020-02-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-156" data-line-number="156"><span class="co">##  rsvd                   1.0.3    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-157" data-line-number="157"><span class="co">##  Rtsne                  0.15     2018-11-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-158" data-line-number="158"><span class="co">##  Rttf2pt1               1.3.8    2020-01-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-159" data-line-number="159"><span class="co">##  rvest                  0.3.5    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-160" data-line-number="160"><span class="co">##  S4Vectors            * 0.26.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-161" data-line-number="161"><span class="co">##  scales                 1.1.1    2020-05-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-162" data-line-number="162"><span class="co">##  scater               * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-163" data-line-number="163"><span class="co">##  scran                * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-164" data-line-number="164"><span class="co">##  sctransform            0.2.1    2019-12-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-165" data-line-number="165"><span class="co">##  sessioninfo            1.1.1    2018-11-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-166" data-line-number="166"><span class="co">##  Seurat               * 3.1.5    2020-04-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-167" data-line-number="167"><span class="co">##  SingleCellExperiment * 1.10.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-168" data-line-number="168"><span class="co">##  statmod                1.4.34   2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-169" data-line-number="169"><span class="co">##  stringi                1.4.6    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-170" data-line-number="170"><span class="co">##  stringr              * 1.4.0    2019-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-171" data-line-number="171"><span class="co">##  SummarizedExperiment * 1.18.1   2020-04-30 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-172" data-line-number="172"><span class="co">##  survival               3.1-12   2020-04-10 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-173" data-line-number="173"><span class="co">##  testthat               2.3.2    2020-03-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-174" data-line-number="174"><span class="co">##  tibble               * 3.0.1    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-175" data-line-number="175"><span class="co">##  tidyr                * 1.0.3    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-176" data-line-number="176"><span class="co">##  tidyselect             1.0.0    2020-01-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-177" data-line-number="177"><span class="co">##  tidyverse            * 1.3.0    2019-11-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-178" data-line-number="178"><span class="co">##  tsne                   0.1-3    2016-07-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-179" data-line-number="179"><span class="co">##  usethis                1.6.1    2020-04-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-180" data-line-number="180"><span class="co">##  utf8                   1.1.4    2018-05-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-181" data-line-number="181"><span class="co">##  uwot                   0.1.8    2020-03-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-182" data-line-number="182"><span class="co">##  vctrs                  0.2.4    2020-03-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-183" data-line-number="183"><span class="co">##  vipor                  0.4.5    2017-03-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-184" data-line-number="184"><span class="co">##  viridis              * 0.5.1    2018-03-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-185" data-line-number="185"><span class="co">##  viridisLite          * 0.3.0    2018-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-186" data-line-number="186"><span class="co">##  withr                  2.2.0    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-187" data-line-number="187"><span class="co">##  xfun                   0.13     2020-04-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-188" data-line-number="188"><span class="co">##  xml2                   1.3.2    2020-04-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-189" data-line-number="189"><span class="co">##  xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-190" data-line-number="190"><span class="co">##  XVector                0.28.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-191" data-line-number="191"><span class="co">##  yaml                   2.2.1    2020-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-192" data-line-number="192"><span class="co">##  zlibbioc               1.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb62-193" data-line-number="193"><span class="co">##  zoo                    1.8-8    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb62-194" data-line-number="194"><span class="co">## </span></a>
<a class="sourceLine" id="cb62-195" data-line-number="195"><span class="co">## [1] D:/R_packages</span></a>
<a class="sourceLine" id="cb62-196" data-line-number="196"><span class="co">## [2] C:/Program Files/R/R-4.0.0/library</span></a></code></pre></div>
</details>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="clustering.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="differentially-expressed-genes.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["NMps_book.pdf", "NMps_book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
