<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 8 Pseudotime | NMps_book.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 8 Pseudotime | NMps_book.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 8 Pseudotime | NMps_book.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="subclustering.html"/>

<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="downloading-data.html"><a href="downloading-data.html"><i class="fa fa-check"></i><b>2</b> Downloading data</a></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a><ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#importing-libraries"><i class="fa fa-check"></i><b>3.2</b> Importing libraries</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#importing-the-data"><i class="fa fa-check"></i><b>3.3</b> Importing the data</a></li>
<li class="chapter" data-level="3.4" data-path="quality-control.html"><a href="quality-control.html#preparing-the-column-metadata"><i class="fa fa-check"></i><b>3.4</b> Preparing the column metadata</a></li>
<li class="chapter" data-level="3.5" data-path="quality-control.html"><a href="quality-control.html#preparing-the-row-metadata"><i class="fa fa-check"></i><b>3.5</b> Preparing the row metadata</a></li>
<li class="chapter" data-level="3.6" data-path="quality-control.html"><a href="quality-control.html#constructing-the-single-cell-experiment-object"><i class="fa fa-check"></i><b>3.6</b> Constructing the single cell experiment object</a></li>
<li class="chapter" data-level="3.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-with-scater"><i class="fa fa-check"></i><b>3.7</b> Quality control with scater</a></li>
<li class="chapter" data-level="3.8" data-path="quality-control.html"><a href="quality-control.html#saving-the-rds-object"><i class="fa fa-check"></i><b>3.8</b> Saving the RDS object</a></li>
<li class="chapter" data-level="3.9" data-path="quality-control.html"><a href="quality-control.html#session-info"><i class="fa fa-check"></i><b>3.9</b> Session Info</a></li>
<li class="chapter" data-level="3.10" data-path="quality-control.html"><a href="quality-control.html#solutions"><i class="fa fa-check"></i><b>3.10</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>4</b> Clustering</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="clustering.html"><a href="clustering.html#importing-libraries-1"><i class="fa fa-check"></i><b>4.2</b> Importing libraries</a></li>
<li class="chapter" data-level="4.3" data-path="clustering.html"><a href="clustering.html#importing-data"><i class="fa fa-check"></i><b>4.3</b> Importing Data</a></li>
<li class="chapter" data-level="4.4" data-path="clustering.html"><a href="clustering.html#useful-seurat-object-interaction-methods"><i class="fa fa-check"></i><b>4.4</b> Useful seurat object interaction methods</a></li>
<li class="chapter" data-level="4.5" data-path="clustering.html"><a href="clustering.html#performing-dimensional-reduction"><i class="fa fa-check"></i><b>4.5</b> Performing dimensional reduction</a></li>
<li class="chapter" data-level="4.6" data-path="clustering.html"><a href="clustering.html#clustering-1"><i class="fa fa-check"></i><b>4.6</b> Clustering</a><ul>
<li class="chapter" data-level="4.6.1" data-path="clustering.html"><a href="clustering.html#constructing-a-shared-nearest-neighbour-snn-graph"><i class="fa fa-check"></i><b>4.6.1</b> Constructing a Shared-nearest neighbour (SNN) graph</a></li>
<li class="chapter" data-level="4.6.2" data-path="clustering.html"><a href="clustering.html#clustering-with-the-louvain-algorithm"><i class="fa fa-check"></i><b>4.6.2</b> Clustering with the Louvain algorithm</a></li>
<li class="chapter" data-level="4.6.3" data-path="clustering.html"><a href="clustering.html#resources"><i class="fa fa-check"></i><b>4.6.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="clustering.html"><a href="clustering.html#umap-with-authors-annotations"><i class="fa fa-check"></i><b>4.7</b> UMAP with authors’ annotations</a></li>
<li class="chapter" data-level="4.8" data-path="clustering.html"><a href="clustering.html#t-sne-with-authors-annotations"><i class="fa fa-check"></i><b>4.8</b> t-SNE with authors’ annotations</a></li>
<li class="chapter" data-level="4.9" data-path="clustering.html"><a href="clustering.html#assigning-our-own-cluster-identities"><i class="fa fa-check"></i><b>4.9</b> Assigning our own cluster identities</a></li>
<li class="chapter" data-level="4.10" data-path="clustering.html"><a href="clustering.html#umap-visualisation-of-our-annotations"><i class="fa fa-check"></i><b>4.10</b> UMAP visualisation of our annotations</a></li>
<li class="chapter" data-level="4.11" data-path="clustering.html"><a href="clustering.html#highlighting-tailbud-cells"><i class="fa fa-check"></i><b>4.11</b> Highlighting tailbud cells</a></li>
<li class="chapter" data-level="4.12" data-path="clustering.html"><a href="clustering.html#saving-rds-object"><i class="fa fa-check"></i><b>4.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="4.13" data-path="clustering.html"><a href="clustering.html#session-info-1"><i class="fa fa-check"></i><b>4.13</b> Session Info</a></li>
<li class="chapter" data-level="4.14" data-path="clustering.html"><a href="clustering.html#solutions-1"><i class="fa fa-check"></i><b>4.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="optimisations.html"><a href="optimisations.html"><i class="fa fa-check"></i><b>5</b> Optimisations</a><ul>
<li class="chapter" data-level="5.1" data-path="optimisations.html"><a href="optimisations.html#importing-libraries-2"><i class="fa fa-check"></i><b>5.1</b> Importing libraries</a></li>
<li class="chapter" data-level="5.2" data-path="optimisations.html"><a href="optimisations.html#importing-data-1"><i class="fa fa-check"></i><b>5.2</b> Importing Data</a></li>
<li class="chapter" data-level="5.3" data-path="optimisations.html"><a href="optimisations.html#number-of-pcs-to-retain"><i class="fa fa-check"></i><b>5.3</b> Number of PCs to retain</a><ul>
<li class="chapter" data-level="5.3.1" data-path="optimisations.html"><a href="optimisations.html#scree-plot"><i class="fa fa-check"></i><b>5.3.1</b> Scree plot</a></li>
<li class="chapter" data-level="5.3.2" data-path="optimisations.html"><a href="optimisations.html#examining-technical-noise"><i class="fa fa-check"></i><b>5.3.2</b> Examining technical noise</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="optimisations.html"><a href="optimisations.html#evaluating-clustering-results"><i class="fa fa-check"></i><b>5.4</b> Evaluating clustering results</a><ul>
<li class="chapter" data-level="5.4.1" data-path="optimisations.html"><a href="optimisations.html#cluster-stability"><i class="fa fa-check"></i><b>5.4.1</b> Cluster stability</a></li>
<li class="chapter" data-level="5.4.2" data-path="optimisations.html"><a href="optimisations.html#cluster-separability"><i class="fa fa-check"></i><b>5.4.2</b> Cluster separability</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="optimisations.html"><a href="optimisations.html#comparison-with-authors-labels"><i class="fa fa-check"></i><b>5.5</b> Comparison with authors’ labels</a><ul>
<li class="chapter" data-level="5.5.1" data-path="optimisations.html"><a href="optimisations.html#concordance-in-tailbud-cluster-assignment"><i class="fa fa-check"></i><b>5.5.1</b> Concordance in tailbud cluster assignment</a></li>
<li class="chapter" data-level="5.5.2" data-path="optimisations.html"><a href="optimisations.html#optimising-the-seurat-clustering-parameters"><i class="fa fa-check"></i><b>5.5.2</b> Optimising the seurat clustering parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="optimisations.html"><a href="optimisations.html#selecting-cells-for-downstream-analysis"><i class="fa fa-check"></i><b>5.6</b> Selecting cells for downstream analysis</a></li>
<li class="chapter" data-level="5.7" data-path="optimisations.html"><a href="optimisations.html#session-info-2"><i class="fa fa-check"></i><b>5.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html"><i class="fa fa-check"></i><b>6</b> Differentially Expressed Genes</a><ul>
<li class="chapter" data-level="6.1" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualisation-of-marker-genes-in-our-dataset"><i class="fa fa-check"></i><b>6.1</b> Visualisation of marker genes in our dataset</a></li>
<li class="chapter" data-level="6.2" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#obtaining-a-list-of-differentially-expressed-genes"><i class="fa fa-check"></i><b>6.2</b> Obtaining a list of differentially expressed genes</a></li>
<li class="chapter" data-level="6.3" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#analysing-output-of-the-wilcoxon-test"><i class="fa fa-check"></i><b>6.3</b> Analysing output of the wilcoxon test</a></li>
<li class="chapter" data-level="6.4" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualising-the-expression-of-the-marker-genes"><i class="fa fa-check"></i><b>6.4</b> Visualising the expression of the marker genes</a></li>
<li class="chapter" data-level="6.5" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#session-info-3"><i class="fa fa-check"></i><b>6.5</b> Session Info</a></li>
<li class="chapter" data-level="6.6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#solutions-2"><i class="fa fa-check"></i><b>6.6</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="subclustering.html"><a href="subclustering.html"><i class="fa fa-check"></i><b>7</b> Subclustering</a><ul>
<li class="chapter" data-level="7.1" data-path="subclustering.html"><a href="subclustering.html#importing-libraries-3"><i class="fa fa-check"></i><b>7.1</b> Importing libraries</a></li>
<li class="chapter" data-level="7.2" data-path="subclustering.html"><a href="subclustering.html#importing-data-2"><i class="fa fa-check"></i><b>7.2</b> Importing Data</a></li>
<li class="chapter" data-level="7.3" data-path="subclustering.html"><a href="subclustering.html#subclustering-the-tb-cells---using-only-cells-that-are-assigned-to-the-shared-cluster-across-all-the-clustering-resolutions."><i class="fa fa-check"></i><b>7.3</b> Subclustering the TB cells - Using only cells that are assigned to the ‘shared’ cluster across all the clustering resolutions.</a></li>
<li class="chapter" data-level="7.4" data-path="subclustering.html"><a href="subclustering.html#dimensional-reduction-with-umap-and-clustering"><i class="fa fa-check"></i><b>7.4</b> Dimensional reduction with UMAP and clustering</a></li>
<li class="chapter" data-level="7.5" data-path="subclustering.html"><a href="subclustering.html#identifying-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.5</b> Identifying marker genes for each cluster</a></li>
<li class="chapter" data-level="7.6" data-path="subclustering.html"><a href="subclustering.html#merging-clusters-that-do-not-appear-distinct"><i class="fa fa-check"></i><b>7.6</b> Merging clusters that do not appear distinct</a></li>
<li class="chapter" data-level="7.7" data-path="subclustering.html"><a href="subclustering.html#plotting-umap-and-assigning-cluster-labels"><i class="fa fa-check"></i><b>7.7</b> Plotting UMAP and assigning cluster labels</a></li>
<li class="chapter" data-level="7.8" data-path="subclustering.html"><a href="subclustering.html#checking-the-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.8</b> Checking the marker genes for each cluster</a></li>
<li class="chapter" data-level="7.9" data-path="subclustering.html"><a href="subclustering.html#sox2-tbxta-coexpression"><i class="fa fa-check"></i><b>7.9</b> Sox2 tbxta coexpression</a></li>
<li class="chapter" data-level="7.10" data-path="subclustering.html"><a href="subclustering.html#viewing-genes-of-interest-with-a-dotplot"><i class="fa fa-check"></i><b>7.10</b> Viewing genes of interest with a dotplot</a></li>
<li class="chapter" data-level="7.11" data-path="subclustering.html"><a href="subclustering.html#exporting-the-violin-plots-for-genes-of-interest."><i class="fa fa-check"></i><b>7.11</b> Exporting the violin plots for genes of interest.</a></li>
<li class="chapter" data-level="7.12" data-path="subclustering.html"><a href="subclustering.html#saving-rds-object-1"><i class="fa fa-check"></i><b>7.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="7.13" data-path="subclustering.html"><a href="subclustering.html#session-info-4"><i class="fa fa-check"></i><b>7.13</b> Session Info</a></li>
<li class="chapter" data-level="7.14" data-path="subclustering.html"><a href="subclustering.html#solutions-3"><i class="fa fa-check"></i><b>7.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="pseudotime.html"><a href="pseudotime.html"><i class="fa fa-check"></i><b>8</b> Pseudotime</a><ul>
<li class="chapter" data-level="8.1" data-path="pseudotime.html"><a href="pseudotime.html#importing-libraries-4"><i class="fa fa-check"></i><b>8.1</b> Importing libraries</a></li>
<li class="chapter" data-level="8.2" data-path="pseudotime.html"><a href="pseudotime.html#importing-data-3"><i class="fa fa-check"></i><b>8.2</b> Importing data</a></li>
<li class="chapter" data-level="8.3" data-path="pseudotime.html"><a href="pseudotime.html#tradeseq-workflow-analyses"><i class="fa fa-check"></i><b>8.3</b> TradeSeq workflow analyses</a></li>
<li class="chapter" data-level="8.4" data-path="pseudotime.html"><a href="pseudotime.html#session-info-5"><i class="fa fa-check"></i><b>8.4</b> Session Info</a></li>
<li class="chapter" data-level="8.5" data-path="pseudotime.html"><a href="pseudotime.html#solutions-4"><i class="fa fa-check"></i><b>8.5</b> Solutions</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="pseudotime" class="section level1">
<h1><span class="header-section-number">Chapter 8</span> Pseudotime</h1>
<p>Our aim is to explore how the expression of marker genes change along the neural and mesodermal trajectories, with the neuromesodermal progenitors acting as the root of the bifurcation.</p>
<p>We will perform <code>pseudotime inference</code> with the slingshot package to assign each cell a measure of biological progress along each trajectory. More precisely, for two cells A and B ordered along a trajectory, if cell B has a larger pseudotime value than cell A, then cell B is interpreted as exhibiting a gene expression profile of a more differentiated cell relative to cell A.</p>
<p>Subsequently, we use the <code>tradeseq</code> package to analyse the change in gene expression along the neural and mesodermal lineage trajectories.</p>
<div id="importing-libraries-4" class="section level2">
<h2><span class="header-section-number">8.1</span> Importing libraries</h2>
<div class="sourceCode" id="cb141"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb141-1" data-line-number="1"><span class="kw">library</span>(SingleCellExperiment)</a>
<a class="sourceLine" id="cb141-2" data-line-number="2"><span class="kw">library</span>(slingshot)</a>
<a class="sourceLine" id="cb141-3" data-line-number="3"><span class="kw">library</span>(RColorBrewer)</a>
<a class="sourceLine" id="cb141-4" data-line-number="4"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb141-5" data-line-number="5"><span class="kw">library</span>(Seurat)</a>
<a class="sourceLine" id="cb141-6" data-line-number="6"><span class="kw">library</span>(tradeSeq)</a>
<a class="sourceLine" id="cb141-7" data-line-number="7"><span class="kw">library</span>(clusterExperiment)</a>
<a class="sourceLine" id="cb141-8" data-line-number="8"><span class="kw">library</span>(cowplot)</a></code></pre></div>
</div>
<div id="importing-data-3" class="section level2">
<h2><span class="header-section-number">8.2</span> Importing data</h2>
<div class="sourceCode" id="cb142"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb142-1" data-line-number="1">TBcells &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;rds/TBcells_subclustered.rds&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb143"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb143-1" data-line-number="1"><span class="co">#Convert to SCE object</span></a>
<a class="sourceLine" id="cb143-2" data-line-number="2">TB_sce &lt;-<span class="st"> </span><span class="kw">as.SingleCellExperiment</span>(<span class="dt">x =</span> TBcells,</a>
<a class="sourceLine" id="cb143-3" data-line-number="3">                                  <span class="dt">assay =</span> <span class="st">&quot;SCT&quot;</span>)</a>
<a class="sourceLine" id="cb143-4" data-line-number="4"></a>
<a class="sourceLine" id="cb143-5" data-line-number="5"><span class="kw">table</span>(<span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names)</a></code></pre></div>
<pre><code>## 
##      Maturation zone      Nascent somites                 NMps 
##                   62                   76                   47 
##         Posterior NT        Posterior PSM Pou5f3+ Posterior NT 
##                  103                  140                   73 
##         Tail somites 
##                   50</code></pre>
<p><em>Q1 How would you subset the data to contain only cells in the NMp, maturation zone, posterior PSM and pou5f3+ posterior NT clusters?</em></p>
<p>To run slingshot on our dataset, we need to provide it with the coordinates of the reduced dimension and the cluster labels. For comparison with our previous work, we will use the <code>UMAP</code> coordinates and the <code>subcluster_names</code> labels, which are both stored in the converted <code>SingleCellExperiment</code> object.</p>
<p>The <code>slingshot</code> wrapper function performs 2 necessary steps in this method of lineage inference. First, it constructs a cluster-based minimum spanning tree (MST) on the cells to obtain a global lineage structure using the <code>getLineages</code> function. This is followed by the fitting of principle curves for each lineage called with the <code>getCurves</code> function.</p>
<div class="sourceCode" id="cb145"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb145-1" data-line-number="1">sce_sling &lt;-<span class="st"> </span><span class="kw">slingshot</span>(TB_sce, <span class="dt">clusterLabels =</span> <span class="st">&quot;subclusters_names&quot;</span>, <span class="dt">reducedDim =</span> <span class="st">&quot;UMAP&quot;</span>, <span class="dt">start.clus =</span> <span class="st">&quot;NMps&quot;</span> )</a></code></pre></div>
<pre><code>## Using full covariance matrix</code></pre>
<div class="sourceCode" id="cb147"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb147-1" data-line-number="1">clusters &lt;-<span class="st"> </span><span class="kw">ifelse</span>(<span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names <span class="op">==</span><span class="st"> &quot;NMps&quot;</span>, <span class="dv">1</span>, </a>
<a class="sourceLine" id="cb147-2" data-line-number="2">                    <span class="kw">ifelse</span>(<span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names <span class="op">==</span><span class="st"> &quot;Maturation zone&quot;</span>,<span class="dv">2</span>,                  <span class="kw">ifelse</span>(<span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names <span class="op">==</span><span class="st"> &quot;Posterior PSM&quot;</span>, <span class="dv">3</span>,</a>
<a class="sourceLine" id="cb147-3" data-line-number="3">                    <span class="kw">ifelse</span>(<span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names <span class="op">==</span><span class="st"> &quot;Pou5f3+ Posterior NT&quot;</span>, <span class="dv">4</span>,<span class="kw">ifelse</span>(<span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names <span class="op">==</span><span class="st"> &quot;Posterior NT&quot;</span>, <span class="dv">5</span>,<span class="kw">ifelse</span>(<span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names <span class="op">==</span><span class="st"> &quot;Nascent somites&quot;</span>, <span class="dv">6</span>,<span class="kw">ifelse</span>(<span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names <span class="op">==</span><span class="st"> &quot;Tail somites&quot;</span>, <span class="dv">7</span>, <span class="ot">NA</span>)))))))</a>
<a class="sourceLine" id="cb147-4" data-line-number="4"></a>
<a class="sourceLine" id="cb147-5" data-line-number="5">clust_col =<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">names =</span> <span class="kw">colData</span>(TB_sce)<span class="op">$</span>subclusters_names, <span class="dt">clusters =</span> clusters, <span class="dt">colors =</span> <span class="kw">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Set1&quot;</span>)[clusters]) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">arrange</span>(clusters) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">distinct</span>()</a>
<a class="sourceLine" id="cb147-6" data-line-number="6"></a>
<a class="sourceLine" id="cb147-7" data-line-number="7"><span class="kw">plot</span>(<span class="kw">reducedDims</span>(sce_sling)<span class="op">$</span>UMAP, <span class="dt">col =</span> <span class="kw">brewer.pal</span>(<span class="dv">9</span>,<span class="st">&quot;Set1&quot;</span>)[clusters], <span class="dt">pch=</span><span class="dv">16</span>, <span class="dt">asp =</span> <span class="dv">1</span>)</a>
<a class="sourceLine" id="cb147-8" data-line-number="8"></a>
<a class="sourceLine" id="cb147-9" data-line-number="9"><span class="kw">lines</span>(<span class="kw">SlingshotDataSet</span>(sce_sling), <span class="dt">lwd=</span><span class="dv">1</span>, <span class="dt">type =</span> <span class="st">&quot;lineages&quot;</span>, <span class="dt">col=</span><span class="st">&#39;black&#39;</span>)</a>
<a class="sourceLine" id="cb147-10" data-line-number="10"></a>
<a class="sourceLine" id="cb147-11" data-line-number="11"><span class="kw">legend</span>(<span class="op">-</span><span class="dv">11</span>, <span class="fl">-0.7</span>, <span class="dt">legend=</span><span class="kw">c</span>(<span class="st">&quot;NMps&quot;</span>, <span class="st">&quot;Maturation Zone&quot;</span>, <span class="st">&quot;Posterior PSM&quot;</span>, <span class="st">&quot;Pou5f3+ posterior NT&quot;</span>, <span class="st">&quot;Posterior NT&quot;</span>, <span class="st">&quot;Nascent Somites&quot;</span>, <span class="st">&quot;Tail Somites&quot;</span>), <span class="dt">col=</span>clust_col<span class="op">$</span>colors, <span class="dt">pch =</span> <span class="dv">16</span>, <span class="dt">pt.cex =</span> <span class="fl">0.8</span>, <span class="dt">cex =</span> <span class="fl">0.8</span>, <span class="dt">bty=</span><span class="st">&#39;n&#39;</span>, <span class="dt">box.lty=</span><span class="dv">0</span>)</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/plot_slingshot-1.png" width="1056" /></p>
<div class="sourceCode" id="cb148"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb148-1" data-line-number="1">TBcells &lt;-<span class="st"> </span><span class="kw">FindVariableFeatures</span>(TBcells)</a>
<a class="sourceLine" id="cb148-2" data-line-number="2">sce_sling &lt;-<span class="st"> </span>sce_sling[<span class="kw">VariableFeatures</span>(TBcells),]</a>
<a class="sourceLine" id="cb148-3" data-line-number="3">counts &lt;-<span class="st"> </span><span class="kw">assays</span>(sce_sling)<span class="op">$</span>counts <span class="op">%&gt;%</span><span class="st"> </span>as.matrix</a>
<a class="sourceLine" id="cb148-4" data-line-number="4">logcounts &lt;-<span class="st"> </span><span class="kw">assays</span>(sce_sling)<span class="op">$</span>logcounts <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">as.matrix</span>()</a></code></pre></div>
</div>
<div id="tradeseq-workflow-analyses" class="section level2">
<h2><span class="header-section-number">8.3</span> TradeSeq workflow analyses</h2>
<p><a href="https://statomics.github.io/tradeSeq/articles/tradeSeq.html">Tradeseq vignette</a></p>
<div class="sourceCode" id="cb149"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb149-1" data-line-number="1"><span class="kw">set.seed</span>(<span class="dv">5</span>)</a>
<a class="sourceLine" id="cb149-2" data-line-number="2">icMat &lt;-<span class="st"> </span><span class="kw">evaluateK</span>(<span class="dt">counts =</span> counts, <span class="dt">sds =</span> <span class="kw">SlingshotDataSet</span>(sce_sling), <span class="dt">k =</span> <span class="dv">3</span><span class="op">:</span><span class="dv">10</span>,</a>
<a class="sourceLine" id="cb149-3" data-line-number="3">                   <span class="dt">nGenes =</span> <span class="dv">200</span>, <span class="dt">verbose =</span> T)</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/evaluateK-1.png" width="672" /></p>
<div class="sourceCode" id="cb150"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb150-1" data-line-number="1"><span class="kw">dim</span>(counts)</a></code></pre></div>
<pre><code>## [1] 2000  551</code></pre>
<div class="sourceCode" id="cb152"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb152-1" data-line-number="1"><span class="kw">dim</span>((sce_sling))</a></code></pre></div>
<pre><code>## [1] 2000  551</code></pre>
<div class="sourceCode" id="cb154"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb154-1" data-line-number="1">sce &lt;-<span class="st"> </span><span class="kw">fitGAM</span>(<span class="dt">counts =</span> counts,</a>
<a class="sourceLine" id="cb154-2" data-line-number="2">                  <span class="dt">sds =</span> <span class="kw">SlingshotDataSet</span>(sce_sling),</a>
<a class="sourceLine" id="cb154-3" data-line-number="3">              <span class="dt">nknots =</span> <span class="dv">10</span>)</a></code></pre></div>
<div class="sourceCode" id="cb155"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb155-1" data-line-number="1"><span class="kw">saveRDS</span>(sce, <span class="dt">file =</span> <span class="st">&quot;rds/6_slingshot.rds&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb156"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb156-1" data-line-number="1">sce &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="st">&quot;rds/6_slingshot.rds&quot;</span>)</a>
<a class="sourceLine" id="cb156-2" data-line-number="2">assoRes &lt;-<span class="st"> </span><span class="kw">associationTest</span>(sce) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb156-3" data-line-number="3"><span class="st">  </span>tibble<span class="op">::</span><span class="kw">rownames_to_column</span>() <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb156-4" data-line-number="4"><span class="st">  </span>dplyr<span class="op">::</span><span class="kw">filter</span>(pvalue <span class="op">&lt;</span><span class="st"> </span><span class="fl">1e-10</span>)</a>
<a class="sourceLine" id="cb156-5" data-line-number="5"></a>
<a class="sourceLine" id="cb156-6" data-line-number="6">asso_genes &lt;-<span class="st"> </span>assoRes<span class="op">$</span>rowname </a></code></pre></div>
<div class="sourceCode" id="cb157"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb157-1" data-line-number="1"><span class="co"># Clustering expression patterns</span></a>
<a class="sourceLine" id="cb157-2" data-line-number="2"></a>
<a class="sourceLine" id="cb157-3" data-line-number="3">nPointsClus &lt;-<span class="st"> </span><span class="dv">20</span></a>
<a class="sourceLine" id="cb157-4" data-line-number="4"><span class="co">#clusterExperiment::listBuiltInFunctions()</span></a>
<a class="sourceLine" id="cb157-5" data-line-number="5"></a>
<a class="sourceLine" id="cb157-6" data-line-number="6">clustPat &lt;-<span class="st"> </span><span class="kw">clusterExpressionPatterns</span>(sce, <span class="dt">nPoints =</span> nPointsClus,</a>
<a class="sourceLine" id="cb157-7" data-line-number="7">                                     <span class="dt">genes =</span> asso_genes)</a></code></pre></div>
<pre><code>## 36 parameter combinations, 36 use sequential method, 36 use subsampling method
## Running Clustering on Parameter Combinations...
## done.</code></pre>
<div class="sourceCode" id="cb159"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb159-1" data-line-number="1"><span class="kw">saveRDS</span>(clustPat, <span class="dt">file =</span> <span class="st">&quot;rds/6_clusterExperiment.rds&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb160"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb160-1" data-line-number="1">clustPat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;rds/6_clusterExperiment.rds&quot;</span>)</a>
<a class="sourceLine" id="cb160-2" data-line-number="2"></a>
<a class="sourceLine" id="cb160-3" data-line-number="3">clusterLabels &lt;-<span class="st"> </span><span class="kw">primaryCluster</span>(clustPat<span class="op">$</span>rsec)</a>
<a class="sourceLine" id="cb160-4" data-line-number="4"><span class="co">#clustPat</span></a>
<a class="sourceLine" id="cb160-5" data-line-number="5">genesWithinClust &lt;-<span class="st"> </span><span class="kw">rownames</span>(clustPat<span class="op">$</span>yhatScaled[clusterLabels <span class="op">==</span><span class="st"> </span><span class="dv">2</span>, ])</a>
<a class="sourceLine" id="cb160-6" data-line-number="6">(genesWithinClust)[<span class="dv">5</span>]</a></code></pre></div>
<pre><code>## [1] &quot;tbx6&quot;</code></pre>
<div class="sourceCode" id="cb162"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb162-1" data-line-number="1">cUniq &lt;-<span class="st"> </span><span class="kw">unique</span>(clusterLabels)</a>
<a class="sourceLine" id="cb162-2" data-line-number="2">cUniq &lt;-<span class="st"> </span>cUniq[<span class="op">!</span>cUniq <span class="op">==</span><span class="st"> </span><span class="dv">-1</span>] <span class="co"># remove unclustered genes</span></a>
<a class="sourceLine" id="cb162-3" data-line-number="3">cUniq &lt;-<span class="st"> </span><span class="kw">sort</span>(cUniq)</a>
<a class="sourceLine" id="cb162-4" data-line-number="4"></a>
<a class="sourceLine" id="cb162-5" data-line-number="5">plots &lt;-<span class="st"> </span><span class="kw">list</span>()</a>
<a class="sourceLine" id="cb162-6" data-line-number="6"><span class="cf">for</span> (xx <span class="cf">in</span> cUniq[<span class="dv">1</span><span class="op">:</span><span class="dv">10</span>]) {</a>
<a class="sourceLine" id="cb162-7" data-line-number="7">  cId &lt;-<span class="st"> </span><span class="kw">which</span>(clusterLabels <span class="op">==</span><span class="st"> </span>xx)</a>
<a class="sourceLine" id="cb162-8" data-line-number="8">  p &lt;-<span class="st"> </span><span class="kw">ggplot</span>(<span class="dt">data =</span> <span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="dv">1</span><span class="op">:</span>nPointsClus,</a>
<a class="sourceLine" id="cb162-9" data-line-number="9">                                <span class="dt">y =</span> <span class="kw">rep</span>(<span class="kw">range</span>(clustPat<span class="op">$</span>yhatScaled[cId, ]),</a>
<a class="sourceLine" id="cb162-10" data-line-number="10">                                        nPointsClus <span class="op">/</span><span class="st"> </span><span class="dv">2</span>)),</a>
<a class="sourceLine" id="cb162-11" data-line-number="11">              <span class="kw">aes</span>(<span class="dt">x =</span> x, <span class="dt">y =</span> y)) <span class="op">+</span></a>
<a class="sourceLine" id="cb162-12" data-line-number="12"><span class="st">    </span><span class="kw">geom_point</span>(<span class="dt">alpha =</span> <span class="dv">0</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb162-13" data-line-number="13"><span class="st">    </span><span class="kw">labs</span>(<span class="dt">title =</span> <span class="kw">paste0</span>(<span class="st">&quot;Cluster &quot;</span>, xx),  <span class="dt">x =</span> <span class="st">&quot;Pseudotime&quot;</span>, <span class="dt">y =</span> <span class="st">&quot;Normalized expression&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb162-14" data-line-number="14"><span class="st">    </span><span class="kw">theme_classic</span>()</a>
<a class="sourceLine" id="cb162-15" data-line-number="15">  <span class="cf">for</span> (ii <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(cId)) {</a>
<a class="sourceLine" id="cb162-16" data-line-number="16">    geneId &lt;-<span class="st"> </span><span class="kw">rownames</span>(clustPat<span class="op">$</span>yhatScaled)[cId[ii]]</a>
<a class="sourceLine" id="cb162-17" data-line-number="17">    p &lt;-<span class="st"> </span>p <span class="op">+</span></a>
<a class="sourceLine" id="cb162-18" data-line-number="18"><span class="st">      </span><span class="kw">geom_line</span>(<span class="dt">data =</span> <span class="kw">data.frame</span>(<span class="dt">x =</span> <span class="kw">rep</span>(<span class="dv">1</span><span class="op">:</span>nPointsClus, <span class="dv">2</span>),</a>
<a class="sourceLine" id="cb162-19" data-line-number="19">                                  <span class="dt">y =</span> clustPat<span class="op">$</span>yhatScaled[geneId, ],</a>
<a class="sourceLine" id="cb162-20" data-line-number="20">                                  <span class="dt">lineage =</span> <span class="kw">rep</span>(<span class="dv">0</span><span class="op">:</span><span class="dv">1</span>, <span class="dt">each =</span> nPointsClus)),</a>
<a class="sourceLine" id="cb162-21" data-line-number="21">                <span class="kw">aes</span>(<span class="dt">col =</span> <span class="kw">as.character</span>(lineage), <span class="dt">group =</span> lineage), <span class="dt">lwd =</span> <span class="fl">1.5</span>)</a>
<a class="sourceLine" id="cb162-22" data-line-number="22">  }</a>
<a class="sourceLine" id="cb162-23" data-line-number="23">  p &lt;-<span class="st"> </span>p <span class="op">+</span><span class="st"> </span><span class="kw">guides</span>(<span class="dt">color =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb162-24" data-line-number="24"><span class="st">    </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> <span class="kw">c</span>(<span class="st">&quot;orange&quot;</span>, <span class="st">&quot;darkseagreen3&quot;</span>),</a>
<a class="sourceLine" id="cb162-25" data-line-number="25">                       <span class="dt">breaks =</span> <span class="kw">c</span>(<span class="st">&quot;0&quot;</span>, <span class="st">&quot;1&quot;</span>))  </a>
<a class="sourceLine" id="cb162-26" data-line-number="26">  plots[[<span class="kw">as.character</span>(xx)]] &lt;-<span class="st"> </span>p</a>
<a class="sourceLine" id="cb162-27" data-line-number="27">}</a>
<a class="sourceLine" id="cb162-28" data-line-number="28">plots<span class="op">$</span>ncol &lt;-<span class="st"> </span><span class="dv">3</span></a>
<a class="sourceLine" id="cb162-29" data-line-number="29"></a>
<a class="sourceLine" id="cb162-30" data-line-number="30"><span class="kw">do.call</span>(cowplot<span class="op">::</span>plot_grid, plots)</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/clusterExperiment_df%20-1.png" width="1152" /></p>
</div>
<div id="session-info-5" class="section level2">
<h2><span class="header-section-number">8.4</span> Session Info</h2>
<details>
<p><summary>View Session Info</summary></p>
<div class="sourceCode" id="cb163"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb163-1" data-line-number="1">devtools<span class="op">::</span><span class="kw">session_info</span>()</a>
<a class="sourceLine" id="cb163-2" data-line-number="2"><span class="co">## - Session info ---------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb163-3" data-line-number="3"><span class="co">##  setting  value                       </span></a>
<a class="sourceLine" id="cb163-4" data-line-number="4"><span class="co">##  version  R version 4.0.0 (2020-04-24)</span></a>
<a class="sourceLine" id="cb163-5" data-line-number="5"><span class="co">##  os       Windows 10 x64              </span></a>
<a class="sourceLine" id="cb163-6" data-line-number="6"><span class="co">##  system   x86_64, mingw32             </span></a>
<a class="sourceLine" id="cb163-7" data-line-number="7"><span class="co">##  ui       RTerm                       </span></a>
<a class="sourceLine" id="cb163-8" data-line-number="8"><span class="co">##  language (EN)                        </span></a>
<a class="sourceLine" id="cb163-9" data-line-number="9"><span class="co">##  collate  English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb163-10" data-line-number="10"><span class="co">##  ctype    English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb163-11" data-line-number="11"><span class="co">##  tz       Asia/Kuala_Lumpur           </span></a>
<a class="sourceLine" id="cb163-12" data-line-number="12"><span class="co">##  date     2020-06-03                  </span></a>
<a class="sourceLine" id="cb163-13" data-line-number="13"><span class="co">## </span></a>
<a class="sourceLine" id="cb163-14" data-line-number="14"><span class="co">## - Packages -------------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb163-15" data-line-number="15"><span class="co">##  package              * version  date       lib source        </span></a>
<a class="sourceLine" id="cb163-16" data-line-number="16"><span class="co">##  ade4                   1.7-15   2020-02-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-17" data-line-number="17"><span class="co">##  annotate               1.66.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-18" data-line-number="18"><span class="co">##  AnnotationDbi        * 1.50.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-19" data-line-number="19"><span class="co">##  ape                    5.3      2019-03-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-20" data-line-number="20"><span class="co">##  assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-21" data-line-number="21"><span class="co">##  backports              1.1.6    2020-04-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-22" data-line-number="22"><span class="co">##  beeswarm               0.2.3    2016-04-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-23" data-line-number="23"><span class="co">##  bibtex                 0.4.2.2  2020-01-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-24" data-line-number="24"><span class="co">##  Biobase              * 2.48.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-25" data-line-number="25"><span class="co">##  BiocGenerics         * 0.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-26" data-line-number="26"><span class="co">##  BiocNeighbors          1.6.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-27" data-line-number="27"><span class="co">##  BiocParallel           1.22.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-28" data-line-number="28"><span class="co">##  BiocSingular           1.4.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-29" data-line-number="29"><span class="co">##  bit                    1.1-15.2 2020-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-30" data-line-number="30"><span class="co">##  bit64                  0.9-7    2017-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-31" data-line-number="31"><span class="co">##  bitops                 1.0-6    2013-08-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-32" data-line-number="32"><span class="co">##  blob                   1.2.1    2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-33" data-line-number="33"><span class="co">##  bookdown               0.19     2020-05-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-34" data-line-number="34"><span class="co">##  broom                  0.5.6    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-35" data-line-number="35"><span class="co">##  callr                  3.4.3    2020-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-36" data-line-number="36"><span class="co">##  cellranger             1.1.0    2016-07-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-37" data-line-number="37"><span class="co">##  cli                    2.0.2    2020-02-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-38" data-line-number="38"><span class="co">##  cluster              * 2.1.0    2019-06-19 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-39" data-line-number="39"><span class="co">##  clusterExperiment    * 2.8.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-40" data-line-number="40"><span class="co">##  codetools              0.2-16   2018-12-24 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-41" data-line-number="41"><span class="co">##  colorspace             1.4-1    2019-03-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-42" data-line-number="42"><span class="co">##  combinat               0.0-8    2012-10-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-43" data-line-number="43"><span class="co">##  cowplot              * 1.0.0    2019-07-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-44" data-line-number="44"><span class="co">##  crayon                 1.3.4    2017-09-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-45" data-line-number="45"><span class="co">##  data.table             1.12.8   2019-12-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-46" data-line-number="46"><span class="co">##  DBI                    1.1.0    2019-12-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-47" data-line-number="47"><span class="co">##  dbplyr                 1.4.3    2020-04-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-48" data-line-number="48"><span class="co">##  DDRTree                0.1.5    2017-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-49" data-line-number="49"><span class="co">##  DelayedArray         * 0.14.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-50" data-line-number="50"><span class="co">##  DelayedMatrixStats     1.10.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-51" data-line-number="51"><span class="co">##  densityClust           0.3      2017-10-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-52" data-line-number="52"><span class="co">##  desc                   1.2.0    2018-05-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-53" data-line-number="53"><span class="co">##  devtools               2.3.0    2020-04-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-54" data-line-number="54"><span class="co">##  digest                 0.6.25   2020-02-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-55" data-line-number="55"><span class="co">##  docopt                 0.6.1    2018-10-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-56" data-line-number="56"><span class="co">##  doParallel             1.0.15   2019-08-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-57" data-line-number="57"><span class="co">##  dplyr                * 0.8.5    2020-03-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-58" data-line-number="58"><span class="co">##  dqrng                  0.2.1    2019-05-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-59" data-line-number="59"><span class="co">##  edgeR                  3.30.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-60" data-line-number="60"><span class="co">##  ellipsis               0.3.0    2019-09-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-61" data-line-number="61"><span class="co">##  evaluate               0.14     2019-05-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-62" data-line-number="62"><span class="co">##  extrafont            * 0.17     2014-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-63" data-line-number="63"><span class="co">##  extrafontdb            1.0      2012-06-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-64" data-line-number="64"><span class="co">##  fansi                  0.4.1    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-65" data-line-number="65"><span class="co">##  farver                 2.0.3    2020-01-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-66" data-line-number="66"><span class="co">##  fastICA                1.2-2    2019-07-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-67" data-line-number="67"><span class="co">##  fitdistrplus           1.0-14   2019-01-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-68" data-line-number="68"><span class="co">##  FNN                    1.1.3    2019-02-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-69" data-line-number="69"><span class="co">##  forcats              * 0.5.0    2020-03-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-70" data-line-number="70"><span class="co">##  foreach                1.5.0    2020-03-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-71" data-line-number="71"><span class="co">##  fs                     1.4.1    2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-72" data-line-number="72"><span class="co">##  future                 1.17.0   2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-73" data-line-number="73"><span class="co">##  future.apply           1.5.0    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-74" data-line-number="74"><span class="co">##  genefilter             1.70.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-75" data-line-number="75"><span class="co">##  generics               0.0.2    2018-11-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-76" data-line-number="76"><span class="co">##  GenomeInfoDb         * 1.24.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-77" data-line-number="77"><span class="co">##  GenomeInfoDbData       1.2.3    2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-78" data-line-number="78"><span class="co">##  GenomicRanges        * 1.40.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-79" data-line-number="79"><span class="co">##  ggbeeswarm             0.6.0    2017-08-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-80" data-line-number="80"><span class="co">##  ggplot2              * 3.3.0    2020-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-81" data-line-number="81"><span class="co">##  ggrepel                0.8.2    2020-03-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-82" data-line-number="82"><span class="co">##  ggridges               0.5.2    2020-01-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-83" data-line-number="83"><span class="co">##  globals                0.12.5   2019-12-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-84" data-line-number="84"><span class="co">##  glue                   1.4.0    2020-04-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-85" data-line-number="85"><span class="co">##  gridBase               0.4-7    2014-02-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-86" data-line-number="86"><span class="co">##  gridExtra            * 2.3      2017-09-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-87" data-line-number="87"><span class="co">##  gtable                 0.3.0    2019-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-88" data-line-number="88"><span class="co">##  gtools                 3.8.2    2020-03-31 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-89" data-line-number="89"><span class="co">##  haven                  2.2.0    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-90" data-line-number="90"><span class="co">##  HDF5Array              1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-91" data-line-number="91"><span class="co">##  highr                  0.8      2019-03-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-92" data-line-number="92"><span class="co">##  hms                    0.5.3    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-93" data-line-number="93"><span class="co">##  howmany                0.3-1    2012-06-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-94" data-line-number="94"><span class="co">##  HSMMSingleCell         1.8.0    2020-05-07 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-95" data-line-number="95"><span class="co">##  htmltools              0.4.0    2019-10-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-96" data-line-number="96"><span class="co">##  htmlwidgets            1.5.1    2019-10-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-97" data-line-number="97"><span class="co">##  httr                   1.4.1    2019-08-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-98" data-line-number="98"><span class="co">##  ica                    1.0-2    2018-05-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-99" data-line-number="99"><span class="co">##  igraph                 1.2.5    2020-03-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-100" data-line-number="100"><span class="co">##  IRanges              * 2.22.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-101" data-line-number="101"><span class="co">##  irlba                  2.3.3    2019-02-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-102" data-line-number="102"><span class="co">##  iterators              1.0.12   2019-07-26 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-103" data-line-number="103"><span class="co">##  jsonlite               1.6.1    2020-02-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-104" data-line-number="104"><span class="co">##  kernlab                0.9-29   2019-11-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-105" data-line-number="105"><span class="co">##  KernSmooth             2.23-16  2019-10-15 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-106" data-line-number="106"><span class="co">##  knitr                  1.28     2020-02-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-107" data-line-number="107"><span class="co">##  labeling               0.3      2014-08-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-108" data-line-number="108"><span class="co">##  lattice                0.20-41  2020-04-02 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-109" data-line-number="109"><span class="co">##  lazyeval               0.2.2    2019-03-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-110" data-line-number="110"><span class="co">##  leiden                 0.3.3    2020-02-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-111" data-line-number="111"><span class="co">##  lifecycle              0.2.0    2020-03-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-112" data-line-number="112"><span class="co">##  limma                  3.44.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-113" data-line-number="113"><span class="co">##  listenv                0.8.0    2019-12-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-114" data-line-number="114"><span class="co">##  lmtest                 0.9-37   2019-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-115" data-line-number="115"><span class="co">##  locfdr                 1.1-8    2015-07-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-116" data-line-number="116"><span class="co">##  locfit                 1.5-9.4  2020-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-117" data-line-number="117"><span class="co">##  lsei                   1.2-0.1  2020-05-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-118" data-line-number="118"><span class="co">##  lubridate              1.7.8    2020-04-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-119" data-line-number="119"><span class="co">##  magrittr               1.5      2014-11-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-120" data-line-number="120"><span class="co">##  MASS                   7.3-51.5 2019-12-20 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-121" data-line-number="121"><span class="co">##  Matrix                 1.2-18   2019-11-27 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-122" data-line-number="122"><span class="co">##  matrixStats          * 0.56.0   2020-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-123" data-line-number="123"><span class="co">##  mclust               * 5.4.6    2020-04-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-124" data-line-number="124"><span class="co">##  memoise                1.1.0    2017-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-125" data-line-number="125"><span class="co">##  mgcv                   1.8-31   2019-11-09 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-126" data-line-number="126"><span class="co">##  modelr                 0.1.7    2020-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-127" data-line-number="127"><span class="co">##  monocle                2.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-128" data-line-number="128"><span class="co">##  munsell                0.5.0    2018-06-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-129" data-line-number="129"><span class="co">##  nlme                   3.1-147  2020-04-13 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-130" data-line-number="130"><span class="co">##  NMF                  * 0.22.0   2020-02-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-131" data-line-number="131"><span class="co">##  npsurv                 0.4-0.1  2020-05-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-132" data-line-number="132"><span class="co">##  org.Dr.eg.db         * 3.11.1   2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-133" data-line-number="133"><span class="co">##  patchwork              1.0.0    2019-12-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-134" data-line-number="134"><span class="co">##  pbapply                1.4-2    2019-08-31 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-135" data-line-number="135"><span class="co">##  pheatmap             * 1.0.12   2019-01-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-136" data-line-number="136"><span class="co">##  phylobase              0.8.10   2020-03-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-137" data-line-number="137"><span class="co">##  pillar                 1.4.4    2020-05-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-138" data-line-number="138"><span class="co">##  pkgbuild               1.0.8    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-139" data-line-number="139"><span class="co">##  pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-140" data-line-number="140"><span class="co">##  pkgload                1.0.2    2018-10-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-141" data-line-number="141"><span class="co">##  pkgmaker             * 0.31.1   2020-03-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-142" data-line-number="142"><span class="co">##  plotly                 4.9.2.1  2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-143" data-line-number="143"><span class="co">##  plyr                   1.8.6    2020-03-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-144" data-line-number="144"><span class="co">##  png                    0.1-7    2013-12-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-145" data-line-number="145"><span class="co">##  prettyunits            1.1.1    2020-01-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-146" data-line-number="146"><span class="co">##  princurve            * 2.1.4    2019-05-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-147" data-line-number="147"><span class="co">##  processx               3.4.2    2020-02-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-148" data-line-number="148"><span class="co">##  progress               1.2.2    2019-05-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-149" data-line-number="149"><span class="co">##  ps                     1.3.3    2020-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-150" data-line-number="150"><span class="co">##  purrr                * 0.3.4    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-151" data-line-number="151"><span class="co">##  qlcMatrix              0.9.7    2018-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-152" data-line-number="152"><span class="co">##  R.methodsS3          * 1.8.0    2020-02-14 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-153" data-line-number="153"><span class="co">##  R.oo                 * 1.23.0   2019-11-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-154" data-line-number="154"><span class="co">##  R.utils              * 2.9.2    2019-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-155" data-line-number="155"><span class="co">##  R6                     2.4.1    2019-11-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-156" data-line-number="156"><span class="co">##  RANN                   2.6.1    2019-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-157" data-line-number="157"><span class="co">##  rappdirs               0.3.1    2016-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-158" data-line-number="158"><span class="co">##  RColorBrewer         * 1.1-2    2014-12-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-159" data-line-number="159"><span class="co">##  Rcpp                   1.0.4.6  2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-160" data-line-number="160"><span class="co">##  RcppAnnoy              0.0.16   2020-03-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-161" data-line-number="161"><span class="co">##  RCurl                  1.98-1.2 2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-162" data-line-number="162"><span class="co">##  readr                * 1.3.1    2018-12-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-163" data-line-number="163"><span class="co">##  readxl                 1.3.1    2019-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-164" data-line-number="164"><span class="co">##  registry             * 0.5-1    2019-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-165" data-line-number="165"><span class="co">##  remotes                2.1.1    2020-02-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-166" data-line-number="166"><span class="co">##  reprex                 0.3.0    2019-05-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-167" data-line-number="167"><span class="co">##  reshape2               1.4.4    2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-168" data-line-number="168"><span class="co">##  reticulate             1.15     2020-04-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-169" data-line-number="169"><span class="co">##  rhdf5                  2.32.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-170" data-line-number="170"><span class="co">##  Rhdf5lib               1.10.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-171" data-line-number="171"><span class="co">##  rlang                  0.4.6    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-172" data-line-number="172"><span class="co">##  rmarkdown              2.1      2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-173" data-line-number="173"><span class="co">##  rncl                   0.8.4    2020-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-174" data-line-number="174"><span class="co">##  RNeXML                 2.4.4    2020-05-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-175" data-line-number="175"><span class="co">##  rngtools             * 1.5      2020-01-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-176" data-line-number="176"><span class="co">##  ROCR                   1.0-11   2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-177" data-line-number="177"><span class="co">##  rprojroot              1.3-2    2018-01-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-178" data-line-number="178"><span class="co">##  RSpectra               0.16-0   2019-12-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-179" data-line-number="179"><span class="co">##  RSQLite                2.2.0    2020-01-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-180" data-line-number="180"><span class="co">##  rstudioapi             0.11     2020-02-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-181" data-line-number="181"><span class="co">##  rsvd                   1.0.3    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-182" data-line-number="182"><span class="co">##  Rtsne                  0.15     2018-11-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-183" data-line-number="183"><span class="co">##  Rttf2pt1               1.3.8    2020-01-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-184" data-line-number="184"><span class="co">##  rvest                  0.3.5    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-185" data-line-number="185"><span class="co">##  S4Vectors            * 0.26.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-186" data-line-number="186"><span class="co">##  scales               * 1.1.1    2020-05-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-187" data-line-number="187"><span class="co">##  scater               * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-188" data-line-number="188"><span class="co">##  scran                * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-189" data-line-number="189"><span class="co">##  sctransform            0.2.1    2019-12-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-190" data-line-number="190"><span class="co">##  sessioninfo            1.1.1    2018-11-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-191" data-line-number="191"><span class="co">##  Seurat               * 3.1.5    2020-04-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-192" data-line-number="192"><span class="co">##  SingleCellExperiment * 1.10.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-193" data-line-number="193"><span class="co">##  slam                   0.1-47   2019-12-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-194" data-line-number="194"><span class="co">##  slingshot            * 1.6.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-195" data-line-number="195"><span class="co">##  softImpute             1.4      2015-04-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-196" data-line-number="196"><span class="co">##  sparsesvd              0.2      2019-07-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-197" data-line-number="197"><span class="co">##  statmod                1.4.34   2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-198" data-line-number="198"><span class="co">##  stringi                1.4.6    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-199" data-line-number="199"><span class="co">##  stringr              * 1.4.0    2019-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-200" data-line-number="200"><span class="co">##  SummarizedExperiment * 1.18.1   2020-04-30 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-201" data-line-number="201"><span class="co">##  survival               3.1-12   2020-04-10 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-202" data-line-number="202"><span class="co">##  testthat               2.3.2    2020-03-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-203" data-line-number="203"><span class="co">##  tibble               * 3.0.1    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-204" data-line-number="204"><span class="co">##  tidyr                * 1.0.3    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-205" data-line-number="205"><span class="co">##  tidyselect             1.0.0    2020-01-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-206" data-line-number="206"><span class="co">##  tidyverse            * 1.3.0    2019-11-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-207" data-line-number="207"><span class="co">##  tradeSeq             * 1.2.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-208" data-line-number="208"><span class="co">##  tsne                   0.1-3    2016-07-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-209" data-line-number="209"><span class="co">##  usethis                1.6.1    2020-04-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-210" data-line-number="210"><span class="co">##  utf8                   1.1.4    2018-05-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-211" data-line-number="211"><span class="co">##  uuid                   0.1-4    2020-02-26 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-212" data-line-number="212"><span class="co">##  uwot                   0.1.8    2020-03-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-213" data-line-number="213"><span class="co">##  vctrs                  0.2.4    2020-03-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-214" data-line-number="214"><span class="co">##  VGAM                   1.1-3    2020-04-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-215" data-line-number="215"><span class="co">##  vipor                  0.4.5    2017-03-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-216" data-line-number="216"><span class="co">##  viridis              * 0.5.1    2018-03-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-217" data-line-number="217"><span class="co">##  viridisLite          * 0.3.0    2018-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-218" data-line-number="218"><span class="co">##  withr                  2.2.0    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-219" data-line-number="219"><span class="co">##  xfun                   0.13     2020-04-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-220" data-line-number="220"><span class="co">##  XML                    3.99-0.3 2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-221" data-line-number="221"><span class="co">##  xml2                   1.3.2    2020-04-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-222" data-line-number="222"><span class="co">##  xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-223" data-line-number="223"><span class="co">##  XVector                0.28.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-224" data-line-number="224"><span class="co">##  yaml                   2.2.1    2020-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-225" data-line-number="225"><span class="co">##  zinbwave               1.10.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-226" data-line-number="226"><span class="co">##  zlibbioc               1.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb163-227" data-line-number="227"><span class="co">##  zoo                    1.8-8    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb163-228" data-line-number="228"><span class="co">## </span></a>
<a class="sourceLine" id="cb163-229" data-line-number="229"><span class="co">## [1] D:/R_packages</span></a>
<a class="sourceLine" id="cb163-230" data-line-number="230"><span class="co">## [2] C:/Program Files/R/R-4.0.0/library</span></a></code></pre></div>
</details>
</div>
<div id="solutions-4" class="section level2">
<h2><span class="header-section-number">8.5</span> Solutions</h2>
<ol style="list-style-type: decimal">
<li>TB_sce &lt;- TB_sce[, colData(TB_sce)<span class="math inline">\(subclusters_names == &quot;NMps&quot; | colData(TB_sce)\)</span>subclusters_names == “Maturation zone”| colData(TB_sce)<span class="math inline">\(subclusters_names == &quot;Posterior PSM&quot; | colData(TB_sce)\)</span>subclusters_names == “Pou5f3+ Posterior NT”]</li>
</ol>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="subclustering.html" class="navigation navigation-prev navigation-unique" aria-label="Previous page"><i class="fa fa-angle-left"></i></a>

    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["NMps_book.pdf", "NMps_book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
