<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 3 Quality Control | NMps_book.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 3 Quality Control | NMps_book.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 3 Quality Control | NMps_book.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="downloading-data.html"/>
<link rel="next" href="clustering.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="downloading-data.html"><a href="downloading-data.html"><i class="fa fa-check"></i><b>2</b> Downloading data</a></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a><ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#importing-libraries"><i class="fa fa-check"></i><b>3.2</b> Importing libraries</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#importing-the-data"><i class="fa fa-check"></i><b>3.3</b> Importing the data</a></li>
<li class="chapter" data-level="3.4" data-path="quality-control.html"><a href="quality-control.html#preparing-the-column-metadata"><i class="fa fa-check"></i><b>3.4</b> Preparing the column metadata</a></li>
<li class="chapter" data-level="3.5" data-path="quality-control.html"><a href="quality-control.html#preparing-the-row-metadata"><i class="fa fa-check"></i><b>3.5</b> Preparing the row metadata</a></li>
<li class="chapter" data-level="3.6" data-path="quality-control.html"><a href="quality-control.html#constructing-the-single-cell-experiment-object"><i class="fa fa-check"></i><b>3.6</b> Constructing the single cell experiment object</a></li>
<li class="chapter" data-level="3.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-with-scater"><i class="fa fa-check"></i><b>3.7</b> Quality control with scater</a></li>
<li class="chapter" data-level="3.8" data-path="quality-control.html"><a href="quality-control.html#saving-the-rds-object"><i class="fa fa-check"></i><b>3.8</b> Saving the RDS object</a></li>
<li class="chapter" data-level="3.9" data-path="quality-control.html"><a href="quality-control.html#session-info"><i class="fa fa-check"></i><b>3.9</b> Session Info</a></li>
<li class="chapter" data-level="3.10" data-path="quality-control.html"><a href="quality-control.html#solutions"><i class="fa fa-check"></i><b>3.10</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>4</b> Clustering</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="clustering.html"><a href="clustering.html#importing-libraries-1"><i class="fa fa-check"></i><b>4.2</b> Importing libraries</a></li>
<li class="chapter" data-level="4.3" data-path="clustering.html"><a href="clustering.html#importing-data"><i class="fa fa-check"></i><b>4.3</b> Importing Data</a></li>
<li class="chapter" data-level="4.4" data-path="clustering.html"><a href="clustering.html#useful-seurat-object-interaction-methods"><i class="fa fa-check"></i><b>4.4</b> Useful seurat object interaction methods</a></li>
<li class="chapter" data-level="4.5" data-path="clustering.html"><a href="clustering.html#performing-dimensional-reduction"><i class="fa fa-check"></i><b>4.5</b> Performing dimensional reduction</a></li>
<li class="chapter" data-level="4.6" data-path="clustering.html"><a href="clustering.html#clustering-1"><i class="fa fa-check"></i><b>4.6</b> Clustering</a><ul>
<li class="chapter" data-level="4.6.1" data-path="clustering.html"><a href="clustering.html#constructing-a-shared-nearest-neighbour-snn-graph"><i class="fa fa-check"></i><b>4.6.1</b> Constructing a Shared-nearest neighbour (SNN) graph</a></li>
<li class="chapter" data-level="4.6.2" data-path="clustering.html"><a href="clustering.html#clustering-with-the-louvain-algorithm"><i class="fa fa-check"></i><b>4.6.2</b> Clustering with the Louvain algorithm</a></li>
<li class="chapter" data-level="4.6.3" data-path="clustering.html"><a href="clustering.html#resources"><i class="fa fa-check"></i><b>4.6.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="clustering.html"><a href="clustering.html#umap-with-authors-annotations"><i class="fa fa-check"></i><b>4.7</b> UMAP with authors’ annotations</a></li>
<li class="chapter" data-level="4.8" data-path="clustering.html"><a href="clustering.html#t-sne-with-authors-annotations"><i class="fa fa-check"></i><b>4.8</b> t-SNE with authors’ annotations</a></li>
<li class="chapter" data-level="4.9" data-path="clustering.html"><a href="clustering.html#assigning-our-own-cluster-identities"><i class="fa fa-check"></i><b>4.9</b> Assigning our own cluster identities</a></li>
<li class="chapter" data-level="4.10" data-path="clustering.html"><a href="clustering.html#umap-visualisation-of-our-annotations"><i class="fa fa-check"></i><b>4.10</b> UMAP visualisation of our annotations</a></li>
<li class="chapter" data-level="4.11" data-path="clustering.html"><a href="clustering.html#highlighting-tailbud-cells"><i class="fa fa-check"></i><b>4.11</b> Highlighting tailbud cells</a></li>
<li class="chapter" data-level="4.12" data-path="clustering.html"><a href="clustering.html#saving-rds-object"><i class="fa fa-check"></i><b>4.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="4.13" data-path="clustering.html"><a href="clustering.html#session-info-1"><i class="fa fa-check"></i><b>4.13</b> Session Info</a></li>
<li class="chapter" data-level="4.14" data-path="clustering.html"><a href="clustering.html#solutions-1"><i class="fa fa-check"></i><b>4.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="optimisations.html"><a href="optimisations.html"><i class="fa fa-check"></i><b>5</b> Optimisations</a><ul>
<li class="chapter" data-level="5.1" data-path="optimisations.html"><a href="optimisations.html#importing-libraries-2"><i class="fa fa-check"></i><b>5.1</b> Importing libraries</a></li>
<li class="chapter" data-level="5.2" data-path="optimisations.html"><a href="optimisations.html#importing-data-1"><i class="fa fa-check"></i><b>5.2</b> Importing Data</a></li>
<li class="chapter" data-level="5.3" data-path="optimisations.html"><a href="optimisations.html#number-of-pcs-to-retain"><i class="fa fa-check"></i><b>5.3</b> Number of PCs to retain</a><ul>
<li class="chapter" data-level="5.3.1" data-path="optimisations.html"><a href="optimisations.html#scree-plot"><i class="fa fa-check"></i><b>5.3.1</b> Scree plot</a></li>
<li class="chapter" data-level="5.3.2" data-path="optimisations.html"><a href="optimisations.html#examining-technical-noise"><i class="fa fa-check"></i><b>5.3.2</b> Examining technical noise</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="optimisations.html"><a href="optimisations.html#evaluating-clustering-results"><i class="fa fa-check"></i><b>5.4</b> Evaluating clustering results</a><ul>
<li class="chapter" data-level="5.4.1" data-path="optimisations.html"><a href="optimisations.html#cluster-stability"><i class="fa fa-check"></i><b>5.4.1</b> Cluster stability</a></li>
<li class="chapter" data-level="5.4.2" data-path="optimisations.html"><a href="optimisations.html#cluster-separability"><i class="fa fa-check"></i><b>5.4.2</b> Cluster separability</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="optimisations.html"><a href="optimisations.html#comparison-with-authors-labels"><i class="fa fa-check"></i><b>5.5</b> Comparison with authors’ labels</a><ul>
<li class="chapter" data-level="5.5.1" data-path="optimisations.html"><a href="optimisations.html#concordance-in-tailbud-cluster-assignment"><i class="fa fa-check"></i><b>5.5.1</b> Concordance in tailbud cluster assignment</a></li>
<li class="chapter" data-level="5.5.2" data-path="optimisations.html"><a href="optimisations.html#optimising-the-seurat-clustering-parameters"><i class="fa fa-check"></i><b>5.5.2</b> Optimising the seurat clustering parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="optimisations.html"><a href="optimisations.html#selecting-cells-for-downstream-analysis"><i class="fa fa-check"></i><b>5.6</b> Selecting cells for downstream analysis</a></li>
<li class="chapter" data-level="5.7" data-path="optimisations.html"><a href="optimisations.html#session-info-2"><i class="fa fa-check"></i><b>5.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html"><i class="fa fa-check"></i><b>6</b> Differentially Expressed Genes</a><ul>
<li class="chapter" data-level="6.1" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualisation-of-marker-genes-in-our-dataset"><i class="fa fa-check"></i><b>6.1</b> Visualisation of marker genes in our dataset</a></li>
<li class="chapter" data-level="6.2" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#obtaining-a-list-of-differentially-expressed-genes"><i class="fa fa-check"></i><b>6.2</b> Obtaining a list of differentially expressed genes</a></li>
<li class="chapter" data-level="6.3" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#analysing-output-of-the-wilcoxon-test"><i class="fa fa-check"></i><b>6.3</b> Analysing output of the wilcoxon test</a></li>
<li class="chapter" data-level="6.4" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualising-the-expression-of-the-marker-genes"><i class="fa fa-check"></i><b>6.4</b> Visualising the expression of the marker genes</a></li>
<li class="chapter" data-level="6.5" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#session-info-3"><i class="fa fa-check"></i><b>6.5</b> Session Info</a></li>
<li class="chapter" data-level="6.6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#solutions-2"><i class="fa fa-check"></i><b>6.6</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="subclustering.html"><a href="subclustering.html"><i class="fa fa-check"></i><b>7</b> Subclustering</a><ul>
<li class="chapter" data-level="7.1" data-path="subclustering.html"><a href="subclustering.html#importing-libraries-3"><i class="fa fa-check"></i><b>7.1</b> Importing libraries</a></li>
<li class="chapter" data-level="7.2" data-path="subclustering.html"><a href="subclustering.html#importing-data-2"><i class="fa fa-check"></i><b>7.2</b> Importing Data</a></li>
<li class="chapter" data-level="7.3" data-path="subclustering.html"><a href="subclustering.html#subclustering-the-tb-cells---using-only-cells-that-are-assigned-to-the-shared-cluster-across-all-the-clustering-resolutions."><i class="fa fa-check"></i><b>7.3</b> Subclustering the TB cells - Using only cells that are assigned to the ‘shared’ cluster across all the clustering resolutions.</a></li>
<li class="chapter" data-level="7.4" data-path="subclustering.html"><a href="subclustering.html#dimensional-reduction-with-umap-and-clustering"><i class="fa fa-check"></i><b>7.4</b> Dimensional reduction with UMAP and clustering</a></li>
<li class="chapter" data-level="7.5" data-path="subclustering.html"><a href="subclustering.html#identifying-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.5</b> Identifying marker genes for each cluster</a></li>
<li class="chapter" data-level="7.6" data-path="subclustering.html"><a href="subclustering.html#merging-clusters-that-do-not-appear-distinct"><i class="fa fa-check"></i><b>7.6</b> Merging clusters that do not appear distinct</a></li>
<li class="chapter" data-level="7.7" data-path="subclustering.html"><a href="subclustering.html#plotting-umap-and-assigning-cluster-labels"><i class="fa fa-check"></i><b>7.7</b> Plotting UMAP and assigning cluster labels</a></li>
<li class="chapter" data-level="7.8" data-path="subclustering.html"><a href="subclustering.html#checking-the-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.8</b> Checking the marker genes for each cluster</a></li>
<li class="chapter" data-level="7.9" data-path="subclustering.html"><a href="subclustering.html#sox2-tbxta-coexpression"><i class="fa fa-check"></i><b>7.9</b> Sox2 tbxta coexpression</a></li>
<li class="chapter" data-level="7.10" data-path="subclustering.html"><a href="subclustering.html#viewing-genes-of-interest-with-a-dotplot"><i class="fa fa-check"></i><b>7.10</b> Viewing genes of interest with a dotplot</a></li>
<li class="chapter" data-level="7.11" data-path="subclustering.html"><a href="subclustering.html#exporting-the-violin-plots-for-genes-of-interest."><i class="fa fa-check"></i><b>7.11</b> Exporting the violin plots for genes of interest.</a></li>
<li class="chapter" data-level="7.12" data-path="subclustering.html"><a href="subclustering.html#saving-rds-object-1"><i class="fa fa-check"></i><b>7.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="7.13" data-path="subclustering.html"><a href="subclustering.html#session-info-4"><i class="fa fa-check"></i><b>7.13</b> Session Info</a></li>
<li class="chapter" data-level="7.14" data-path="subclustering.html"><a href="subclustering.html#solutions-3"><i class="fa fa-check"></i><b>7.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="pseudotime.html"><a href="pseudotime.html"><i class="fa fa-check"></i><b>8</b> Pseudotime</a><ul>
<li class="chapter" data-level="8.1" data-path="pseudotime.html"><a href="pseudotime.html#importing-libraries-4"><i class="fa fa-check"></i><b>8.1</b> Importing libraries</a></li>
<li class="chapter" data-level="8.2" data-path="pseudotime.html"><a href="pseudotime.html#importing-data-3"><i class="fa fa-check"></i><b>8.2</b> Importing data</a></li>
<li class="chapter" data-level="8.3" data-path="pseudotime.html"><a href="pseudotime.html#tradeseq-workflow-analyses"><i class="fa fa-check"></i><b>8.3</b> TradeSeq workflow analyses</a></li>
<li class="chapter" data-level="8.4" data-path="pseudotime.html"><a href="pseudotime.html#session-info-5"><i class="fa fa-check"></i><b>8.4</b> Session Info</a></li>
<li class="chapter" data-level="8.5" data-path="pseudotime.html"><a href="pseudotime.html#solutions-4"><i class="fa fa-check"></i><b>8.5</b> Solutions</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="quality-control" class="section level1">
<h1><span class="header-section-number">Chapter 3</span> Quality Control</h1>
<div id="overview" class="section level2">
<h2><span class="header-section-number">3.1</span> Overview</h2>
<p>In this section, we will construct the SingleCellExperiment object, and then perform quality control on the dataset. The workflow below, taken from chapter 5 of this text <a href="https://osca.bioconductor.org/overview.html#introduction-1">Orchestrating Single-Cell Analysis</a>, neatly summarises the entire scRNA-seq workflow.</p>
<p><em>Q1) Which steps of the workflow are we executing in this notebook?</em></p>
<div class="figure"><span id="fig:unnamed-chunk-3"></span>
<img src="1_workflow.png" alt="Figure 1: Schematic of a typical scRNA-seq analysis workflow. Each stage (separated by dashed lines) consists of a number of specific steps, many of which operate on and modify a SingleCellExperiment instance." width="735" />
<p class="caption">
Figure 3.1: Figure 1: Schematic of a typical scRNA-seq analysis workflow. Each stage (separated by dashed lines) consists of a number of specific steps, many of which operate on and modify a SingleCellExperiment instance.
</p>
</div>
</div>
<div id="importing-libraries" class="section level2">
<h2><span class="header-section-number">3.2</span> Importing libraries</h2>
<div class="sourceCode" id="cb3"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb3-1" data-line-number="1"><span class="co">###  Clearing workspace and loading the required packages</span></a>
<a class="sourceLine" id="cb3-2" data-line-number="2"><span class="kw">library</span>(scater) <span class="co">#For quality control</span></a>
<a class="sourceLine" id="cb3-3" data-line-number="3"><span class="kw">library</span>(SingleCellExperiment) <span class="co"># Data structure to house out data</span></a>
<a class="sourceLine" id="cb3-4" data-line-number="4"><span class="kw">library</span>(tidyverse) </a>
<a class="sourceLine" id="cb3-5" data-line-number="5"><span class="kw">library</span>(org.Dr.eg.db) <span class="co">#Zebrafish annotation database</span></a></code></pre></div>
</div>
<div id="importing-the-data" class="section level2">
<h2><span class="header-section-number">3.3</span> Importing the data</h2>
<p><em>Q2) What does the %&gt;% symbol do?</em></p>
<div class="sourceCode" id="cb4"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb4-1" data-line-number="1"><span class="co"># Importing count matrix, cluster IDs and cluster names.</span></a>
<a class="sourceLine" id="cb4-2" data-line-number="2">count_file &lt;-<span class="st"> &quot;./hpf18_count_matrix.csv&quot;</span> </a>
<a class="sourceLine" id="cb4-3" data-line-number="3">clustid_file &lt;-<span class="st"> &quot;./hpf18_clustID.txt&quot;</span></a>
<a class="sourceLine" id="cb4-4" data-line-number="4">clust_names_file &lt;-<span class="st"> &quot;./hpf18_clustNames.csv&quot;</span></a>
<a class="sourceLine" id="cb4-5" data-line-number="5"></a>
<a class="sourceLine" id="cb4-6" data-line-number="6">count_matrix &lt;-<span class="st"> </span><span class="kw">read_csv</span>(count_file) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">column_to_rownames</span>(<span class="dt">var =</span> <span class="st">&#39;Row&#39;</span>)</a>
<a class="sourceLine" id="cb4-7" data-line-number="7">clustid &lt;-<span class="st"> </span><span class="kw">read_table</span>(clustid_file, <span class="dt">col_names =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb4-8" data-line-number="8">clust_names &lt;-<span class="st"> </span><span class="kw">read_csv</span>(clust_names_file, <span class="dt">col_types =</span> <span class="kw">cols</span>(</a>
<a class="sourceLine" id="cb4-9" data-line-number="9">  <span class="st">&quot;TimePoint(hpf)&quot;</span>=<span class="st"> </span><span class="kw">col_double</span>(), </a>
<a class="sourceLine" id="cb4-10" data-line-number="10"> <span class="st">&quot;ClusterID&quot;</span> =<span class="st"> </span><span class="kw">col_double</span>(), </a>
<a class="sourceLine" id="cb4-11" data-line-number="11">  <span class="st">&quot;ClusterName&quot;</span> =<span class="st"> </span><span class="kw">col_character</span>()))</a></code></pre></div>
</div>
<div id="preparing-the-column-metadata" class="section level2">
<h2><span class="header-section-number">3.4</span> Preparing the column metadata</h2>
<p>In the next code chunk, we will prepare the column metadata which contains information about each cell in the dataset. We will want to append the zebrafish stage as a prefix to the cell names, in case we want to keep track of cells from different developmental stages in our downstream analyses. We will also extract the library names from the cell names using a simple <code>regular expression</code> that succinctly represents our search target. Lastly, we will combine the updated cell names, library names, cluster IDs and cluster names into a single dataframe, and rename the columns accordingly. To do this, we use the <code>reduce</code> function, which iteratively applies the function <code>cbind</code> (short for column bind) to a list containing our columns of interest. We also used the <code>left_join</code> function from the <code>dplyr</code> package. <a href="https://statisticsglobe.com/r-dplyr-join-inner-left-right-full-semi-anti#left_join">Join functions</a></p>
<p><em>Q3) What does the [0-9]* syntax within the regular expression represent?</em></p>
<p><em>Q4) Must we use left_join to achieve our goal here?</em></p>
<div class="sourceCode" id="cb5"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb5-1" data-line-number="1"><span class="co"># Preparing column data file</span></a>
<a class="sourceLine" id="cb5-2" data-line-number="2">hpf18_clusternames &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">filter</span>(clust_names, <span class="st">`</span><span class="dt">TimePoint(hpf)</span><span class="st">`</span> <span class="op">==</span><span class="st"> </span><span class="dv">18</span>)</a>
<a class="sourceLine" id="cb5-3" data-line-number="3"></a>
<a class="sourceLine" id="cb5-4" data-line-number="4"><span class="co"># Appending 18hpf to cell names </span></a>
<a class="sourceLine" id="cb5-5" data-line-number="5"><span class="kw">colnames</span>(count_matrix) &lt;-<span class="st"> </span><span class="kw">paste0</span>(<span class="st">&quot;18hpf_&quot;</span>, <span class="kw">colnames</span>(count_matrix))</a>
<a class="sourceLine" id="cb5-6" data-line-number="6">cellnames &lt;-<span class="st"> </span><span class="kw">colnames</span>(count_matrix)</a>
<a class="sourceLine" id="cb5-7" data-line-number="7"></a>
<a class="sourceLine" id="cb5-8" data-line-number="8"><span class="co"># Extracting library names from cell names</span></a>
<a class="sourceLine" id="cb5-9" data-line-number="9">pattern=<span class="st">&quot;DEW[0-9]*&quot;</span></a>
<a class="sourceLine" id="cb5-10" data-line-number="10">libnames &lt;-<span class="st"> </span><span class="kw">regmatches</span>(cellnames,<span class="kw">regexec</span>(pattern,cellnames)) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">unlist</span>()</a>
<a class="sourceLine" id="cb5-11" data-line-number="11"></a>
<a class="sourceLine" id="cb5-12" data-line-number="12"><span class="co"># Merging all 3 dataframes and renaming columns</span></a>
<a class="sourceLine" id="cb5-13" data-line-number="13"></a>
<a class="sourceLine" id="cb5-14" data-line-number="14">name_cluster_lib &lt;-<span class="st"> </span><span class="kw">Reduce</span>(cbind, <span class="kw">list</span>(cellnames, clustid, libnames))</a>
<a class="sourceLine" id="cb5-15" data-line-number="15">colDataMatrix &lt;-<span class="st"> </span>dplyr<span class="op">::</span><span class="kw">left_join</span>(name_cluster_lib, hpf18_clusternames, <span class="dt">by =</span> <span class="kw">c</span>(<span class="st">&quot;X1&quot;</span> =<span class="st"> &quot;ClusterID&quot;</span>)) <span class="op">%&gt;%</span><span class="st"> </span></a>
<a class="sourceLine" id="cb5-16" data-line-number="16"><span class="st">                </span>dplyr<span class="op">::</span><span class="kw">rename</span>(<span class="dt">cells =</span> init, </a>
<a class="sourceLine" id="cb5-17" data-line-number="17">                <span class="dt">clusterID =</span> X1,</a>
<a class="sourceLine" id="cb5-18" data-line-number="18">                <span class="dt">libraryName =</span> <span class="st">`</span><span class="dt">x[[i]]</span><span class="st">`</span>,</a>
<a class="sourceLine" id="cb5-19" data-line-number="19">                <span class="dt">timepoint =</span> <span class="st">`</span><span class="dt">TimePoint(hpf)</span><span class="st">`</span>,</a>
<a class="sourceLine" id="cb5-20" data-line-number="20">                <span class="dt">clusterNames =</span> ClusterName)</a></code></pre></div>
<div class="sourceCode" id="cb6"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb6-1" data-line-number="1"><span class="kw">dim</span>(colDataMatrix)</a></code></pre></div>
<pre><code>## [1] 6962    5</code></pre>
</div>
<div id="preparing-the-row-metadata" class="section level2">
<h2><span class="header-section-number">3.5</span> Preparing the row metadata</h2>
<p>We use the <code>AnnotationDbi</code> package below to obtain the ensembl gene identifiers for our gene symbols. <a href="https://www.r-bloggers.com/converting-gene-names-in-r-with-annotationdbi/">Converting Gene names</a>. For more information about annotation packages, see <a href="https://www.bioconductor.org/help/course-materials/2014/SeattleOct2014/B02.4_Annotation.html">Annotations</a></p>
<div class="sourceCode" id="cb8"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb8-1" data-line-number="1"><span class="co"># Preparing row data file</span></a>
<a class="sourceLine" id="cb8-2" data-line-number="2"><span class="co">#keytypes(org.Dr.eg.db)</span></a>
<a class="sourceLine" id="cb8-3" data-line-number="3">ensembl &lt;-<span class="st"> </span>AnnotationDbi<span class="op">::</span><span class="kw">mapIds</span>(org.Dr.eg.db, <span class="dt">keys=</span><span class="kw">rownames</span>(count_matrix), <span class="dt">keytype =</span> <span class="st">&quot;SYMBOL&quot;</span>,<span class="dt">column =</span> <span class="st">&quot;ENSEMBL&quot;</span>)</a></code></pre></div>
<pre><code>## &#39;select()&#39; returned 1:many mapping between keys and columns</code></pre>
<div class="sourceCode" id="cb10"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb10-1" data-line-number="1"><span class="kw">tail</span>(ensembl)</a></code></pre></div>
<pre><code>##               zwilch                zyg11              zymnd12 
## &quot;ENSDARG00000057100&quot; &quot;ENSDARG00000007737&quot;                   NA 
##                  zyx                zzef1                 zzz3 
## &quot;ENSDARG00000075733&quot; &quot;ENSDARG00000094380&quot; &quot;ENSDARG00000077581&quot;</code></pre>
</div>
<div id="constructing-the-single-cell-experiment-object" class="section level2">
<h2><span class="header-section-number">3.6</span> Constructing the single cell experiment object</h2>
<p>At this stage, we have the following prepared:</p>
<ol style="list-style-type: decimal">
<li>The raw count matrix of the 18hpf zebrafish dataset</li>
<li>The column metadata which describes information regarding each cell such as its Dropseq library name, clusterID assigned by the authors and the associated cluster name</li>
<li>The row metadata which contains the ensembl gene names.</li>
</ol>
<p>We will parse all 3 datasets as arguments into the <code>SingleCellExperiment()</code> constructor.</p>
<div class="sourceCode" id="cb12"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb12-1" data-line-number="1"><span class="co"># Creating SCE object and obtaining Ensembl IDs</span></a>
<a class="sourceLine" id="cb12-2" data-line-number="2"></a>
<a class="sourceLine" id="cb12-3" data-line-number="3">hpf18 &lt;-<span class="st"> </span><span class="kw">SingleCellExperiment</span>(<span class="dt">assays =</span> <span class="kw">list</span>(<span class="dt">counts =</span> <span class="kw">as.matrix</span>(count_matrix)), <span class="dt">colData =</span> colDataMatrix)</a>
<a class="sourceLine" id="cb12-4" data-line-number="4"></a>
<a class="sourceLine" id="cb12-5" data-line-number="5"><span class="kw">rowData</span>(hpf18)<span class="op">$</span>ensembl &lt;-<span class="st"> </span>ensembl</a></code></pre></div>
</div>
<div id="quality-control-with-scater" class="section level2">
<h2><span class="header-section-number">3.7</span> Quality control with scater</h2>
<p>Using the scater package, we obtain useful QC metrics for the dataset and discard cells with abnormally low total number of RNA or expressed features.</p>
<p><em>Q5) How does the <code>quickPerCellQC</code> function determine which cells are outliers? What is the median absolute deviation?</em></p>
<div class="sourceCode" id="cb13"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb13-1" data-line-number="1">stats &lt;-<span class="st"> </span><span class="kw">perCellQCMetrics</span>(hpf18, <span class="dt">exprs_values =</span> <span class="st">&quot;counts&quot;</span>, <span class="dt">percent_top =</span> <span class="kw">c</span>(<span class="dv">50</span>, <span class="dv">100</span>, <span class="dv">200</span>, <span class="dv">500</span>),</a>
<a class="sourceLine" id="cb13-2" data-line-number="2"><span class="dt">detection_limit =</span> <span class="dv">0</span>)</a>
<a class="sourceLine" id="cb13-3" data-line-number="3"></a>
<a class="sourceLine" id="cb13-4" data-line-number="4">qc &lt;-<span class="st"> </span><span class="kw">quickPerCellQC</span>(stats, <span class="dt">batch =</span> hpf18<span class="op">$</span>libraryName)</a>
<a class="sourceLine" id="cb13-5" data-line-number="5"><span class="kw">colData</span>(hpf18) &lt;-<span class="st"> </span><span class="kw">cbind</span>(<span class="kw">colData</span>(hpf18), stats)</a>
<a class="sourceLine" id="cb13-6" data-line-number="6"></a>
<a class="sourceLine" id="cb13-7" data-line-number="7">hpf18<span class="op">$</span>discard &lt;-<span class="st"> </span>qc<span class="op">$</span>discard</a>
<a class="sourceLine" id="cb13-8" data-line-number="8">gridExtra<span class="op">::</span><span class="kw">grid.arrange</span>(</a>
<a class="sourceLine" id="cb13-9" data-line-number="9">    <span class="kw">plotColData</span>(hpf18, <span class="dt">x =</span> <span class="st">&quot;libraryName&quot;</span>, <span class="dt">y=</span><span class="st">&quot;sum&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;discard&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-10" data-line-number="10"><span class="st">        </span><span class="kw">scale_y_log10</span>() <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;Total count&quot;</span>),</a>
<a class="sourceLine" id="cb13-11" data-line-number="11">    <span class="kw">plotColData</span>(hpf18, <span class="dt">x =</span> <span class="st">&quot;libraryName&quot;</span>, <span class="dt">y=</span><span class="st">&quot;detected&quot;</span>, <span class="dt">colour_by=</span><span class="st">&quot;discard&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb13-12" data-line-number="12"><span class="st">        </span><span class="kw">scale_y_log10</span>() <span class="op">+</span><span class="st"> </span><span class="kw">ggtitle</span>(<span class="st">&quot;Detected features&quot;</span>)</a>
<a class="sourceLine" id="cb13-13" data-line-number="13">)</a></code></pre></div>
<div class="figure"><span id="fig:unnamed-chunk-6"></span>
<img src="NMps_book_files/figure-html/unnamed-chunk-6-1.png" alt="Figure 2: Quality control plots from the 18hpf zebrafish data, with the total RNA count and detected features as the quality metrics. Each point is a cell and colored according to whether it is identified as an outlier within each batch (library names) or not." width="672" />
<p class="caption">
Figure 3.2: Figure 2: Quality control plots from the 18hpf zebrafish data, with the total RNA count and detected features as the quality metrics. Each point is a cell and colored according to whether it is identified as an outlier within each batch (library names) or not.
</p>
</div>
<p>Notice that there are only a few cells that are discarded. This is because the authors have already performed a first pass in the quality control.</p>
<div class="sourceCode" id="cb14"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb14-1" data-line-number="1"><span class="kw">colSums</span>(<span class="kw">as.matrix</span>(qc))</a></code></pre></div>
<pre><code>##   low_lib_size low_n_features        discard 
##              0              8              8</code></pre>
<p>Apart from removing the outlier cells and thus reducing the dimensions of the column space, we can also reduce the number of features. We take a conservative approach here: we retain all genes that are expressed in at least 1 cell in the dataset.</p>
<div class="sourceCode" id="cb16"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb16-1" data-line-number="1"><span class="co"># Retaining genes that are expressed in at least 1 cell in the dataset</span></a>
<a class="sourceLine" id="cb16-2" data-line-number="2">ave_counts &lt;-<span class="st"> </span><span class="kw">rowSums</span>(<span class="kw">assay</span>(hpf18, <span class="st">&quot;counts&quot;</span>)) <span class="op">/</span><span class="st"> </span><span class="kw">dim</span>(hpf18)[<span class="dv">2</span>]</a>
<a class="sourceLine" id="cb16-3" data-line-number="3">to_keep &lt;-<span class="st"> </span>ave_counts <span class="op">&gt;</span><span class="st"> </span><span class="dv">0</span></a>
<a class="sourceLine" id="cb16-4" data-line-number="4"><span class="kw">sum</span>(<span class="op">!</span>to_keep)</a></code></pre></div>
<pre><code>## [1] 381</code></pre>
<p>We subset our dataset with the 2 boolean conditions established above and save our results.</p>
<div class="sourceCode" id="cb18"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb18-1" data-line-number="1">hpf18&lt;-<span class="st"> </span>hpf18[to_keep, <span class="op">!</span>qc<span class="op">$</span>discard]</a>
<a class="sourceLine" id="cb18-2" data-line-number="2">hpf18</a></code></pre></div>
<pre><code>## class: SingleCellExperiment 
## dim: 30296 6954 
## metadata(0):
## assays(1): counts
## rownames(30296): LOC100000009 LOC100000024 ... zzef1 zzz3
## rowData names(1): ensembl
## colnames(6954): 18hpf_DEW038_TGAGAAGCACT_CAACGTGG
##   18hpf_DEW038_TGAACGCTCAG_AACCCTTG ...
##   18hpf_DEW003_TGACAAGTCAT_GTTTACGT 18hpf_DEW003_TGCAAGGG_GGGAACGA
## colData names(13): cells clusterID ... total discard
## reducedDimNames(0):
## altExpNames(0):</code></pre>
</div>
<div id="saving-the-rds-object" class="section level2">
<h2><span class="header-section-number">3.8</span> Saving the RDS object</h2>
<div class="sourceCode" id="cb20"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb20-1" data-line-number="1"><span class="kw">saveRDS</span>(hpf18, <span class="dt">file =</span> <span class="st">&quot;./rds/hpf18_QCed_new.rds&quot;</span> )</a>
<a class="sourceLine" id="cb20-2" data-line-number="2"><span class="kw">write_csv</span>(<span class="kw">as_tibble</span>(<span class="kw">colData</span>(hpf18)), <span class="dt">path =</span> <span class="st">&quot;./hpf18_coldata.csv&quot;</span>)</a></code></pre></div>
</div>
<div id="session-info" class="section level2">
<h2><span class="header-section-number">3.9</span> Session Info</h2>
<details>
<p><summary>View Session Info</summary></p>
<div class="sourceCode" id="cb21"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb21-1" data-line-number="1">devtools<span class="op">::</span><span class="kw">session_info</span>()</a>
<a class="sourceLine" id="cb21-2" data-line-number="2"><span class="co">## - Session info ---------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb21-3" data-line-number="3"><span class="co">##  setting  value                       </span></a>
<a class="sourceLine" id="cb21-4" data-line-number="4"><span class="co">##  version  R version 4.0.0 (2020-04-24)</span></a>
<a class="sourceLine" id="cb21-5" data-line-number="5"><span class="co">##  os       Windows 10 x64              </span></a>
<a class="sourceLine" id="cb21-6" data-line-number="6"><span class="co">##  system   x86_64, mingw32             </span></a>
<a class="sourceLine" id="cb21-7" data-line-number="7"><span class="co">##  ui       RTerm                       </span></a>
<a class="sourceLine" id="cb21-8" data-line-number="8"><span class="co">##  language (EN)                        </span></a>
<a class="sourceLine" id="cb21-9" data-line-number="9"><span class="co">##  collate  English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb21-10" data-line-number="10"><span class="co">##  ctype    English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb21-11" data-line-number="11"><span class="co">##  tz       Asia/Kuala_Lumpur           </span></a>
<a class="sourceLine" id="cb21-12" data-line-number="12"><span class="co">##  date     2020-06-03                  </span></a>
<a class="sourceLine" id="cb21-13" data-line-number="13"><span class="co">## </span></a>
<a class="sourceLine" id="cb21-14" data-line-number="14"><span class="co">## - Packages -------------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb21-15" data-line-number="15"><span class="co">##  package              * version  date       lib source        </span></a>
<a class="sourceLine" id="cb21-16" data-line-number="16"><span class="co">##  AnnotationDbi        * 1.50.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-17" data-line-number="17"><span class="co">##  assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-18" data-line-number="18"><span class="co">##  backports              1.1.6    2020-04-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-19" data-line-number="19"><span class="co">##  beeswarm               0.2.3    2016-04-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-20" data-line-number="20"><span class="co">##  Biobase              * 2.48.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-21" data-line-number="21"><span class="co">##  BiocGenerics         * 0.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-22" data-line-number="22"><span class="co">##  BiocNeighbors          1.6.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-23" data-line-number="23"><span class="co">##  BiocParallel           1.22.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-24" data-line-number="24"><span class="co">##  BiocSingular           1.4.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-25" data-line-number="25"><span class="co">##  bit                    1.1-15.2 2020-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-26" data-line-number="26"><span class="co">##  bit64                  0.9-7    2017-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-27" data-line-number="27"><span class="co">##  bitops                 1.0-6    2013-08-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-28" data-line-number="28"><span class="co">##  blob                   1.2.1    2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-29" data-line-number="29"><span class="co">##  bookdown               0.19     2020-05-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-30" data-line-number="30"><span class="co">##  broom                  0.5.6    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-31" data-line-number="31"><span class="co">##  callr                  3.4.3    2020-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-32" data-line-number="32"><span class="co">##  cellranger             1.1.0    2016-07-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-33" data-line-number="33"><span class="co">##  cli                    2.0.2    2020-02-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-34" data-line-number="34"><span class="co">##  colorspace             1.4-1    2019-03-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-35" data-line-number="35"><span class="co">##  cowplot                1.0.0    2019-07-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-36" data-line-number="36"><span class="co">##  crayon                 1.3.4    2017-09-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-37" data-line-number="37"><span class="co">##  DBI                    1.1.0    2019-12-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-38" data-line-number="38"><span class="co">##  dbplyr                 1.4.3    2020-04-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-39" data-line-number="39"><span class="co">##  DelayedArray         * 0.14.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-40" data-line-number="40"><span class="co">##  DelayedMatrixStats     1.10.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-41" data-line-number="41"><span class="co">##  desc                   1.2.0    2018-05-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-42" data-line-number="42"><span class="co">##  devtools               2.3.0    2020-04-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-43" data-line-number="43"><span class="co">##  digest                 0.6.25   2020-02-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-44" data-line-number="44"><span class="co">##  dplyr                * 0.8.5    2020-03-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-45" data-line-number="45"><span class="co">##  ellipsis               0.3.0    2019-09-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-46" data-line-number="46"><span class="co">##  evaluate               0.14     2019-05-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-47" data-line-number="47"><span class="co">##  fansi                  0.4.1    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-48" data-line-number="48"><span class="co">##  farver                 2.0.3    2020-01-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-49" data-line-number="49"><span class="co">##  forcats              * 0.5.0    2020-03-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-50" data-line-number="50"><span class="co">##  fs                     1.4.1    2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-51" data-line-number="51"><span class="co">##  generics               0.0.2    2018-11-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-52" data-line-number="52"><span class="co">##  GenomeInfoDb         * 1.24.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-53" data-line-number="53"><span class="co">##  GenomeInfoDbData       1.2.3    2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-54" data-line-number="54"><span class="co">##  GenomicRanges        * 1.40.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-55" data-line-number="55"><span class="co">##  ggbeeswarm             0.6.0    2017-08-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-56" data-line-number="56"><span class="co">##  ggplot2              * 3.3.0    2020-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-57" data-line-number="57"><span class="co">##  glue                   1.4.0    2020-04-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-58" data-line-number="58"><span class="co">##  gridExtra              2.3      2017-09-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-59" data-line-number="59"><span class="co">##  gtable                 0.3.0    2019-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-60" data-line-number="60"><span class="co">##  haven                  2.2.0    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-61" data-line-number="61"><span class="co">##  highr                  0.8      2019-03-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-62" data-line-number="62"><span class="co">##  hms                    0.5.3    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-63" data-line-number="63"><span class="co">##  htmltools              0.4.0    2019-10-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-64" data-line-number="64"><span class="co">##  httr                   1.4.1    2019-08-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-65" data-line-number="65"><span class="co">##  IRanges              * 2.22.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-66" data-line-number="66"><span class="co">##  irlba                  2.3.3    2019-02-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-67" data-line-number="67"><span class="co">##  jsonlite               1.6.1    2020-02-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-68" data-line-number="68"><span class="co">##  knitr                  1.28     2020-02-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-69" data-line-number="69"><span class="co">##  lattice                0.20-41  2020-04-02 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-70" data-line-number="70"><span class="co">##  lifecycle              0.2.0    2020-03-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-71" data-line-number="71"><span class="co">##  lubridate              1.7.8    2020-04-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-72" data-line-number="72"><span class="co">##  magrittr               1.5      2014-11-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-73" data-line-number="73"><span class="co">##  Matrix                 1.2-18   2019-11-27 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-74" data-line-number="74"><span class="co">##  matrixStats          * 0.56.0   2020-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-75" data-line-number="75"><span class="co">##  memoise                1.1.0    2017-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-76" data-line-number="76"><span class="co">##  modelr                 0.1.7    2020-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-77" data-line-number="77"><span class="co">##  munsell                0.5.0    2018-06-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-78" data-line-number="78"><span class="co">##  nlme                   3.1-147  2020-04-13 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-79" data-line-number="79"><span class="co">##  org.Dr.eg.db         * 3.11.1   2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-80" data-line-number="80"><span class="co">##  pillar                 1.4.4    2020-05-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-81" data-line-number="81"><span class="co">##  pkgbuild               1.0.8    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-82" data-line-number="82"><span class="co">##  pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-83" data-line-number="83"><span class="co">##  pkgload                1.0.2    2018-10-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-84" data-line-number="84"><span class="co">##  png                    0.1-7    2013-12-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-85" data-line-number="85"><span class="co">##  prettyunits            1.1.1    2020-01-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-86" data-line-number="86"><span class="co">##  processx               3.4.2    2020-02-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-87" data-line-number="87"><span class="co">##  ps                     1.3.3    2020-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-88" data-line-number="88"><span class="co">##  purrr                * 0.3.4    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-89" data-line-number="89"><span class="co">##  R.methodsS3          * 1.8.0    2020-02-14 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-90" data-line-number="90"><span class="co">##  R.oo                 * 1.23.0   2019-11-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-91" data-line-number="91"><span class="co">##  R.utils              * 2.9.2    2019-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-92" data-line-number="92"><span class="co">##  R6                     2.4.1    2019-11-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-93" data-line-number="93"><span class="co">##  Rcpp                   1.0.4.6  2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-94" data-line-number="94"><span class="co">##  RCurl                  1.98-1.2 2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-95" data-line-number="95"><span class="co">##  readr                * 1.3.1    2018-12-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-96" data-line-number="96"><span class="co">##  readxl                 1.3.1    2019-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-97" data-line-number="97"><span class="co">##  remotes                2.1.1    2020-02-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-98" data-line-number="98"><span class="co">##  reprex                 0.3.0    2019-05-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-99" data-line-number="99"><span class="co">##  rlang                  0.4.6    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-100" data-line-number="100"><span class="co">##  rmarkdown              2.1      2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-101" data-line-number="101"><span class="co">##  rprojroot              1.3-2    2018-01-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-102" data-line-number="102"><span class="co">##  RSQLite                2.2.0    2020-01-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-103" data-line-number="103"><span class="co">##  rstudioapi             0.11     2020-02-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-104" data-line-number="104"><span class="co">##  rsvd                   1.0.3    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-105" data-line-number="105"><span class="co">##  rvest                  0.3.5    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-106" data-line-number="106"><span class="co">##  S4Vectors            * 0.26.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-107" data-line-number="107"><span class="co">##  scales                 1.1.1    2020-05-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-108" data-line-number="108"><span class="co">##  scater               * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-109" data-line-number="109"><span class="co">##  sessioninfo            1.1.1    2018-11-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-110" data-line-number="110"><span class="co">##  SingleCellExperiment * 1.10.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-111" data-line-number="111"><span class="co">##  stringi                1.4.6    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-112" data-line-number="112"><span class="co">##  stringr              * 1.4.0    2019-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-113" data-line-number="113"><span class="co">##  SummarizedExperiment * 1.18.1   2020-04-30 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-114" data-line-number="114"><span class="co">##  testthat               2.3.2    2020-03-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-115" data-line-number="115"><span class="co">##  tibble               * 3.0.1    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-116" data-line-number="116"><span class="co">##  tidyr                * 1.0.3    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-117" data-line-number="117"><span class="co">##  tidyselect             1.0.0    2020-01-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-118" data-line-number="118"><span class="co">##  tidyverse            * 1.3.0    2019-11-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-119" data-line-number="119"><span class="co">##  usethis                1.6.1    2020-04-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-120" data-line-number="120"><span class="co">##  vctrs                  0.2.4    2020-03-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-121" data-line-number="121"><span class="co">##  vipor                  0.4.5    2017-03-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-122" data-line-number="122"><span class="co">##  viridis                0.5.1    2018-03-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-123" data-line-number="123"><span class="co">##  viridisLite            0.3.0    2018-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-124" data-line-number="124"><span class="co">##  withr                  2.2.0    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-125" data-line-number="125"><span class="co">##  xfun                   0.13     2020-04-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-126" data-line-number="126"><span class="co">##  xml2                   1.3.2    2020-04-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-127" data-line-number="127"><span class="co">##  XVector                0.28.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-128" data-line-number="128"><span class="co">##  yaml                   2.2.1    2020-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb21-129" data-line-number="129"><span class="co">##  zlibbioc               1.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb21-130" data-line-number="130"><span class="co">## </span></a>
<a class="sourceLine" id="cb21-131" data-line-number="131"><span class="co">## [1] D:/R_packages</span></a>
<a class="sourceLine" id="cb21-132" data-line-number="132"><span class="co">## [2] C:/Program Files/R/R-4.0.0/library</span></a></code></pre></div>
</details>
</div>
<div id="solutions" class="section level2">
<h2><span class="header-section-number">3.10</span> Solutions</h2>
<ol style="list-style-type: decimal">
<li>Construction of SCE and quality control</li>
<li>%&gt;% is the pipe operator from the dplyr package. It is very useful when the output of one function acts as an input for another, thus facilitating the chaining of our functions. It saves us from assigning new variables each time.</li>
<li>[0-9] means any digit from 0-9 and * means that the preceding expression is repeated for any number of times.</li>
<li>Depends. We want to make sure that all the clusterIDs in our dataset are mapped to a corresponding clustername, hence we use left_join. If there is an unmatched clusterID, it will return an NA. In this case, we know that all the clusterIDs in our dataset are present in the txt file, so any join method will work.</li>
<li>A cell is an outlier with respect to a metric if it is greater than 3 median absolute deviations <a href="https://en.wikipedia.org/wiki/Median_absolute_deviation">MAD</a></li>
</ol>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="downloading-data.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="clustering.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["NMps_book.pdf", "NMps_book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
