<!DOCTYPE html>
<html lang="" xml:lang="">
<head>

  <meta charset="utf-8" />
  <meta http-equiv="X-UA-Compatible" content="IE=edge" />
  <title>Chapter 7 Subclustering | NMps_book.utf8</title>
  <meta name="description" content="" />
  <meta name="generator" content="bookdown 0.19 and GitBook 2.6.7" />

  <meta property="og:title" content="Chapter 7 Subclustering | NMps_book.utf8" />
  <meta property="og:type" content="book" />
  
  
  
  

  <meta name="twitter:card" content="summary" />
  <meta name="twitter:title" content="Chapter 7 Subclustering | NMps_book.utf8" />
  
  
  




  <meta name="viewport" content="width=device-width, initial-scale=1" />
  <meta name="apple-mobile-web-app-capable" content="yes" />
  <meta name="apple-mobile-web-app-status-bar-style" content="black" />
  
  
<link rel="prev" href="differentially-expressed-genes.html"/>
<link rel="next" href="pseudotime.html"/>
<script src="libs/jquery-2.2.3/jquery.min.js"></script>
<link href="libs/gitbook-2.6.7/css/style.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-table.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-bookdown.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-highlight.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-search.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-fontsettings.css" rel="stylesheet" />
<link href="libs/gitbook-2.6.7/css/plugin-clipboard.css" rel="stylesheet" />











<style type="text/css">
a.sourceLine { display: inline-block; line-height: 1.25; }
a.sourceLine { pointer-events: none; color: inherit; text-decoration: inherit; }
a.sourceLine:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode { white-space: pre; position: relative; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
code.sourceCode { white-space: pre-wrap; }
a.sourceLine { text-indent: -1em; padding-left: 1em; }
}
pre.numberSource a.sourceLine
  { position: relative; left: -4em; }
pre.numberSource a.sourceLine::before
  { content: attr(data-line-number);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; pointer-events: all; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
    color: #aaaaaa;
  }
pre.numberSource { margin-left: 3em; border-left: 1px solid #aaaaaa;  padding-left: 4px; }
div.sourceCode
  {  }
@media screen {
a.sourceLine::before { text-decoration: underline; }
}
code span.al { color: #ff0000; font-weight: bold; } /* Alert */
code span.an { color: #60a0b0; font-weight: bold; font-style: italic; } /* Annotation */
code span.at { color: #7d9029; } /* Attribute */
code span.bn { color: #40a070; } /* BaseN */
code span.bu { } /* BuiltIn */
code span.cf { color: #007020; font-weight: bold; } /* ControlFlow */
code span.ch { color: #4070a0; } /* Char */
code span.cn { color: #880000; } /* Constant */
code span.co { color: #60a0b0; font-style: italic; } /* Comment */
code span.cv { color: #60a0b0; font-weight: bold; font-style: italic; } /* CommentVar */
code span.do { color: #ba2121; font-style: italic; } /* Documentation */
code span.dt { color: #902000; } /* DataType */
code span.dv { color: #40a070; } /* DecVal */
code span.er { color: #ff0000; font-weight: bold; } /* Error */
code span.ex { } /* Extension */
code span.fl { color: #40a070; } /* Float */
code span.fu { color: #06287e; } /* Function */
code span.im { } /* Import */
code span.in { color: #60a0b0; font-weight: bold; font-style: italic; } /* Information */
code span.kw { color: #007020; font-weight: bold; } /* Keyword */
code span.op { color: #666666; } /* Operator */
code span.ot { color: #007020; } /* Other */
code span.pp { color: #bc7a00; } /* Preprocessor */
code span.sc { color: #4070a0; } /* SpecialChar */
code span.ss { color: #bb6688; } /* SpecialString */
code span.st { color: #4070a0; } /* String */
code span.va { color: #19177c; } /* Variable */
code span.vs { color: #4070a0; } /* VerbatimString */
code span.wa { color: #60a0b0; font-weight: bold; font-style: italic; } /* Warning */
</style>

<link rel="stylesheet" href="style.css" type="text/css" />
</head>

<body>



  <div class="book without-animation with-summary font-size-2 font-family-1" data-basepath=".">

    <div class="book-summary">
      <nav role="navigation">

<ul class="summary">
<li><a href="./">A Minimal Book Example</a></li>

<li class="divider"></li>
<li class="chapter" data-level="1" data-path="index.html"><a href="index.html"><i class="fa fa-check"></i><b>1</b> Preface</a></li>
<li class="chapter" data-level="2" data-path="downloading-data.html"><a href="downloading-data.html"><i class="fa fa-check"></i><b>2</b> Downloading data</a></li>
<li class="chapter" data-level="3" data-path="quality-control.html"><a href="quality-control.html"><i class="fa fa-check"></i><b>3</b> Quality Control</a><ul>
<li class="chapter" data-level="3.1" data-path="quality-control.html"><a href="quality-control.html#overview"><i class="fa fa-check"></i><b>3.1</b> Overview</a></li>
<li class="chapter" data-level="3.2" data-path="quality-control.html"><a href="quality-control.html#importing-libraries"><i class="fa fa-check"></i><b>3.2</b> Importing libraries</a></li>
<li class="chapter" data-level="3.3" data-path="quality-control.html"><a href="quality-control.html#importing-the-data"><i class="fa fa-check"></i><b>3.3</b> Importing the data</a></li>
<li class="chapter" data-level="3.4" data-path="quality-control.html"><a href="quality-control.html#preparing-the-column-metadata"><i class="fa fa-check"></i><b>3.4</b> Preparing the column metadata</a></li>
<li class="chapter" data-level="3.5" data-path="quality-control.html"><a href="quality-control.html#preparing-the-row-metadata"><i class="fa fa-check"></i><b>3.5</b> Preparing the row metadata</a></li>
<li class="chapter" data-level="3.6" data-path="quality-control.html"><a href="quality-control.html#constructing-the-single-cell-experiment-object"><i class="fa fa-check"></i><b>3.6</b> Constructing the single cell experiment object</a></li>
<li class="chapter" data-level="3.7" data-path="quality-control.html"><a href="quality-control.html#quality-control-with-scater"><i class="fa fa-check"></i><b>3.7</b> Quality control with scater</a></li>
<li class="chapter" data-level="3.8" data-path="quality-control.html"><a href="quality-control.html#saving-the-rds-object"><i class="fa fa-check"></i><b>3.8</b> Saving the RDS object</a></li>
<li class="chapter" data-level="3.9" data-path="quality-control.html"><a href="quality-control.html#session-info"><i class="fa fa-check"></i><b>3.9</b> Session Info</a></li>
<li class="chapter" data-level="3.10" data-path="quality-control.html"><a href="quality-control.html#solutions"><i class="fa fa-check"></i><b>3.10</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="4" data-path="clustering.html"><a href="clustering.html"><i class="fa fa-check"></i><b>4</b> Clustering</a><ul>
<li class="chapter" data-level="4.1" data-path="clustering.html"><a href="clustering.html#introduction"><i class="fa fa-check"></i><b>4.1</b> Introduction</a></li>
<li class="chapter" data-level="4.2" data-path="clustering.html"><a href="clustering.html#importing-libraries-1"><i class="fa fa-check"></i><b>4.2</b> Importing libraries</a></li>
<li class="chapter" data-level="4.3" data-path="clustering.html"><a href="clustering.html#importing-data"><i class="fa fa-check"></i><b>4.3</b> Importing Data</a></li>
<li class="chapter" data-level="4.4" data-path="clustering.html"><a href="clustering.html#useful-seurat-object-interaction-methods"><i class="fa fa-check"></i><b>4.4</b> Useful seurat object interaction methods</a></li>
<li class="chapter" data-level="4.5" data-path="clustering.html"><a href="clustering.html#performing-dimensional-reduction"><i class="fa fa-check"></i><b>4.5</b> Performing dimensional reduction</a></li>
<li class="chapter" data-level="4.6" data-path="clustering.html"><a href="clustering.html#clustering-1"><i class="fa fa-check"></i><b>4.6</b> Clustering</a><ul>
<li class="chapter" data-level="4.6.1" data-path="clustering.html"><a href="clustering.html#constructing-a-shared-nearest-neighbour-snn-graph"><i class="fa fa-check"></i><b>4.6.1</b> Constructing a Shared-nearest neighbour (SNN) graph</a></li>
<li class="chapter" data-level="4.6.2" data-path="clustering.html"><a href="clustering.html#clustering-with-the-louvain-algorithm"><i class="fa fa-check"></i><b>4.6.2</b> Clustering with the Louvain algorithm</a></li>
<li class="chapter" data-level="4.6.3" data-path="clustering.html"><a href="clustering.html#resources"><i class="fa fa-check"></i><b>4.6.3</b> Resources</a></li>
</ul></li>
<li class="chapter" data-level="4.7" data-path="clustering.html"><a href="clustering.html#umap-with-authors-annotations"><i class="fa fa-check"></i><b>4.7</b> UMAP with authors’ annotations</a></li>
<li class="chapter" data-level="4.8" data-path="clustering.html"><a href="clustering.html#t-sne-with-authors-annotations"><i class="fa fa-check"></i><b>4.8</b> t-SNE with authors’ annotations</a></li>
<li class="chapter" data-level="4.9" data-path="clustering.html"><a href="clustering.html#assigning-our-own-cluster-identities"><i class="fa fa-check"></i><b>4.9</b> Assigning our own cluster identities</a></li>
<li class="chapter" data-level="4.10" data-path="clustering.html"><a href="clustering.html#umap-visualisation-of-our-annotations"><i class="fa fa-check"></i><b>4.10</b> UMAP visualisation of our annotations</a></li>
<li class="chapter" data-level="4.11" data-path="clustering.html"><a href="clustering.html#highlighting-tailbud-cells"><i class="fa fa-check"></i><b>4.11</b> Highlighting tailbud cells</a></li>
<li class="chapter" data-level="4.12" data-path="clustering.html"><a href="clustering.html#saving-rds-object"><i class="fa fa-check"></i><b>4.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="4.13" data-path="clustering.html"><a href="clustering.html#session-info-1"><i class="fa fa-check"></i><b>4.13</b> Session Info</a></li>
<li class="chapter" data-level="4.14" data-path="clustering.html"><a href="clustering.html#solutions-1"><i class="fa fa-check"></i><b>4.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="5" data-path="optimisations.html"><a href="optimisations.html"><i class="fa fa-check"></i><b>5</b> Optimisations</a><ul>
<li class="chapter" data-level="5.1" data-path="optimisations.html"><a href="optimisations.html#importing-libraries-2"><i class="fa fa-check"></i><b>5.1</b> Importing libraries</a></li>
<li class="chapter" data-level="5.2" data-path="optimisations.html"><a href="optimisations.html#importing-data-1"><i class="fa fa-check"></i><b>5.2</b> Importing Data</a></li>
<li class="chapter" data-level="5.3" data-path="optimisations.html"><a href="optimisations.html#number-of-pcs-to-retain"><i class="fa fa-check"></i><b>5.3</b> Number of PCs to retain</a><ul>
<li class="chapter" data-level="5.3.1" data-path="optimisations.html"><a href="optimisations.html#scree-plot"><i class="fa fa-check"></i><b>5.3.1</b> Scree plot</a></li>
<li class="chapter" data-level="5.3.2" data-path="optimisations.html"><a href="optimisations.html#examining-technical-noise"><i class="fa fa-check"></i><b>5.3.2</b> Examining technical noise</a></li>
</ul></li>
<li class="chapter" data-level="5.4" data-path="optimisations.html"><a href="optimisations.html#evaluating-clustering-results"><i class="fa fa-check"></i><b>5.4</b> Evaluating clustering results</a><ul>
<li class="chapter" data-level="5.4.1" data-path="optimisations.html"><a href="optimisations.html#cluster-stability"><i class="fa fa-check"></i><b>5.4.1</b> Cluster stability</a></li>
<li class="chapter" data-level="5.4.2" data-path="optimisations.html"><a href="optimisations.html#cluster-separability"><i class="fa fa-check"></i><b>5.4.2</b> Cluster separability</a></li>
</ul></li>
<li class="chapter" data-level="5.5" data-path="optimisations.html"><a href="optimisations.html#comparison-with-authors-labels"><i class="fa fa-check"></i><b>5.5</b> Comparison with authors’ labels</a><ul>
<li class="chapter" data-level="5.5.1" data-path="optimisations.html"><a href="optimisations.html#concordance-in-tailbud-cluster-assignment"><i class="fa fa-check"></i><b>5.5.1</b> Concordance in tailbud cluster assignment</a></li>
<li class="chapter" data-level="5.5.2" data-path="optimisations.html"><a href="optimisations.html#optimising-the-seurat-clustering-parameters"><i class="fa fa-check"></i><b>5.5.2</b> Optimising the seurat clustering parameters</a></li>
</ul></li>
<li class="chapter" data-level="5.6" data-path="optimisations.html"><a href="optimisations.html#selecting-cells-for-downstream-analysis"><i class="fa fa-check"></i><b>5.6</b> Selecting cells for downstream analysis</a></li>
<li class="chapter" data-level="5.7" data-path="optimisations.html"><a href="optimisations.html#session-info-2"><i class="fa fa-check"></i><b>5.7</b> Session Info</a></li>
</ul></li>
<li class="chapter" data-level="6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html"><i class="fa fa-check"></i><b>6</b> Differentially Expressed Genes</a><ul>
<li class="chapter" data-level="6.1" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualisation-of-marker-genes-in-our-dataset"><i class="fa fa-check"></i><b>6.1</b> Visualisation of marker genes in our dataset</a></li>
<li class="chapter" data-level="6.2" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#obtaining-a-list-of-differentially-expressed-genes"><i class="fa fa-check"></i><b>6.2</b> Obtaining a list of differentially expressed genes</a></li>
<li class="chapter" data-level="6.3" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#analysing-output-of-the-wilcoxon-test"><i class="fa fa-check"></i><b>6.3</b> Analysing output of the wilcoxon test</a></li>
<li class="chapter" data-level="6.4" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#visualising-the-expression-of-the-marker-genes"><i class="fa fa-check"></i><b>6.4</b> Visualising the expression of the marker genes</a></li>
<li class="chapter" data-level="6.5" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#session-info-3"><i class="fa fa-check"></i><b>6.5</b> Session Info</a></li>
<li class="chapter" data-level="6.6" data-path="differentially-expressed-genes.html"><a href="differentially-expressed-genes.html#solutions-2"><i class="fa fa-check"></i><b>6.6</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="7" data-path="subclustering.html"><a href="subclustering.html"><i class="fa fa-check"></i><b>7</b> Subclustering</a><ul>
<li class="chapter" data-level="7.1" data-path="subclustering.html"><a href="subclustering.html#importing-libraries-3"><i class="fa fa-check"></i><b>7.1</b> Importing libraries</a></li>
<li class="chapter" data-level="7.2" data-path="subclustering.html"><a href="subclustering.html#importing-data-2"><i class="fa fa-check"></i><b>7.2</b> Importing Data</a></li>
<li class="chapter" data-level="7.3" data-path="subclustering.html"><a href="subclustering.html#subclustering-the-tb-cells---using-only-cells-that-are-assigned-to-the-shared-cluster-across-all-the-clustering-resolutions."><i class="fa fa-check"></i><b>7.3</b> Subclustering the TB cells - Using only cells that are assigned to the ‘shared’ cluster across all the clustering resolutions.</a></li>
<li class="chapter" data-level="7.4" data-path="subclustering.html"><a href="subclustering.html#dimensional-reduction-with-umap-and-clustering"><i class="fa fa-check"></i><b>7.4</b> Dimensional reduction with UMAP and clustering</a></li>
<li class="chapter" data-level="7.5" data-path="subclustering.html"><a href="subclustering.html#identifying-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.5</b> Identifying marker genes for each cluster</a></li>
<li class="chapter" data-level="7.6" data-path="subclustering.html"><a href="subclustering.html#merging-clusters-that-do-not-appear-distinct"><i class="fa fa-check"></i><b>7.6</b> Merging clusters that do not appear distinct</a></li>
<li class="chapter" data-level="7.7" data-path="subclustering.html"><a href="subclustering.html#plotting-umap-and-assigning-cluster-labels"><i class="fa fa-check"></i><b>7.7</b> Plotting UMAP and assigning cluster labels</a></li>
<li class="chapter" data-level="7.8" data-path="subclustering.html"><a href="subclustering.html#checking-the-marker-genes-for-each-cluster"><i class="fa fa-check"></i><b>7.8</b> Checking the marker genes for each cluster</a></li>
<li class="chapter" data-level="7.9" data-path="subclustering.html"><a href="subclustering.html#sox2-tbxta-coexpression"><i class="fa fa-check"></i><b>7.9</b> Sox2 tbxta coexpression</a></li>
<li class="chapter" data-level="7.10" data-path="subclustering.html"><a href="subclustering.html#viewing-genes-of-interest-with-a-dotplot"><i class="fa fa-check"></i><b>7.10</b> Viewing genes of interest with a dotplot</a></li>
<li class="chapter" data-level="7.11" data-path="subclustering.html"><a href="subclustering.html#exporting-the-violin-plots-for-genes-of-interest."><i class="fa fa-check"></i><b>7.11</b> Exporting the violin plots for genes of interest.</a></li>
<li class="chapter" data-level="7.12" data-path="subclustering.html"><a href="subclustering.html#saving-rds-object-1"><i class="fa fa-check"></i><b>7.12</b> Saving RDS object</a></li>
<li class="chapter" data-level="7.13" data-path="subclustering.html"><a href="subclustering.html#session-info-4"><i class="fa fa-check"></i><b>7.13</b> Session Info</a></li>
<li class="chapter" data-level="7.14" data-path="subclustering.html"><a href="subclustering.html#solutions-3"><i class="fa fa-check"></i><b>7.14</b> Solutions</a></li>
</ul></li>
<li class="chapter" data-level="8" data-path="pseudotime.html"><a href="pseudotime.html"><i class="fa fa-check"></i><b>8</b> Pseudotime</a><ul>
<li class="chapter" data-level="8.1" data-path="pseudotime.html"><a href="pseudotime.html#importing-libraries-4"><i class="fa fa-check"></i><b>8.1</b> Importing libraries</a></li>
<li class="chapter" data-level="8.2" data-path="pseudotime.html"><a href="pseudotime.html#importing-data-3"><i class="fa fa-check"></i><b>8.2</b> Importing data</a></li>
<li class="chapter" data-level="8.3" data-path="pseudotime.html"><a href="pseudotime.html#tradeseq-workflow-analyses"><i class="fa fa-check"></i><b>8.3</b> TradeSeq workflow analyses</a></li>
<li class="chapter" data-level="8.4" data-path="pseudotime.html"><a href="pseudotime.html#session-info-5"><i class="fa fa-check"></i><b>8.4</b> Session Info</a></li>
<li class="chapter" data-level="8.5" data-path="pseudotime.html"><a href="pseudotime.html#solutions-4"><i class="fa fa-check"></i><b>8.5</b> Solutions</a></li>
</ul></li>
<li class="divider"></li>
<li><a href="https://github.com/rstudio/bookdown" target="blank">Published with bookdown</a></li>

</ul>

      </nav>
    </div>

    <div class="book-body">
      <div class="body-inner">
        <div class="book-header" role="navigation">
          <h1>
            <i class="fa fa-circle-o-notch fa-spin"></i><a href="./"></a>
          </h1>
        </div>

        <div class="page-wrapper" tabindex="-1" role="main">
          <div class="page-inner">

            <section class="normal" id="section-">
<div id="subclustering" class="section level1">
<h1><span class="header-section-number">Chapter 7</span> Subclustering</h1>
<div id="importing-libraries-3" class="section level2">
<h2><span class="header-section-number">7.1</span> Importing libraries</h2>
<div class="sourceCode" id="cb96"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb96-1" data-line-number="1"><span class="kw">library</span>(Seurat)</a>
<a class="sourceLine" id="cb96-2" data-line-number="2"><span class="kw">library</span>(tidyverse)</a>
<a class="sourceLine" id="cb96-3" data-line-number="3"><span class="kw">library</span>(gridExtra)</a>
<a class="sourceLine" id="cb96-4" data-line-number="4"><span class="kw">library</span>(RColorBrewer)</a></code></pre></div>
</div>
<div id="importing-data-2" class="section level2">
<h2><span class="header-section-number">7.2</span> Importing Data</h2>
<div class="sourceCode" id="cb97"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb97-1" data-line-number="1">hpf18_seurat &lt;-<span class="st"> </span><span class="kw">readRDS</span>(<span class="dt">file =</span> <span class="st">&quot;rds/hpf18_seurat_optimized551.rds&quot;</span> )</a></code></pre></div>
</div>
<div id="subclustering-the-tb-cells---using-only-cells-that-are-assigned-to-the-shared-cluster-across-all-the-clustering-resolutions." class="section level2">
<h2><span class="header-section-number">7.3</span> Subclustering the TB cells - Using only cells that are assigned to the ‘shared’ cluster across all the clustering resolutions.</h2>
<div class="sourceCode" id="cb98"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb98-1" data-line-number="1">TBcells &lt;-<span class="st"> </span><span class="kw">subset</span>(hpf18_seurat, <span class="dt">subset =</span> TB <span class="op">==</span><span class="st"> </span><span class="ot">TRUE</span>)</a>
<a class="sourceLine" id="cb98-2" data-line-number="2">TBcells</a></code></pre></div>
<pre><code>## An object of class Seurat 
## 60208 features across 551 samples within 2 assays 
## Active assay: SCT (29912 features, 3000 variable features)
##  1 other assay present: RNA
##  3 dimensional reductions calculated: pca, tsne, umap</code></pre>
</div>
<div id="dimensional-reduction-with-umap-and-clustering" class="section level2">
<h2><span class="header-section-number">7.4</span> Dimensional reduction with UMAP and clustering</h2>
<div class="sourceCode" id="cb100"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb100-1" data-line-number="1"><span class="co"># Running PCA</span></a>
<a class="sourceLine" id="cb100-2" data-line-number="2">TBcells &lt;-<span class="st"> </span><span class="kw">RunPCA</span>(TBcells, <span class="dt">npcs =</span> <span class="dv">50</span>)</a></code></pre></div>
<pre><code>## PC_ 1 
## Positive:  apoc1l, msgn1, fn1b, tbx24, tbx6, myf5, her1, apoeb, si:dkey-103d23.5, itm2cb 
##     tbx16, hoxd12a, her7, hm:zewp0073, actc1a, cdx4, LOC103910167, cxcl12b, tuba8l2, wu:fb77a09 
##     tcf15, efemp2b, hes6, meox1, phkg2, zgc:56585, foxc1a, cyp26a1, hoxb10a, kif26ab 
## Negative:  tmsb4x, id1, actb2, fosab, fabp3, krt8, rrm2, prtga, her15.1, junbb 
##     her2, wu:fb18c02, nova2, marcksl1b, pcna, zfp36l1a, LOC100534909, epcam, wu:fb25b09, elavl3 
##     pfn1, krt18, cfl1l, krt4, mdka, jun, sox3, tuba1a, atf3, her6 
## PC_ 2 
## Positive:  cdx4, hoxd12a, wnt8-2, wnt8a, hoxc13a, cyp26a1, hoxc13b, LOC103910167, kif26ab, sp5l 
##     LOC100001360, hoxb9a, scg3, eve1, her12, ldlrad2, cx43.4, ta, hes6, itm2cb 
##     fgf8a, tbx16, ptmab, hoxd9a, hoxa9b, angptl2b, hoxa13b, phkg2, hoxa10b, tbx6 
## Negative:  tbx24, hm:zewp0073, actc1a, efemp2b, hsp90aa1.1, apoc1l, meox1, tcf15, fn1b, myf5 
##     apoeb, LOC101886965, ripply1, wif1, ripply2, cpn1, LOC101886882, ntd5, kazald2, mespab 
##     LOC100535159, si:dkey-103d23.5, tuba8l2, net1, unc45b, rbm24a, aldh1a2, ppap2a, LOC100536992, dmrt2a 
## PC_ 3 
## Positive:  msgn1, tbx6, itm2cb, tbx16, cyp26a1, hoxc13b, phkg2, hes6, snai1a, wu:fb77a09 
##     rbm38, hoxc13a, fn1a, ved, mgat4b, actb2, her1, aplnrb, tmsb4x, tob1a 
##     palld, her7, cacna2d4a, wnt8-2, wnt8a, krt8, si:dkey-261h17.1, foxc1a, hoxa13a, ephb3 
## Negative:  ldlrad2, LOC100001360, scg3, LOC103910167, efemp2b, hoxb9a, wnt11r, hsp90aa1.1, her12, tcf15 
##     hoxc3a, si:dkey-103d23.5, meox1, LOC101886965, ppp2r2ca, zic2a, LOC103911885, LOC101882117, olig4, alcamb 
##     cdx4, LOC101886882, ripply1, col5a2a, bicc2, si:ch211-170d8.2, her3, hspb1, cdh6, hm:zewp0073 
## PC_ 4 
## Positive:  wnt8-2, wnt8a, fgf8a, hoxc13b, ta, LOC101886965, efemp2b, cyp26a1, LOC101886882, eve1 
##     hoxc13a, LOC100535159, cpn1, aldh1a2, tagln3b, thbs4b, kif26ab, hoxa13b, sp5l, kazald2 
##     ripply1, hsp90aa1.1, angptl7, myod1, ppap2a, LOC100536992, rdh10a, zgc:158328, apela, rprmb 
## Negative:  msgn1, tbx24, ism1, tbx6, her1, ldlrad2, foxc1a, her7, wnt11r, tbx16 
##     zgc:56585, ripply2, hoxb1b, apoeb, palld, lbx2, wif1, phc2a, si:dkey-103d23.5, LOC100334443 
##     foxc1b, LOC561719, igsf9, scg3, mgat4b, dact2, zic3, itm2cb, cacna2d4a, her2 
## PC_ 5 
## Positive:  her7, her1, ripply2, mespab, her12, apoeb, tbx24, hoxc13a, wif1, dlc 
##     her11, LOC101883520, mespaa, LOC557301, hoxc13b, fgf8a, ta, hoxb7a, eve1, dld 
##     hoxa13b, sp5l, hoxa13a, ntd5, fkbp7, hoxc3a, tagln3b, wnt8a, wnt8-2, net1 
## Negative:  si:dkey-103d23.5, ephb3, mgat4b, fn1b, cacna2d4a, LOC103910167, LOC100006216, itm2cb, tbx6, angptl7 
##     LOC103911885, si:ch211-168h21.3, ppp2r2ca, LOC100537766, thbs4b, LOC103908989, tbx16, foxc1a, zgc:123035, phkg2 
##     LOC101886965, cpn1, cdx1a, msgn1, crabp2b, rdh10a, zgc:162707, her15.1, uncx4.1, tgfbi</code></pre>
<div class="sourceCode" id="cb102"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb102-1" data-line-number="1"><span class="co"># Elbow plot to determine number of PCs for dimension reduction</span></a>
<a class="sourceLine" id="cb102-2" data-line-number="2"></a>
<a class="sourceLine" id="cb102-3" data-line-number="3">ep &lt;-<span class="st"> </span><span class="kw">ElbowPlot</span>(TBcells, <span class="dt">ndims =</span> <span class="dv">50</span>, <span class="dt">reduction =</span> <span class="st">&quot;pca&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb102-4" data-line-number="4"><span class="st">  </span><span class="kw">geom_vline</span>(<span class="dt">xintercept =</span> <span class="dv">25</span>,  <span class="dt">color =</span> <span class="st">&quot;red&quot;</span>, <span class="dt">size=</span><span class="fl">1.5</span>, <span class="dt">linetype=</span><span class="st">&quot;dotted&quot;</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb102-5" data-line-number="5"><span class="st">  </span><span class="kw">ylab</span>(<span class="dt">label =</span> <span class="st">&quot;Percentage of variance explained&quot;</span>) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb102-6" data-line-number="6"><span class="st">  </span><span class="kw">theme_bw</span>() <span class="op">+</span><span class="st"> </span><span class="kw">theme</span>(<span class="dt">axis.line =</span> <span class="kw">element_line</span>(<span class="dt">colour =</span> <span class="st">&quot;black&quot;</span>),</a>
<a class="sourceLine" id="cb102-7" data-line-number="7">    <span class="dt">panel.grid.major =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb102-8" data-line-number="8">    <span class="dt">panel.grid.minor =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb102-9" data-line-number="9">    <span class="dt">panel.border =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb102-10" data-line-number="10">    <span class="dt">panel.background =</span> <span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb102-11" data-line-number="11">    <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>))</a>
<a class="sourceLine" id="cb102-12" data-line-number="12"></a>
<a class="sourceLine" id="cb102-13" data-line-number="13"><span class="co"># UMAP embedding and clustering</span></a>
<a class="sourceLine" id="cb102-14" data-line-number="14">ep</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/unnamed-chunk-28-1.png" width="672" /></p>
<div class="sourceCode" id="cb103"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb103-1" data-line-number="1">TBcells &lt;-<span class="st"> </span><span class="kw">RunUMAP</span>(TBcells, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">25</span>,  <span class="dt">min.dist =</span> <span class="fl">0.7</span>, <span class="dt">n.neighbors =</span> 30L, <span class="dt">verbose =</span> <span class="ot">TRUE</span>)</a></code></pre></div>
<pre><code>## 17:42:38 UMAP embedding parameters a = 0.3208 b = 1.563</code></pre>
<pre><code>## 17:42:38 Read 551 rows and found 25 numeric columns</code></pre>
<pre><code>## 17:42:38 Using Annoy for neighbor search, n_neighbors = 30</code></pre>
<pre><code>## 17:42:38 Building Annoy index with metric = cosine, n_trees = 50</code></pre>
<pre><code>## 0%   10   20   30   40   50   60   70   80   90   100%</code></pre>
<pre><code>## [----|----|----|----|----|----|----|----|----|----|</code></pre>
<pre><code>## **************************************************|
## 17:42:38 Writing NN index file to temp file C:\Users\User\AppData\Local\Temp\RtmpUv6AgP\file439ce965a11
## 17:42:38 Searching Annoy index using 1 thread, search_k = 3000
## 17:42:38 Annoy recall = 100%
## 17:42:39 Commencing smooth kNN distance calibration using 1 thread
## 17:42:40 Initializing from normalized Laplacian + noise
## 17:42:40 Commencing optimization for 500 epochs, with 20478 positive edges
## 17:42:41 Optimization finished</code></pre>
<div class="sourceCode" id="cb111"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb111-1" data-line-number="1">TBcells &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">FindNeighbors</span>(TBcells, <span class="dt">dims =</span> <span class="dv">1</span><span class="op">:</span><span class="dv">25</span>, <span class="dt">k.param =</span> <span class="dv">20</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>) </a>
<a class="sourceLine" id="cb111-2" data-line-number="2">TBcells &lt;-<span class="st"> </span>Seurat<span class="op">::</span><span class="kw">FindClusters</span>(TBcells, <span class="dt">algorithm =</span> <span class="dv">1</span>, <span class="dt">resolution =</span> <span class="dv">1</span>, <span class="dt">verbose =</span> <span class="ot">FALSE</span>)</a></code></pre></div>
</div>
<div id="identifying-marker-genes-for-each-cluster" class="section level2">
<h2><span class="header-section-number">7.5</span> Identifying marker genes for each cluster</h2>
<div class="sourceCode" id="cb112"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb112-1" data-line-number="1"><span class="co"># Setting identity for cells to seurat clustering results</span></a>
<a class="sourceLine" id="cb112-2" data-line-number="2"></a>
<a class="sourceLine" id="cb112-3" data-line-number="3">TBcells&lt;-<span class="st"> </span><span class="kw">SetIdent</span>(TBcells, <span class="dt">value =</span> TBcells<span class="op">@</span>meta.data<span class="op">$</span>seurat_clusters) </a>
<a class="sourceLine" id="cb112-4" data-line-number="4"></a>
<a class="sourceLine" id="cb112-5" data-line-number="5">all_clusters &lt;-<span class="st"> </span><span class="kw">FindAllMarkers</span>(TBcells, <span class="dt">test.use =</span> <span class="st">&quot;wilcox&quot;</span>, <span class="dt">only.pos =</span> <span class="ot">TRUE</span>) <span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb112-6" data-line-number="6"><span class="st">  </span><span class="kw">group_by</span>(cluster)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb112-7" data-line-number="7"><span class="st">  </span><span class="kw">filter</span>(p_val_adj <span class="op">&lt;</span><span class="st"> </span><span class="fl">0.05</span>)<span class="op">%&gt;%</span></a>
<a class="sourceLine" id="cb112-8" data-line-number="8"><span class="st">  </span><span class="kw">arrange</span>(cluster, <span class="kw">desc</span>(<span class="kw">abs</span>(avg_logFC)),p_val_adj)</a></code></pre></div>
<pre><code>## Calculating cluster 0</code></pre>
<pre><code>## Calculating cluster 1</code></pre>
<pre><code>## Calculating cluster 2</code></pre>
<pre><code>## Calculating cluster 3</code></pre>
<pre><code>## Calculating cluster 4</code></pre>
<pre><code>## Calculating cluster 5</code></pre>
<pre><code>## Calculating cluster 6</code></pre>
<pre><code>## Calculating cluster 7</code></pre>
<div class="sourceCode" id="cb121"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb121-1" data-line-number="1"><span class="co">#write.csv(all_clusters,file = &quot;Files/2g)MarkerGenes_9Clusters.csv&quot;)</span></a>
<a class="sourceLine" id="cb121-2" data-line-number="2"></a>
<a class="sourceLine" id="cb121-3" data-line-number="3"><span class="co"># Viewing number of cells in each cluster</span></a>
<a class="sourceLine" id="cb121-4" data-line-number="4"></a>
<a class="sourceLine" id="cb121-5" data-line-number="5"><span class="kw">table</span>(TBcells<span class="op">@</span>meta.data<span class="op">$</span>seurat_clusters)</a></code></pre></div>
<pre><code>## 
##   0   1   2   3   4   5   6   7 
## 103  90  76  73  62  50  50  47</code></pre>
<div class="sourceCode" id="cb123"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb123-1" data-line-number="1"><span class="co"># Plot a heatmap to visualize the top 7 marker genes in each cluster. </span></a>
<a class="sourceLine" id="cb123-2" data-line-number="2"></a>
<a class="sourceLine" id="cb123-3" data-line-number="3">top7_wilcox &lt;-<span class="st"> </span>all_clusters <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">group_by</span>(cluster) <span class="op">%&gt;%</span><span class="st"> </span><span class="kw">top_n</span>(<span class="dv">7</span>, avg_logFC)</a>
<a class="sourceLine" id="cb123-4" data-line-number="4"><span class="kw">DoHeatmap</span>(<span class="dt">object =</span> TBcells, <span class="dt">features =</span> <span class="kw">as.character</span>(top7_wilcox<span class="op">$</span>gene), <span class="dt">label =</span> <span class="ot">FALSE</span>) <span class="op">+</span></a>
<a class="sourceLine" id="cb123-5" data-line-number="5"><span class="st">   </span><span class="kw">theme</span>(<span class="dt">legend.title =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">25</span>),</a>
<a class="sourceLine" id="cb123-6" data-line-number="6">         <span class="dt">axis.line=</span><span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb123-7" data-line-number="7">      <span class="dt">axis.text.x=</span><span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb123-8" data-line-number="8">      <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">15</span>),</a>
<a class="sourceLine" id="cb123-9" data-line-number="9">      <span class="dt">axis.ticks=</span><span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb123-10" data-line-number="10">      <span class="dt">axis.title.x=</span><span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb123-11" data-line-number="11">      <span class="dt">axis.title.y=</span><span class="kw">element_blank</span>(),</a>
<a class="sourceLine" id="cb123-12" data-line-number="12">      <span class="dt">legend.text=</span><span class="kw">element_text</span>(<span class="dt">face=</span><span class="st">&quot;bold&quot;</span>, <span class="dt">size =</span> <span class="dv">25</span>), </a>
<a class="sourceLine" id="cb123-13" data-line-number="13">      <span class="dt">text=</span><span class="kw">element_text</span>(<span class="dt">family=</span><span class="st">&quot;Arial&quot;</span>))  </a></code></pre></div>
<p><img src="NMps_book_files/figure-html/subclusters_DEG-1.png" width="1152" /></p>
<div class="sourceCode" id="cb124"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb124-1" data-line-number="1"><span class="co">#ggsave(&quot;2g)AllClusters_markerGeneHeatmap.tiff&quot;, hm, dpi = 500, width = 17, height = 17,units = c(&quot;cm&quot;))</span></a></code></pre></div>
<div class="sourceCode" id="cb125"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb125-1" data-line-number="1">top7_wilcox</a></code></pre></div>
<pre><code>## # A tibble: 56 x 7
## # Groups:   cluster [8]
##       p_val avg_logFC pct.1 pct.2 p_val_adj cluster gene       
##       &lt;dbl&gt;     &lt;dbl&gt; &lt;dbl&gt; &lt;dbl&gt;     &lt;dbl&gt; &lt;fct&gt;   &lt;chr&gt;      
##  1 5.41e-54     1.55  0.796 0.118  1.62e-49 0       gfap       
##  2 8.71e-29     1.44  0.689 0.201  2.60e-24 0       slc1a3a    
##  3 1.69e-31     1.41  0.893 0.44   5.06e-27 0       prtga      
##  4 5.01e-44     1.30  0.816 0.181  1.50e-39 0       wnt11r     
##  5 3.60e-42     1.28  0.903 0.268  1.08e-37 0       nova2      
##  6 3.88e-42     1.27  0.854 0.219  1.16e-37 0       sox3       
##  7 7.61e-22     1.13  0.932 0.766  2.28e-17 0       mdka       
##  8 6.98e-16     1.13  0.922 0.482  2.09e-11 1       actc1a     
##  9 8.80e-16     1.12  0.922 0.482  2.63e-11 1       hm:zewp0073
## 10 7.98e-26     0.954 1     0.62   2.39e-21 1       fn1b       
## # ... with 46 more rows</code></pre>
</div>
<div id="merging-clusters-that-do-not-appear-distinct" class="section level2">
<h2><span class="header-section-number">7.6</span> Merging clusters that do not appear distinct</h2>
<div class="sourceCode" id="cb127"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb127-1" data-line-number="1"><span class="co"># Merging the two mesodermal clusters 1 and 4</span></a>
<a class="sourceLine" id="cb127-2" data-line-number="2"></a>
<a class="sourceLine" id="cb127-3" data-line-number="3">meso &lt;-<span class="st"> </span><span class="kw">subset</span>(<span class="dt">x =</span> TBcells, <span class="dt">subset =</span> (seurat_clusters <span class="op">==</span><span class="st"> &quot;1&quot;</span> <span class="op">|</span><span class="st"> </span>seurat_clusters <span class="op">==</span><span class="st"> &quot;5&quot;</span>))</a>
<a class="sourceLine" id="cb127-4" data-line-number="4">meso_ids &lt;-<span class="st"> </span><span class="kw">rownames</span>(meso<span class="op">@</span>meta.data)</a>
<a class="sourceLine" id="cb127-5" data-line-number="5">TBcells<span class="op">@</span>meta.data<span class="op">$</span>seurat_clusters[<span class="kw">rownames</span>(TBcells<span class="op">@</span>meta.data) <span class="op">%in%</span><span class="st"> </span>meso_ids] &lt;-<span class="st"> </span><span class="dv">1</span></a></code></pre></div>
</div>
<div id="plotting-umap-and-assigning-cluster-labels" class="section level2">
<h2><span class="header-section-number">7.7</span> Plotting UMAP and assigning cluster labels</h2>
<div class="sourceCode" id="cb128"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb128-1" data-line-number="1">labelList &lt;-<span class="st"> </span><span class="kw">data.frame</span>(<span class="dt">clusterID =</span> <span class="kw">c</span>(<span class="dv">0</span>,<span class="dv">1</span>,<span class="dv">2</span>,<span class="dv">3</span>,<span class="dv">4</span>,<span class="dv">6</span>,<span class="dv">7</span>), </a>
<a class="sourceLine" id="cb128-2" data-line-number="2">                        <span class="dt">clusterNames =</span> <span class="kw">c</span>(<span class="st">&quot;Posterior NT&quot;</span>, </a>
<a class="sourceLine" id="cb128-3" data-line-number="3">                                         <span class="st">&quot;Posterior PSM&quot;</span>,  </a>
<a class="sourceLine" id="cb128-4" data-line-number="4">                                        <span class="st">&quot;Nascent somites&quot;</span>,</a>
<a class="sourceLine" id="cb128-5" data-line-number="5">                                        <span class="st">&quot;Pou5f3+ Posterior NT&quot;</span>,</a>
<a class="sourceLine" id="cb128-6" data-line-number="6">                                        <span class="st">&quot;Maturation zone&quot;</span>, </a>
<a class="sourceLine" id="cb128-7" data-line-number="7">                                        <span class="st">&quot;Tail somites&quot;</span>, </a>
<a class="sourceLine" id="cb128-8" data-line-number="8">                                        <span class="st">&quot;NMps&quot;</span>))</a>
<a class="sourceLine" id="cb128-9" data-line-number="9"></a>
<a class="sourceLine" id="cb128-10" data-line-number="10">TBcells<span class="op">$</span>subclusters_names &lt;-<span class="st"> </span>labelList<span class="op">$</span>clusterNames[<span class="kw">match</span>(TBcells<span class="op">$</span>seurat_clusters, labelList<span class="op">$</span>clusterID)]</a>
<a class="sourceLine" id="cb128-11" data-line-number="11">TBcells<span class="op">$</span>subclusters_names &lt;-<span class="st"> </span><span class="kw">factor</span>(TBcells<span class="op">$</span>subclusters_names)</a>
<a class="sourceLine" id="cb128-12" data-line-number="12">TBcells&lt;-<span class="st"> </span><span class="kw">SetIdent</span>(TBcells, <span class="dt">value =</span> TBcells<span class="op">@</span>meta.data<span class="op">$</span>subclusters_names)</a></code></pre></div>
</div>
<div id="checking-the-marker-genes-for-each-cluster" class="section level2">
<h2><span class="header-section-number">7.8</span> Checking the marker genes for each cluster</h2>
<div class="sourceCode" id="cb129"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb129-1" data-line-number="1">marker_genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;sox2&quot;</span>, <span class="st">&quot;ta&quot;</span>, <span class="st">&quot;pou5f3&quot;</span>, <span class="st">&quot;olig4&quot;</span>, <span class="st">&quot;zic5&quot;</span>,<span class="st">&quot;tbx6&quot;</span>, <span class="st">&quot;tbx16&quot;</span>,<span class="st">&quot;msgn1&quot;</span>,<span class="st">&quot;meox1&quot;</span>, <span class="st">&quot;mespab&quot;</span>, <span class="st">&quot;cyp26a1&quot;</span>, <span class="st">&quot;ripply1&quot;</span>)</a>
<a class="sourceLine" id="cb129-2" data-line-number="2"><span class="kw">VlnPlot</span>(<span class="dt">object =</span> TBcells, <span class="dt">features =</span> marker_genes, <span class="dt">ncol =</span> <span class="dv">3</span>)</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/check_subclusters_violin-1.png" width="768" /></p>
<div class="sourceCode" id="cb130"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb130-1" data-line-number="1"><span class="co"># Plotting UMAP dimplot with cluster labels</span></a>
<a class="sourceLine" id="cb130-2" data-line-number="2"></a>
<a class="sourceLine" id="cb130-3" data-line-number="3">cols &lt;-<span class="st"> </span><span class="kw">c</span>(scales<span class="op">::</span><span class="kw">hue_pal</span>()(<span class="dv">7</span>), <span class="st">&quot;#FFFFFF&quot;</span>)</a>
<a class="sourceLine" id="cb130-4" data-line-number="4"></a>
<a class="sourceLine" id="cb130-5" data-line-number="5"><span class="kw">DimPlot</span>(TBcells, <span class="dt">label =</span> <span class="ot">TRUE</span>, <span class="dt">label.size =</span> <span class="dv">5</span>, <span class="dt">pt.size=</span><span class="fl">0.6</span>)<span class="op">+</span></a>
<a class="sourceLine" id="cb130-6" data-line-number="6"><span class="st">  </span><span class="kw">scale_color_manual</span>(<span class="dt">values =</span> cols, <span class="dt">name =</span> <span class="st">&quot;Cluster Identities&quot;</span>)<span class="op">+</span></a>
<a class="sourceLine" id="cb130-7" data-line-number="7"><span class="st">  </span>mytheme</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/UMAP_subclusters-1.png" width="768" /></p>
<div class="sourceCode" id="cb131"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb131-1" data-line-number="1"><span class="co">#ggsave(&quot;2g)TBsubset.tiff&quot;, tbsubset_plt, dpi = 500, width = 17, height = 13,units = c(&quot;cm&quot;))</span></a>
<a class="sourceLine" id="cb131-2" data-line-number="2"><span class="co">#saveRDS(TBcells, file = &quot;rds/2g)TBcells_24.10.19.rds&quot;)</span></a></code></pre></div>
</div>
<div id="sox2-tbxta-coexpression" class="section level2">
<h2><span class="header-section-number">7.9</span> Sox2 tbxta coexpression</h2>
<div class="sourceCode" id="cb132"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb132-1" data-line-number="1"><span class="co"># Plotting co-expression of sox2 and ta in TBcells</span></a>
<a class="sourceLine" id="cb132-2" data-line-number="2"></a>
<a class="sourceLine" id="cb132-3" data-line-number="3"><span class="kw">FeaturePlot</span>(TBcells, <span class="dt">features =</span> <span class="kw">c</span>(<span class="st">&quot;sox2&quot;</span>, <span class="st">&quot;ta&quot;</span>), <span class="dt">blend =</span> <span class="ot">TRUE</span>, <span class="dt">pt.size =</span> <span class="dv">1</span>, <span class="dt">blend.threshold =</span> <span class="fl">0.5</span>)</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/sox2ta_coexp-1.png" width="672" /></p>
<div class="sourceCode" id="cb133"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb133-1" data-line-number="1"><span class="co">#ggsave(&quot;2g)sox2_ta_coexp.tiff&quot;, coexp, dpi = 400, width = 12, height = 6,units = c(&quot;cm&quot;))</span></a></code></pre></div>
</div>
<div id="viewing-genes-of-interest-with-a-dotplot" class="section level2">
<h2><span class="header-section-number">7.10</span> Viewing genes of interest with a dotplot</h2>
<div class="sourceCode" id="cb134"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb134-1" data-line-number="1"><span class="co"># Looking at only cells in the NM decision</span></a>
<a class="sourceLine" id="cb134-2" data-line-number="2"></a>
<a class="sourceLine" id="cb134-3" data-line-number="3">TBcells_nmdecision &lt;-<span class="st"> </span><span class="kw">subset</span>(TBcells, <span class="dt">subset =</span> subclusters_names <span class="op">==</span><span class="st"> &quot;Neuromesodermal Progenitors&quot;</span> <span class="op">|</span></a>
<a class="sourceLine" id="cb134-4" data-line-number="4"><span class="st">                              </span>subclusters_names <span class="op">==</span><span class="st"> &quot;Maturation zone&quot;</span> <span class="op">|</span><span class="st"> </span></a>
<a class="sourceLine" id="cb134-5" data-line-number="5"><span class="st">                              </span>subclusters_names <span class="op">==</span><span class="st"> &quot;Pou5f3+ Posterior NT&quot;</span> <span class="op">|</span></a>
<a class="sourceLine" id="cb134-6" data-line-number="6"><span class="st">                              </span>subclusters_names <span class="op">==</span><span class="st"> &quot;Posterior NT&quot;</span>)</a>
<a class="sourceLine" id="cb134-7" data-line-number="7"><span class="co"># Specifying genes to visualize</span></a>
<a class="sourceLine" id="cb134-8" data-line-number="8">nm_genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;cyp26a1&quot;</span>,<span class="st">&quot;eve1&quot;</span>,<span class="st">&quot;fgf8a&quot;</span>,<span class="st">&quot;hoxa13b&quot;</span>,<span class="st">&quot;hoxc13a&quot;</span>,<span class="st">&quot;hoxc13b&quot;</span>,<span class="st">&quot;sox2&quot;</span>,<span class="st">&quot;ta&quot;</span>,<span class="st">&quot;wnt8a&quot;</span>)</a>
<a class="sourceLine" id="cb134-9" data-line-number="9">meso_genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;aldh1a2&quot;</span>, <span class="st">&quot;apoc1l&quot;</span>,<span class="st">&quot;bmp10&quot;</span>, <span class="st">&quot;her1&quot;</span>, <span class="st">&quot;her7&quot;</span>, <span class="st">&quot;msgn1&quot;</span>,<span class="st">&quot;snai1a&quot;</span>,<span class="st">&quot;tbx6&quot;</span>,<span class="st">&quot;tbx16&quot;</span>,<span class="st">&quot;tbx24&quot;</span>,<span class="st">&quot;uncx4.1&quot;</span>, <span class="st">&quot;meox1&quot;</span>, <span class="st">&quot;mespab&quot;</span>)</a>
<a class="sourceLine" id="cb134-10" data-line-number="10">neural_genes &lt;-<span class="st"> </span><span class="kw">c</span>(<span class="st">&quot;gfap&quot;</span>, <span class="st">&quot;sox3&quot;</span>, <span class="st">&quot;prtga&quot;</span>, <span class="st">&quot;zic2a&quot;</span>, <span class="st">&quot;zic5&quot;</span>)</a></code></pre></div>
<div class="sourceCode" id="cb135"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb135-1" data-line-number="1"><span class="kw">DotPlot</span>(TBcells, <span class="dt">features =</span> nm_genes) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb135-2" data-line-number="2"><span class="st">    </span><span class="kw">theme</span>(<span class="dt">text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>),</a>
<a class="sourceLine" id="cb135-3" data-line-number="3">          <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>),</a>
<a class="sourceLine" id="cb135-4" data-line-number="4">          <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>, <span class="dt">angle =</span> <span class="dv">35</span>))</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/nm_dotplot-1.png" width="672" /></p>
<div class="sourceCode" id="cb136"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb136-1" data-line-number="1"><span class="kw">DotPlot</span>(TBcells, <span class="dt">features =</span> meso_genes) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb136-2" data-line-number="2"><span class="st">    </span><span class="kw">theme</span>(<span class="dt">text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>),</a>
<a class="sourceLine" id="cb136-3" data-line-number="3">          <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>),</a>
<a class="sourceLine" id="cb136-4" data-line-number="4">          <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>, <span class="dt">angle =</span> <span class="dv">35</span>))</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/meso_dotplot-1.png" width="672" /></p>
<div class="sourceCode" id="cb137"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb137-1" data-line-number="1"><span class="kw">DotPlot</span>(TBcells, <span class="dt">features =</span> neural_genes) <span class="op">+</span><span class="st"> </span></a>
<a class="sourceLine" id="cb137-2" data-line-number="2"><span class="st">    </span><span class="kw">theme</span>(<span class="dt">text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>),</a>
<a class="sourceLine" id="cb137-3" data-line-number="3">          <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>),</a>
<a class="sourceLine" id="cb137-4" data-line-number="4">          <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">8</span>, <span class="dt">angle =</span> <span class="dv">35</span>))</a></code></pre></div>
<p><img src="NMps_book_files/figure-html/neural_dotplot-1.png" width="672" /></p>
<p><em>Q1) How would you save the images in the working directory?</em></p>
</div>
<div id="exporting-the-violin-plots-for-genes-of-interest." class="section level2">
<h2><span class="header-section-number">7.11</span> Exporting the violin plots for genes of interest.</h2>
<p>We loop over each violin plot using the <code>lapply</code> function, which merits additional comment.The apply family of functions is very useful to perform an action repeatedly on a collection of objects such as a dataframe, vector, list etc. Normally iteration is done with a for loop, which is rather time-consuming. The apply function performs the iteration without the explicit use of a for loop. <code>lapply</code> returns a list.</p>
<p><em>Q2) What is the difference between lapply and sapply?</em></p>
<div class="sourceCode" id="cb138"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb138-1" data-line-number="1"><span class="co"># Looking at expression of nmp markers over all clusters in individual violinPlots</span></a>
<a class="sourceLine" id="cb138-2" data-line-number="2">nmp_markers &lt;-<span class="st"> </span><span class="kw">c</span>(</a>
<a class="sourceLine" id="cb138-3" data-line-number="3"><span class="st">&quot;fgf8a&quot;</span>,</a>
<a class="sourceLine" id="cb138-4" data-line-number="4"><span class="st">&quot;wnt8-2&quot;</span>,</a>
<a class="sourceLine" id="cb138-5" data-line-number="5"><span class="st">&quot;wnt8a&quot;</span>,</a>
<a class="sourceLine" id="cb138-6" data-line-number="6"><span class="st">&quot;ta&quot;</span>,</a>
<a class="sourceLine" id="cb138-7" data-line-number="7"><span class="st">&quot;apela&quot;</span>,</a>
<a class="sourceLine" id="cb138-8" data-line-number="8"><span class="st">&quot;eve1&quot;</span>,</a>
<a class="sourceLine" id="cb138-9" data-line-number="9"><span class="st">&quot;ndnf&quot;</span>,</a>
<a class="sourceLine" id="cb138-10" data-line-number="10"><span class="st">&quot;cyp26a1&quot;</span>,</a>
<a class="sourceLine" id="cb138-11" data-line-number="11"><span class="st">&quot;hoxc13b&quot;</span>,</a>
<a class="sourceLine" id="cb138-12" data-line-number="12"><span class="st">&quot;id3&quot;</span>,</a>
<a class="sourceLine" id="cb138-13" data-line-number="13"><span class="st">&quot;hoxc13a&quot;</span>,</a>
<a class="sourceLine" id="cb138-14" data-line-number="14"><span class="st">&quot;thbs2&quot;</span>,</a>
<a class="sourceLine" id="cb138-15" data-line-number="15"><span class="st">&quot;sp5l&quot;</span>,</a>
<a class="sourceLine" id="cb138-16" data-line-number="16"><span class="st">&quot;tagln3b&quot;</span>,</a>
<a class="sourceLine" id="cb138-17" data-line-number="17"><span class="st">&quot;mcamb&quot;</span>,</a>
<a class="sourceLine" id="cb138-18" data-line-number="18"><span class="st">&quot;her12&quot;</span>,</a>
<a class="sourceLine" id="cb138-19" data-line-number="19"><span class="st">&quot;hoxa13b&quot;</span>,</a>
<a class="sourceLine" id="cb138-20" data-line-number="20"><span class="st">&quot;hoxd12a&quot;</span>,</a>
<a class="sourceLine" id="cb138-21" data-line-number="21"><span class="st">&quot;kif26ab&quot;</span>,</a>
<a class="sourceLine" id="cb138-22" data-line-number="22"><span class="st">&quot;sox2&quot;</span>,</a>
<a class="sourceLine" id="cb138-23" data-line-number="23"><span class="st">&quot;prickle1b&quot;</span>,</a>
<a class="sourceLine" id="cb138-24" data-line-number="24"><span class="st">&quot;si:dkey-261h17.1&quot;</span>,</a>
<a class="sourceLine" id="cb138-25" data-line-number="25"><span class="st">&quot;cdh6&quot;</span>,</a>
<a class="sourceLine" id="cb138-26" data-line-number="26"><span class="st">&quot;cx43.4&quot;</span>,</a>
<a class="sourceLine" id="cb138-27" data-line-number="27"><span class="st">&quot;zgc:158328&quot;</span>,</a>
<a class="sourceLine" id="cb138-28" data-line-number="28"><span class="st">&quot;fndc3ba&quot;</span>,</a>
<a class="sourceLine" id="cb138-29" data-line-number="29"><span class="st">&quot;mnx1&quot;</span>,</a>
<a class="sourceLine" id="cb138-30" data-line-number="30"><span class="st">&quot;hes6&quot;</span>,</a>
<a class="sourceLine" id="cb138-31" data-line-number="31"><span class="st">&quot;znf703&quot;</span>,</a>
<a class="sourceLine" id="cb138-32" data-line-number="32"><span class="st">&quot;ube2e2&quot;</span>,</a>
<a class="sourceLine" id="cb138-33" data-line-number="33"><span class="st">&quot;angptl2b&quot;</span>,</a>
<a class="sourceLine" id="cb138-34" data-line-number="34"><span class="st">&quot;depdc7&quot;</span>,</a>
<a class="sourceLine" id="cb138-35" data-line-number="35"><span class="st">&quot;efhd2&quot;</span>,</a>
<a class="sourceLine" id="cb138-36" data-line-number="36"><span class="st">&quot;kdm6a&quot;</span>,</a>
<a class="sourceLine" id="cb138-37" data-line-number="37"><span class="st">&quot;nradd&quot;</span>,</a>
<a class="sourceLine" id="cb138-38" data-line-number="38"><span class="st">&quot;enc1&quot;</span>,</a>
<a class="sourceLine" id="cb138-39" data-line-number="39"><span class="st">&quot;snai1a&quot;</span>,</a>
<a class="sourceLine" id="cb138-40" data-line-number="40"><span class="st">&quot;arf1&quot;</span>,</a>
<a class="sourceLine" id="cb138-41" data-line-number="41"><span class="st">&quot;flrt3&quot;</span>,</a>
<a class="sourceLine" id="cb138-42" data-line-number="42"><span class="st">&quot;phgdh&quot;</span>,</a>
<a class="sourceLine" id="cb138-43" data-line-number="43"><span class="st">&quot;traf4a&quot;</span>,</a>
<a class="sourceLine" id="cb138-44" data-line-number="44"><span class="st">&quot;wls&quot;</span>,</a>
<a class="sourceLine" id="cb138-45" data-line-number="45"><span class="st">&quot;sepn1&quot;</span>,</a>
<a class="sourceLine" id="cb138-46" data-line-number="46"><span class="st">&quot;esrrga&quot;</span>,</a>
<a class="sourceLine" id="cb138-47" data-line-number="47"><span class="st">&quot;ptmab&quot;</span>,</a>
<a class="sourceLine" id="cb138-48" data-line-number="48"><span class="st">&quot;ptbp1b&quot;</span>,</a>
<a class="sourceLine" id="cb138-49" data-line-number="49"><span class="st">&quot;ptmaa&quot;</span>,</a>
<a class="sourceLine" id="cb138-50" data-line-number="50"><span class="st">&quot;mgat1b&quot;</span>,</a>
<a class="sourceLine" id="cb138-51" data-line-number="51"><span class="st">&quot;ubl3a&quot;</span>,</a>
<a class="sourceLine" id="cb138-52" data-line-number="52"><span class="st">&quot;sall4&quot;</span>,</a>
<a class="sourceLine" id="cb138-53" data-line-number="53"><span class="st">&quot;tuba8l4&quot;</span>,</a>
<a class="sourceLine" id="cb138-54" data-line-number="54"><span class="st">&quot;hoxd13a&quot;</span>,</a>
<a class="sourceLine" id="cb138-55" data-line-number="55"><span class="st">&quot;LOC100537138&quot;</span>,</a>
<a class="sourceLine" id="cb138-56" data-line-number="56"><span class="st">&quot;nog2&quot;</span>,</a>
<a class="sourceLine" id="cb138-57" data-line-number="57"><span class="st">&quot;h3f3a&quot;</span>,</a>
<a class="sourceLine" id="cb138-58" data-line-number="58"><span class="st">&quot;myclb&quot;</span>,</a>
<a class="sourceLine" id="cb138-59" data-line-number="59"><span class="st">&quot;sult6b1&quot;</span>,</a>
<a class="sourceLine" id="cb138-60" data-line-number="60"><span class="st">&quot;galnt2&quot;</span>,</a>
<a class="sourceLine" id="cb138-61" data-line-number="61"><span class="st">&quot;fgf4&quot;</span>,</a>
<a class="sourceLine" id="cb138-62" data-line-number="62"><span class="st">&quot;foxa&quot;</span>,</a>
<a class="sourceLine" id="cb138-63" data-line-number="63"><span class="st">&quot;bbc3&quot;</span>,</a>
<a class="sourceLine" id="cb138-64" data-line-number="64"><span class="st">&quot;hoxd11a&quot;</span>,</a>
<a class="sourceLine" id="cb138-65" data-line-number="65"><span class="st">&quot;rcn3&quot;</span>,</a>
<a class="sourceLine" id="cb138-66" data-line-number="66"><span class="st">&quot;plk4&quot;</span>,</a>
<a class="sourceLine" id="cb138-67" data-line-number="67"><span class="st">&quot;foxd3&quot;</span>,</a>
<a class="sourceLine" id="cb138-68" data-line-number="68"><span class="st">&quot;pip5k1ca&quot;</span>,</a>
<a class="sourceLine" id="cb138-69" data-line-number="69"><span class="st">&quot;sec22bb&quot;</span>,</a>
<a class="sourceLine" id="cb138-70" data-line-number="70"><span class="st">&quot;slc12a7b&quot;</span>,</a>
<a class="sourceLine" id="cb138-71" data-line-number="71"><span class="st">&quot;marveld1&quot;</span>,</a>
<a class="sourceLine" id="cb138-72" data-line-number="72"><span class="st">&quot;pcp4a&quot;</span></a>
<a class="sourceLine" id="cb138-73" data-line-number="73">)</a>
<a class="sourceLine" id="cb138-74" data-line-number="74"></a>
<a class="sourceLine" id="cb138-75" data-line-number="75"><span class="co">#write.table(nmp_markers, file = &quot;nmp_markergenes.txt&quot;)</span></a>
<a class="sourceLine" id="cb138-76" data-line-number="76">plots &lt;-<span class="st"> </span><span class="kw">VlnPlot</span>(<span class="dt">object =</span> TBcells, <span class="dt">features =</span>nmp_markers, <span class="dt">combine =</span> <span class="ot">FALSE</span>)</a>
<a class="sourceLine" id="cb138-77" data-line-number="77">plots &lt;-<span class="st"> </span><span class="kw">lapply</span>(<span class="dt">X =</span> plots, <span class="dt">FUN =</span> <span class="cf">function</span>(x) x <span class="op">+</span><span class="st">  </span><span class="kw">theme</span>(<span class="dt">plot.title =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">30</span>),</a>
<a class="sourceLine" id="cb138-78" data-line-number="78">                        <span class="dt">text =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">15</span>),</a>
<a class="sourceLine" id="cb138-79" data-line-number="79">                        <span class="dt">axis.text.y =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">12</span>),</a>
<a class="sourceLine" id="cb138-80" data-line-number="80">                        <span class="dt">axis.text.x =</span> <span class="kw">element_text</span>(<span class="dt">size =</span> <span class="dv">12</span>, <span class="dt">angle =</span> <span class="dv">35</span>)))</a>
<a class="sourceLine" id="cb138-81" data-line-number="81"></a>
<a class="sourceLine" id="cb138-82" data-line-number="82"><span class="cf">for</span> (i <span class="cf">in</span> <span class="dv">1</span><span class="op">:</span><span class="kw">length</span>(plots)) {</a>
<a class="sourceLine" id="cb138-83" data-line-number="83">  <span class="kw">ggsave</span>(<span class="dt">plot =</span> plots[[i]], <span class="dt">path =</span> <span class="st">&quot;violin_plots&quot;</span>, <span class="dt">filename =</span> <span class="kw">paste0</span>(<span class="st">&quot;violin_&quot;</span>,nmp_markers[i],<span class="st">&quot;.png&quot;</span>), <span class="dt">width =</span> <span class="dv">14</span>, <span class="dt">height =</span> <span class="dv">10</span>, <span class="dt">units =</span> <span class="st">&quot;cm&quot;</span>)</a>
<a class="sourceLine" id="cb138-84" data-line-number="84">}</a></code></pre></div>
</div>
<div id="saving-rds-object-1" class="section level2">
<h2><span class="header-section-number">7.12</span> Saving RDS object</h2>
<div class="sourceCode" id="cb139"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb139-1" data-line-number="1"><span class="kw">saveRDS</span>(TBcells, <span class="dt">file =</span> <span class="st">&quot;rds/TBcells_subclustered.rds&quot;</span>)</a></code></pre></div>
</div>
<div id="session-info-4" class="section level2">
<h2><span class="header-section-number">7.13</span> Session Info</h2>
<details>
<p><summary>View Session Info</summary></p>
<div class="sourceCode" id="cb140"><pre class="sourceCode r"><code class="sourceCode r"><a class="sourceLine" id="cb140-1" data-line-number="1">devtools<span class="op">::</span><span class="kw">session_info</span>()</a>
<a class="sourceLine" id="cb140-2" data-line-number="2"><span class="co">## - Session info ---------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb140-3" data-line-number="3"><span class="co">##  setting  value                       </span></a>
<a class="sourceLine" id="cb140-4" data-line-number="4"><span class="co">##  version  R version 4.0.0 (2020-04-24)</span></a>
<a class="sourceLine" id="cb140-5" data-line-number="5"><span class="co">##  os       Windows 10 x64              </span></a>
<a class="sourceLine" id="cb140-6" data-line-number="6"><span class="co">##  system   x86_64, mingw32             </span></a>
<a class="sourceLine" id="cb140-7" data-line-number="7"><span class="co">##  ui       RTerm                       </span></a>
<a class="sourceLine" id="cb140-8" data-line-number="8"><span class="co">##  language (EN)                        </span></a>
<a class="sourceLine" id="cb140-9" data-line-number="9"><span class="co">##  collate  English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb140-10" data-line-number="10"><span class="co">##  ctype    English_United Kingdom.1252 </span></a>
<a class="sourceLine" id="cb140-11" data-line-number="11"><span class="co">##  tz       Asia/Kuala_Lumpur           </span></a>
<a class="sourceLine" id="cb140-12" data-line-number="12"><span class="co">##  date     2020-06-03                  </span></a>
<a class="sourceLine" id="cb140-13" data-line-number="13"><span class="co">## </span></a>
<a class="sourceLine" id="cb140-14" data-line-number="14"><span class="co">## - Packages -------------------------------------------------------------------</span></a>
<a class="sourceLine" id="cb140-15" data-line-number="15"><span class="co">##  package              * version  date       lib source        </span></a>
<a class="sourceLine" id="cb140-16" data-line-number="16"><span class="co">##  AnnotationDbi        * 1.50.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-17" data-line-number="17"><span class="co">##  ape                    5.3      2019-03-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-18" data-line-number="18"><span class="co">##  assertthat             0.2.1    2019-03-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-19" data-line-number="19"><span class="co">##  backports              1.1.6    2020-04-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-20" data-line-number="20"><span class="co">##  beeswarm               0.2.3    2016-04-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-21" data-line-number="21"><span class="co">##  bibtex                 0.4.2.2  2020-01-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-22" data-line-number="22"><span class="co">##  Biobase              * 2.48.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-23" data-line-number="23"><span class="co">##  BiocGenerics         * 0.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-24" data-line-number="24"><span class="co">##  BiocNeighbors          1.6.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-25" data-line-number="25"><span class="co">##  BiocParallel           1.22.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-26" data-line-number="26"><span class="co">##  BiocSingular           1.4.0    2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-27" data-line-number="27"><span class="co">##  bit                    1.1-15.2 2020-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-28" data-line-number="28"><span class="co">##  bit64                  0.9-7    2017-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-29" data-line-number="29"><span class="co">##  bitops                 1.0-6    2013-08-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-30" data-line-number="30"><span class="co">##  blob                   1.2.1    2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-31" data-line-number="31"><span class="co">##  bookdown               0.19     2020-05-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-32" data-line-number="32"><span class="co">##  broom                  0.5.6    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-33" data-line-number="33"><span class="co">##  callr                  3.4.3    2020-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-34" data-line-number="34"><span class="co">##  cellranger             1.1.0    2016-07-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-35" data-line-number="35"><span class="co">##  cli                    2.0.2    2020-02-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-36" data-line-number="36"><span class="co">##  cluster              * 2.1.0    2019-06-19 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-37" data-line-number="37"><span class="co">##  codetools              0.2-16   2018-12-24 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-38" data-line-number="38"><span class="co">##  colorspace             1.4-1    2019-03-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-39" data-line-number="39"><span class="co">##  cowplot                1.0.0    2019-07-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-40" data-line-number="40"><span class="co">##  crayon                 1.3.4    2017-09-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-41" data-line-number="41"><span class="co">##  data.table             1.12.8   2019-12-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-42" data-line-number="42"><span class="co">##  DBI                    1.1.0    2019-12-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-43" data-line-number="43"><span class="co">##  dbplyr                 1.4.3    2020-04-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-44" data-line-number="44"><span class="co">##  DelayedArray         * 0.14.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-45" data-line-number="45"><span class="co">##  DelayedMatrixStats     1.10.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-46" data-line-number="46"><span class="co">##  desc                   1.2.0    2018-05-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-47" data-line-number="47"><span class="co">##  devtools               2.3.0    2020-04-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-48" data-line-number="48"><span class="co">##  digest                 0.6.25   2020-02-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-49" data-line-number="49"><span class="co">##  doParallel             1.0.15   2019-08-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-50" data-line-number="50"><span class="co">##  dplyr                * 0.8.5    2020-03-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-51" data-line-number="51"><span class="co">##  dqrng                  0.2.1    2019-05-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-52" data-line-number="52"><span class="co">##  edgeR                  3.30.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-53" data-line-number="53"><span class="co">##  ellipsis               0.3.0    2019-09-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-54" data-line-number="54"><span class="co">##  evaluate               0.14     2019-05-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-55" data-line-number="55"><span class="co">##  extrafont            * 0.17     2014-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-56" data-line-number="56"><span class="co">##  extrafontdb            1.0      2012-06-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-57" data-line-number="57"><span class="co">##  fansi                  0.4.1    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-58" data-line-number="58"><span class="co">##  farver                 2.0.3    2020-01-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-59" data-line-number="59"><span class="co">##  fitdistrplus           1.0-14   2019-01-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-60" data-line-number="60"><span class="co">##  forcats              * 0.5.0    2020-03-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-61" data-line-number="61"><span class="co">##  foreach                1.5.0    2020-03-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-62" data-line-number="62"><span class="co">##  fs                     1.4.1    2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-63" data-line-number="63"><span class="co">##  future                 1.17.0   2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-64" data-line-number="64"><span class="co">##  future.apply           1.5.0    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-65" data-line-number="65"><span class="co">##  generics               0.0.2    2018-11-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-66" data-line-number="66"><span class="co">##  GenomeInfoDb         * 1.24.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-67" data-line-number="67"><span class="co">##  GenomeInfoDbData       1.2.3    2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-68" data-line-number="68"><span class="co">##  GenomicRanges        * 1.40.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-69" data-line-number="69"><span class="co">##  ggbeeswarm             0.6.0    2017-08-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-70" data-line-number="70"><span class="co">##  ggplot2              * 3.3.0    2020-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-71" data-line-number="71"><span class="co">##  ggrepel                0.8.2    2020-03-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-72" data-line-number="72"><span class="co">##  ggridges               0.5.2    2020-01-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-73" data-line-number="73"><span class="co">##  globals                0.12.5   2019-12-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-74" data-line-number="74"><span class="co">##  glue                   1.4.0    2020-04-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-75" data-line-number="75"><span class="co">##  gridBase               0.4-7    2014-02-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-76" data-line-number="76"><span class="co">##  gridExtra            * 2.3      2017-09-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-77" data-line-number="77"><span class="co">##  gtable                 0.3.0    2019-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-78" data-line-number="78"><span class="co">##  gtools                 3.8.2    2020-03-31 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-79" data-line-number="79"><span class="co">##  haven                  2.2.0    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-80" data-line-number="80"><span class="co">##  highr                  0.8      2019-03-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-81" data-line-number="81"><span class="co">##  hms                    0.5.3    2020-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-82" data-line-number="82"><span class="co">##  htmltools              0.4.0    2019-10-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-83" data-line-number="83"><span class="co">##  htmlwidgets            1.5.1    2019-10-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-84" data-line-number="84"><span class="co">##  httr                   1.4.1    2019-08-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-85" data-line-number="85"><span class="co">##  ica                    1.0-2    2018-05-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-86" data-line-number="86"><span class="co">##  igraph                 1.2.5    2020-03-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-87" data-line-number="87"><span class="co">##  IRanges              * 2.22.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-88" data-line-number="88"><span class="co">##  irlba                  2.3.3    2019-02-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-89" data-line-number="89"><span class="co">##  iterators              1.0.12   2019-07-26 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-90" data-line-number="90"><span class="co">##  jsonlite               1.6.1    2020-02-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-91" data-line-number="91"><span class="co">##  KernSmooth             2.23-16  2019-10-15 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-92" data-line-number="92"><span class="co">##  knitr                  1.28     2020-02-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-93" data-line-number="93"><span class="co">##  labeling               0.3      2014-08-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-94" data-line-number="94"><span class="co">##  lattice                0.20-41  2020-04-02 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-95" data-line-number="95"><span class="co">##  lazyeval               0.2.2    2019-03-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-96" data-line-number="96"><span class="co">##  leiden                 0.3.3    2020-02-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-97" data-line-number="97"><span class="co">##  lifecycle              0.2.0    2020-03-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-98" data-line-number="98"><span class="co">##  limma                  3.44.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-99" data-line-number="99"><span class="co">##  listenv                0.8.0    2019-12-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-100" data-line-number="100"><span class="co">##  lmtest                 0.9-37   2019-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-101" data-line-number="101"><span class="co">##  locfit                 1.5-9.4  2020-03-25 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-102" data-line-number="102"><span class="co">##  lsei                   1.2-0.1  2020-05-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-103" data-line-number="103"><span class="co">##  lubridate              1.7.8    2020-04-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-104" data-line-number="104"><span class="co">##  magrittr               1.5      2014-11-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-105" data-line-number="105"><span class="co">##  MASS                   7.3-51.5 2019-12-20 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-106" data-line-number="106"><span class="co">##  Matrix                 1.2-18   2019-11-27 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-107" data-line-number="107"><span class="co">##  matrixStats          * 0.56.0   2020-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-108" data-line-number="108"><span class="co">##  mclust               * 5.4.6    2020-04-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-109" data-line-number="109"><span class="co">##  memoise                1.1.0    2017-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-110" data-line-number="110"><span class="co">##  modelr                 0.1.7    2020-04-30 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-111" data-line-number="111"><span class="co">##  munsell                0.5.0    2018-06-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-112" data-line-number="112"><span class="co">##  nlme                   3.1-147  2020-04-13 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-113" data-line-number="113"><span class="co">##  NMF                  * 0.22.0   2020-02-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-114" data-line-number="114"><span class="co">##  npsurv                 0.4-0.1  2020-05-06 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-115" data-line-number="115"><span class="co">##  org.Dr.eg.db         * 3.11.1   2020-05-11 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-116" data-line-number="116"><span class="co">##  patchwork              1.0.0    2019-12-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-117" data-line-number="117"><span class="co">##  pbapply                1.4-2    2019-08-31 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-118" data-line-number="118"><span class="co">##  pheatmap             * 1.0.12   2019-01-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-119" data-line-number="119"><span class="co">##  pillar                 1.4.4    2020-05-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-120" data-line-number="120"><span class="co">##  pkgbuild               1.0.8    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-121" data-line-number="121"><span class="co">##  pkgconfig              2.0.3    2019-09-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-122" data-line-number="122"><span class="co">##  pkgload                1.0.2    2018-10-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-123" data-line-number="123"><span class="co">##  pkgmaker             * 0.31.1   2020-03-19 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-124" data-line-number="124"><span class="co">##  plotly                 4.9.2.1  2020-04-04 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-125" data-line-number="125"><span class="co">##  plyr                   1.8.6    2020-03-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-126" data-line-number="126"><span class="co">##  png                    0.1-7    2013-12-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-127" data-line-number="127"><span class="co">##  prettyunits            1.1.1    2020-01-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-128" data-line-number="128"><span class="co">##  processx               3.4.2    2020-02-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-129" data-line-number="129"><span class="co">##  ps                     1.3.3    2020-05-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-130" data-line-number="130"><span class="co">##  purrr                * 0.3.4    2020-04-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-131" data-line-number="131"><span class="co">##  R.methodsS3          * 1.8.0    2020-02-14 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-132" data-line-number="132"><span class="co">##  R.oo                 * 1.23.0   2019-11-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-133" data-line-number="133"><span class="co">##  R.utils              * 2.9.2    2019-12-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-134" data-line-number="134"><span class="co">##  R6                     2.4.1    2019-11-12 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-135" data-line-number="135"><span class="co">##  RANN                   2.6.1    2019-01-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-136" data-line-number="136"><span class="co">##  rappdirs               0.3.1    2016-03-28 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-137" data-line-number="137"><span class="co">##  RColorBrewer         * 1.1-2    2014-12-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-138" data-line-number="138"><span class="co">##  Rcpp                   1.0.4.6  2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-139" data-line-number="139"><span class="co">##  RcppAnnoy              0.0.16   2020-03-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-140" data-line-number="140"><span class="co">##  RCurl                  1.98-1.2 2020-04-18 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-141" data-line-number="141"><span class="co">##  readr                * 1.3.1    2018-12-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-142" data-line-number="142"><span class="co">##  readxl                 1.3.1    2019-03-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-143" data-line-number="143"><span class="co">##  registry             * 0.5-1    2019-03-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-144" data-line-number="144"><span class="co">##  remotes                2.1.1    2020-02-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-145" data-line-number="145"><span class="co">##  reprex                 0.3.0    2019-05-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-146" data-line-number="146"><span class="co">##  reshape2               1.4.4    2020-04-09 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-147" data-line-number="147"><span class="co">##  reticulate             1.15     2020-04-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-148" data-line-number="148"><span class="co">##  rlang                  0.4.6    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-149" data-line-number="149"><span class="co">##  rmarkdown              2.1      2020-01-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-150" data-line-number="150"><span class="co">##  rngtools             * 1.5      2020-01-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-151" data-line-number="151"><span class="co">##  ROCR                   1.0-11   2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-152" data-line-number="152"><span class="co">##  rprojroot              1.3-2    2018-01-03 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-153" data-line-number="153"><span class="co">##  RSpectra               0.16-0   2019-12-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-154" data-line-number="154"><span class="co">##  RSQLite                2.2.0    2020-01-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-155" data-line-number="155"><span class="co">##  rstudioapi             0.11     2020-02-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-156" data-line-number="156"><span class="co">##  rsvd                   1.0.3    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-157" data-line-number="157"><span class="co">##  Rtsne                  0.15     2018-11-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-158" data-line-number="158"><span class="co">##  Rttf2pt1               1.3.8    2020-01-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-159" data-line-number="159"><span class="co">##  rvest                  0.3.5    2019-11-08 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-160" data-line-number="160"><span class="co">##  S4Vectors            * 0.26.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-161" data-line-number="161"><span class="co">##  scales               * 1.1.1    2020-05-11 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-162" data-line-number="162"><span class="co">##  scater               * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-163" data-line-number="163"><span class="co">##  scran                * 1.16.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-164" data-line-number="164"><span class="co">##  sctransform            0.2.1    2019-12-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-165" data-line-number="165"><span class="co">##  sessioninfo            1.1.1    2018-11-05 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-166" data-line-number="166"><span class="co">##  Seurat               * 3.1.5    2020-04-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-167" data-line-number="167"><span class="co">##  SingleCellExperiment * 1.10.1   2020-04-28 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-168" data-line-number="168"><span class="co">##  statmod                1.4.34   2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-169" data-line-number="169"><span class="co">##  stringi                1.4.6    2020-02-17 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-170" data-line-number="170"><span class="co">##  stringr              * 1.4.0    2019-02-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-171" data-line-number="171"><span class="co">##  SummarizedExperiment * 1.18.1   2020-04-30 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-172" data-line-number="172"><span class="co">##  survival               3.1-12   2020-04-10 [2] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-173" data-line-number="173"><span class="co">##  testthat               2.3.2    2020-03-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-174" data-line-number="174"><span class="co">##  tibble               * 3.0.1    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-175" data-line-number="175"><span class="co">##  tidyr                * 1.0.3    2020-05-07 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-176" data-line-number="176"><span class="co">##  tidyselect             1.0.0    2020-01-27 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-177" data-line-number="177"><span class="co">##  tidyverse            * 1.3.0    2019-11-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-178" data-line-number="178"><span class="co">##  tsne                   0.1-3    2016-07-15 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-179" data-line-number="179"><span class="co">##  usethis                1.6.1    2020-04-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-180" data-line-number="180"><span class="co">##  utf8                   1.1.4    2018-05-24 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-181" data-line-number="181"><span class="co">##  uwot                   0.1.8    2020-03-16 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-182" data-line-number="182"><span class="co">##  vctrs                  0.2.4    2020-03-10 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-183" data-line-number="183"><span class="co">##  vipor                  0.4.5    2017-03-22 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-184" data-line-number="184"><span class="co">##  viridis              * 0.5.1    2018-03-29 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-185" data-line-number="185"><span class="co">##  viridisLite          * 0.3.0    2018-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-186" data-line-number="186"><span class="co">##  withr                  2.2.0    2020-04-20 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-187" data-line-number="187"><span class="co">##  xfun                   0.13     2020-04-13 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-188" data-line-number="188"><span class="co">##  xml2                   1.3.2    2020-04-23 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-189" data-line-number="189"><span class="co">##  xtable                 1.8-4    2019-04-21 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-190" data-line-number="190"><span class="co">##  XVector                0.28.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-191" data-line-number="191"><span class="co">##  yaml                   2.2.1    2020-02-01 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-192" data-line-number="192"><span class="co">##  zlibbioc               1.34.0   2020-04-27 [1] Bioconductor  </span></a>
<a class="sourceLine" id="cb140-193" data-line-number="193"><span class="co">##  zoo                    1.8-8    2020-05-02 [1] CRAN (R 4.0.0)</span></a>
<a class="sourceLine" id="cb140-194" data-line-number="194"><span class="co">## </span></a>
<a class="sourceLine" id="cb140-195" data-line-number="195"><span class="co">## [1] D:/R_packages</span></a>
<a class="sourceLine" id="cb140-196" data-line-number="196"><span class="co">## [2] C:/Program Files/R/R-4.0.0/library</span></a></code></pre></div>
</details>
</div>
<div id="solutions-3" class="section level2">
<h2><span class="header-section-number">7.14</span> Solutions</h2>
<ol style="list-style-type: decimal">
<li>Use the ggsave function.</li>
</ol>
<ul>
<li>ggsave(“2g)nm_dotplot.tiff”, nm_dotplot, dpi = 400, width = 12, height = 8,units = c(“cm”))</li>
<li>ggsave(“2g)meso_dotplot.tiff”, meso_dotplot, dpi = 400, width = 12, height = 8,units = c(“cm”))</li>
<li>ggsave(“2g)neural_dotplot.tiff”, neural_dotplot, dpi = 400, width = 12, height = 8,units = c(“cm”))</li>
</ul>
<ol start="2" style="list-style-type: decimal">
<li>lapply returns a list of the same length as the input vector. sapply (simple apply) returns a vector.</li>
</ol>

</div>
</div>
            </section>

          </div>
        </div>
      </div>
<a href="differentially-expressed-genes.html" class="navigation navigation-prev " aria-label="Previous page"><i class="fa fa-angle-left"></i></a>
<a href="pseudotime.html" class="navigation navigation-next " aria-label="Next page"><i class="fa fa-angle-right"></i></a>
    </div>
  </div>
<script src="libs/gitbook-2.6.7/js/app.min.js"></script>
<script src="libs/gitbook-2.6.7/js/lunr.js"></script>
<script src="libs/gitbook-2.6.7/js/clipboard.min.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-search.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-sharing.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-fontsettings.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-bookdown.js"></script>
<script src="libs/gitbook-2.6.7/js/jquery.highlight.js"></script>
<script src="libs/gitbook-2.6.7/js/plugin-clipboard.js"></script>
<script>
gitbook.require(["gitbook"], function(gitbook) {
gitbook.start({
"sharing": {
"github": false,
"facebook": true,
"twitter": true,
"linkedin": false,
"weibo": false,
"instapaper": false,
"vk": false,
"all": ["facebook", "twitter", "linkedin", "weibo", "instapaper"]
},
"fontsettings": {
"theme": "white",
"family": "sans",
"size": 2
},
"edit": {
"link": null,
"text": null
},
"history": {
"link": null,
"text": null
},
"view": {
"link": null,
"text": null
},
"download": ["NMps_book.pdf", "NMps_book.epub"],
"toc": {
"collapse": "subsection"
}
});
});
</script>

<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    var src = "true";
    if (src === "" || src === "true") src = "https://mathjax.rstudio.com/latest/MathJax.js?config=TeX-MML-AM_CHTML";
    if (location.protocol !== "file:")
      if (/^https?:/.test(src))
        src = src.replace(/^https?:/, '');
    script.src = src;
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>
</body>

</html>
